{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CNS AP-axis VAE model\n",
    "\n",
    "This notebook runs through the CNS analysis used in the VAE paper (ref: ).  \n",
    "\n",
    "We use genes that were found to be differentially expressed and aim to understand how different groups of genes responded to PRC2 knockout.  \n",
    "\n",
    "We do this by using a variational autoencoder (VAE).  \n",
    "\n",
    "Our VAE model uses a three layer network, and as input data, takes in the normalised expression matrix including the wild type and response to Eed-cKO.  \n",
    "\n",
    "\n",
    "### Sections:\n",
    "\n",
    "#### Part 1: VAE setup\n",
    "    1) Function setup\n",
    "    2) Read in processed dataframe and organise this\n",
    "    3) Build VAE model\n",
    "    \n",
    "#### Part 2: VAE as a ranking method\n",
    "    4) Setup data for visualisations\n",
    "    5) Quantify separability of latent space\n",
    "    6) Save VAE ranks to csv to run GSEA in R\n",
    "    \n",
    "#### Part 3: VAE to identify coordinating groups of genes\n",
    "    6) Build gene groups using VAE latent space\n",
    "    7) Save gene groups for ORA in R\n",
    "    8) Inspecting the groups in context of chromHMM annotations \n",
    "    9) Inspect groups in terms of annotations for Epi paper\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "--------------------------------------------------------\n",
    "                     Imports\n",
    "--------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "import os, sys\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "import venn\n",
    "from matplotlib_venn import venn3\n",
    "from sklearn.decomposition import PCA\n",
    "from scipy import stats\n",
    "from sciviso import *\n",
    "from sciutil import SciUtil\n",
    "from scivae import Vis as VVis\n",
    "import string\n",
    "u = SciUtil()\n",
    "\n",
    "module_path = os.path.abspath(os.path.join(''))\n",
    "sys.path.append(module_path)\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "--------------------------------------------------------\n",
    "                     Global variables\n",
    "--------------------------------------------------------\n",
    "\"\"\"\n",
    "date = '20210217'\n",
    "\n",
    "gene_id = 'entrezgene_id'\n",
    "gene_name = 'external_gene_name'\n",
    "\n",
    "sci_colour = ['#483873', '#1BD8A6', '#B117B7', '#AAC7E2', '#FFC107', '#016957', '#9785C0', \n",
    "             '#D09139', '#338A03', '#FF69A1', '#5930B1', '#FFE884', '#35B567', '#1E88E5', \n",
    "             '#ACAD60', '#A2FFB4', '#B618F5', '#854A9C']\n",
    "\n",
    "hist_colour = '#483873'\n",
    "\n",
    "e11_colour = '#BEC0C2'\n",
    "e13_colour = '#A7A9AC'\n",
    "e15_colour = '#7D7E81'\n",
    "e18_colour = '#58595B'\n",
    "\n",
    "fb_colour = '#ffbf80'\n",
    "mb_colour = '#ff8c1a'\n",
    "hb_colour = '#b35900'\n",
    "sc_colour = '#663300'\n",
    "fb_color = '#ffbf80'\n",
    "mb_color = '#ff8c1a'\n",
    "hb_color = '#b35900'\n",
    "sc_color = '#663300'\n",
    "\n",
    "h3k36me3_colour = '#CEF471'\n",
    "h3k27me3_colour = '#9F71F4'\n",
    "h3k4me3_colour = '#9F00FA'\n",
    "h3k4me2_colour = '#5930B1'\n",
    "h3k4me1_colour = '#FFE884'\n",
    "h3k27ac_colour = '#35B567'\n",
    "h3k9me3_colour = '#1E88E5'\n",
    "h3k9ac_colour = '#A2FFB4'\n",
    "           \n",
    "wt_colour = '#AADFF1'\n",
    "ko_colour = '#A53736'\n",
    "\n",
    "sns.palplot(sci_colour)\n",
    "sns.color_palette(sci_colour)\n",
    "\n",
    "project_name = '6_node_consistent_genes'\n",
    "\n",
    "data_dir = '../../data/'\n",
    "r_dir = f'{data_dir}results/deseq2/'\n",
    "fig_dir = f'../../figures/{project_name}/vae/'\n",
    "output_dir = f'{data_dir}results/{project_name}/vae/'\n",
    "input_dir = f'{data_dir}results/prelim/'\n",
    "rna_dir = f'{input_dir}feature-counts_04052020/'\n",
    "supp_dir = f'{data_dir}input/supps/'\n",
    "logging_dir = 'logging/'\n",
    "ora_dir = '../../data/results/6_node_consistent_genes/functional/'\n",
    "\n",
    "grey = '#bdbdbd'\n",
    "hist_cmap = 'Greens'\n",
    "rna_cmap = 'Purples'\n",
    "div_cmap = 'seismic'\n",
    "\n",
    "experiment_name = project_name\n",
    "from_saved = True  # Load from a saved VAE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = pd.read_csv(f'{input_dir}df-all_epi-2500_20210124.csv')\n",
    "df_training = pd.read_csv(f'{input_dir}df-consistent_epi-2500_20210124.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1) Function setup\n",
    "\n",
    "Here we just initialise the functions that will be used in the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "--------------------------------------------------------\n",
    "                     Colour/style getters\n",
    "--------------------------------------------------------\n",
    "\"\"\"\n",
    "def get_time_colour(c):\n",
    "    if '11' in c or '10' in c:\n",
    "        return e11_colour\n",
    "    elif '13' in c or '12' in c:\n",
    "        return e13_colour\n",
    "    elif '15' in c or '14' in c:\n",
    "        return e15_colour\n",
    "    elif '18' in c or '16' in c:\n",
    "        return e18_colour\n",
    "    return '#FFFFFF'\n",
    "\n",
    "def get_tissue_colour(c):\n",
    "    if 'sc' in c or 'spinal' in c:\n",
    "        return sc_colour\n",
    "    elif 'hb' in c or 'hindbrain' in c:\n",
    "        return hb_colour\n",
    "    elif 'fb' in c or 'forebrain' in c:\n",
    "        return fb_colour\n",
    "    elif 'mb' in c or 'midbrain' in c:\n",
    "        return mb_colour\n",
    "    return '#FFFFFF'\n",
    "\n",
    "def get_cond_colour(c):\n",
    "    if 'ko' in c:\n",
    "        return ko_colour\n",
    "    elif 'wt' in c:\n",
    "        return wt_colour\n",
    "    return '#FFFFFF'\n",
    "\n",
    "def get_mark_colour(c):\n",
    "    if '36me3' in c:\n",
    "        return h3k36me3_colour\n",
    "    elif '27me3' in c:\n",
    "        return h3k27me3_colour\n",
    "    elif 'K4me3' in c:\n",
    "        return h3k4me3_colour\n",
    "    elif 'K4me2' in c:\n",
    "        return h3k4me2_colour\n",
    "    elif 'K4me1' in c:\n",
    "        return h3k4me1_colour\n",
    "    elif '27ac' in c:\n",
    "        return h3k27ac_colour\n",
    "    elif 'K9me3' in c:\n",
    "        return h3k9me3_colour\n",
    "    elif 'K9ac' in c:\n",
    "        return h3k9ac_colour\n",
    "    return '#FFFFFF'\n",
    "\n",
    "\"\"\"\n",
    "--------------------------------------------------------\n",
    "                     Plotting\n",
    "--------------------------------------------------------\n",
    "\"\"\"\n",
    "def pplot():\n",
    "    return\n",
    "        \n",
    "def cplot():\n",
    "    return\n",
    "    \n",
    "def save_fig(title):\n",
    "    plt.savefig(f'{fig_dir}{title.replace(\" \", \"-\")}_{experiment_name}_{date}.svg')\n",
    "    \n",
    "\"\"\"\n",
    "--------------------------------------------------------\n",
    "                     Heatmap and vis function\n",
    "--------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "def plot_gene_heatmap(df, fb_genes, mb_genes, hb_genes, sc_genes, \n",
    "                      input_cols, vae_data, vae, gene_name, mark, method='input', \n",
    "                      title=\"\", merge_reps=False, input_col_ids=None, cmap='', genes=None, \n",
    "                      values=None):\n",
    "    brain_genes = []\n",
    "    colours = []\n",
    "    gene_idxs = []\n",
    "    genes = genes if genes is not None else df[gene_name].values\n",
    "    i = 0\n",
    "    for g in genes:\n",
    "        if g in fb_genes:\n",
    "            brain_genes.append(g)\n",
    "            colours.append(fb_colour)\n",
    "            gene_idxs.append(i)\n",
    "        i += 1\n",
    "    i = 0\n",
    "    for g in genes:    \n",
    "        if g in mb_genes:\n",
    "            brain_genes.append(g)\n",
    "            colours.append(mb_colour)\n",
    "            gene_idxs.append(i)\n",
    "        i += 1\n",
    "    i = 0\n",
    "    for g in genes:    \n",
    "        if g in hb_genes:\n",
    "            brain_genes.append(g)\n",
    "            colours.append(hb_colour)\n",
    "            gene_idxs.append(i)\n",
    "        i += 1\n",
    "    i = 0\n",
    "    for g in genes:    \n",
    "        if g in sc_genes:\n",
    "            brain_genes.append(g)\n",
    "            colours.append(sc_colour)\n",
    "            gene_idxs.append(i)\n",
    "        i += 1\n",
    "    heatmap_df = pd.DataFrame()\n",
    "    heatmap_df[gene_name] = brain_genes\n",
    "\n",
    "    if method == 'decoding':\n",
    "        deocded = vae.decoder.predict(vae_data)\n",
    "        deocded = scaler.fit_transform(deocded)\n",
    "        vals = deocded[gene_idxs]\n",
    "    elif method == 'input':\n",
    "        vals = df[input_cols].values[gene_idxs]\n",
    "    else:\n",
    "        vals = vae_data[gene_idxs]\n",
    "    vals = values[gene_idxs] if values is not None else vals\n",
    "    \n",
    "    vals = np.nan_to_num(vals)\n",
    "    min_r = np.min(vals, axis = 1)\n",
    "    max_r = np.max(vals, axis = 1)\n",
    "    ov_c = max_r - min_r\n",
    "    if not merge_reps:\n",
    "        cols = []\n",
    "        for i, c in enumerate(input_cols):\n",
    "            if input_col_ids is not None:\n",
    "                if i in input_col_ids:\n",
    "                    heatmap_df[c] = np.nan_to_num(vals[:,i])\n",
    "                    cols.append(c)\n",
    "            else:\n",
    "                heatmap_df[c] = np.nan_to_num(vals[:,i])\n",
    "                cols.append(c)\n",
    "        heatmap_cols = cols\n",
    "    else:\n",
    "        i = 0\n",
    "        offset = 0\n",
    "        max_c = len(input_cols)\n",
    "        if input_col_ids is not None:\n",
    "            offset = input_col_ids[0]\n",
    "            max_c = len(input_col_ids)\n",
    "        heatmap_cols = []\n",
    "        \n",
    "        while(i < max_c):\n",
    "            if input_col_ids is not None:\n",
    "                heatmap_df[input_cols[input_col_ids[i]]] = (vals[:, input_col_ids[i]] - min_r)/(max_r - min_r)\n",
    "                heatmap_cols.append(input_cols[input_col_ids[i]])\n",
    "            else:\n",
    "                heatmap_df[input_cols[i]] = (vals[:, i + offset] - min_r)/(max_r - min_r)\n",
    "                heatmap_cols.append(input_cols[i])\n",
    "            i += 1\n",
    "            \n",
    "    print(heatmap_df.head())\n",
    "    cols = []\n",
    "    col_colours = []\n",
    "    cond_colours = []\n",
    "    tissue_colours = []\n",
    "    time_colours = []\n",
    "    for c in heatmap_df.columns:\n",
    "        if 'wt' in c or 'ko' in c:\n",
    "            cols.append(c)\n",
    "            time_colours.append(get_time_colour(c))\n",
    "            tissue_colours.append(get_tissue_colour(c))\n",
    "            cond_colours.append(get_cond_colour(c))\n",
    "        if mark in c and hist_metric in c:\n",
    "            time_colours.append(get_time_colour(c))\n",
    "            tissue_colours.append(get_tissue_colour(c))\n",
    "            cond_colours.append('white')\n",
    "            cols.append(c)\n",
    "        elif 'log2' in c:\n",
    "            time_colours.append(get_time_colour(c))\n",
    "            tissue_colours.append(get_tissue_colour(c))\n",
    "            cond_colours.append(get_mark_colour(c))\n",
    "            cols.append(c)\n",
    "    col_colours = [cond_colours, tissue_colours, time_colours]\n",
    "\n",
    "    # Now plot the rnaseq response as a heatmap\n",
    "    vmin = None\n",
    "    vmax = None\n",
    "    if cmap == hist_cmap or cmap == rna_cmap:\n",
    "        vmin = 0\n",
    "        vmax = 10\n",
    "    elif cmap == div_cmap:\n",
    "        vmin = -3\n",
    "        vmax = 3\n",
    "    heatmap = Heatmap(heatmap_df, heatmap_cols, gene_name,\n",
    "                      title=title, cluster_cols=False, cluster_rows=False, vmin=vmin, vmax=vmax,\n",
    "                      figsize=(5, 5),\n",
    "                      cmap=cmap)\n",
    "    heatmap.cmap = cmap\n",
    "    heatmap.plot()\n",
    "    pplot()\n",
    "    save_fig(title)\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2) Read in processed dataframe\n",
    "\n",
    "Here we import the results from running the AP-axis dataset generation script.  \n",
    "\n",
    "This used the output from differential analysis and also annotated histone marks to genes.  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove genes that weren't able to map to a gene name\n",
    "gene_mask = []\n",
    "for g in df_training[gene_name].values:\n",
    "    if g == None:\n",
    "        gene_mask.append(0)\n",
    "    else:\n",
    "        try:\n",
    "            e = float(g)\n",
    "            gene_mask.append(1)\n",
    "        except ValueError:\n",
    "            gene_mask.append(0)\n",
    "\n",
    "df_training['gene_mask'] = gene_mask\n",
    "print(len(df_training))\n",
    "print(len(df_training[df_training['gene_mask'] == 0]))\n",
    "df_training = df_training[df_training['gene_mask'] == 0]\n",
    "\n",
    "# Smooth out the columns in the data frame i.e. for the clones we only put in the mean of the two replicates\n",
    "\n",
    "cols_to_merge = [c for c in df_training.columns if 'wt' in c or 'ko' in c]\n",
    "col_names = []\n",
    "col_values = []\n",
    "values = df_training[cols_to_merge].values\n",
    "i = 0\n",
    "while(i < len(cols_to_merge)):\n",
    "    df_training[f'{cols_to_merge[i][:-1]}_merged-rep'] = 0.5 * (df_training[cols_to_merge[i]].values +\n",
    "                                                                df_training[cols_to_merge[i + 1]].values)\n",
    "    print(\"merged\", cols_to_merge[i], cols_to_merge[i + 1])\n",
    "    i += 2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols_to_merge = [c for c in df_all.columns if 'wt' in c or 'ko' in c]\n",
    "col_names = []\n",
    "col_values = []\n",
    "values = df_all[cols_to_merge].values\n",
    "i = 0\n",
    "while(i < len(cols_to_merge)):\n",
    "    df_all[f'{cols_to_merge[i][:-1]}_merged-rep'] = 0.5 * (df_all[cols_to_merge[i]].values +\n",
    "                                                                df_all[cols_to_merge[i + 1]].values)\n",
    "    print(\"merged\", cols_to_merge[i], cols_to_merge[i + 1])\n",
    "    i += 2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3) Build VAE model\n",
    "\n",
    "Since the VAEs are stochastic rather than deterministic we want to create our clusters based on a number of iterations.\n",
    "We also are interested in how much the genes differ in various latent spaces."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scivae import VAE\n",
    "\n",
    "hist_metric = 'signal'\n",
    "labels = df_training[gene_name].values\n",
    "\n",
    "\n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Set the input data space\n",
    "# -----------------------------------------------------------------------------------\n",
    "def get_input_data(df: pd.DataFrame, label_col: str):\n",
    "    scaler = MinMaxScaler(copy=True)\n",
    "    \"\"\" Here we define the data for the VAE we choose to log2 the signal \"\"\"\n",
    "    cols_bin = []\n",
    "    tmp_df = pd.DataFrame()\n",
    "    non_nan_df = df.copy()\n",
    "    non_nan_df = non_nan_df.fillna(0)\n",
    "    for c in df.columns:\n",
    "        if 'H3K27me3' in c and 'signal' in c and 'brain' in c:\n",
    "            v = np.log2(non_nan_df[c].values + 1)\n",
    "            nn_min = np.nanmin(v)\n",
    "            nn_max = np.nanmax(v)\n",
    "            tmp_df[f'{c}_log2'] = (v-nn_min) / (nn_max - nn_min)\n",
    "            cols_bin.append(f'{c}_log2')\n",
    "        elif ('merged-rep' not in c) and ('log2FoldChange' in c or 'wt' in c or 'ko' in c): \n",
    "            cols_bin.append(c)\n",
    "            v = non_nan_df[c].values\n",
    "            nn_min = np.nanmin(v)\n",
    "            nn_max = np.nanmax(v)\n",
    "            tmp_df[c] = (v-nn_min) / (nn_max - nn_min)\n",
    "    vae_values = tmp_df[cols_bin].values\n",
    "    vae_values = np.nan_to_num(vae_values)\n",
    "    tmp_df[label_col] = df[label_col].values\n",
    "    return vae_values, tmp_df, df[label_col].values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # -----------------------------------------------------------------------------------\n",
    "# #                         Run a test to work out the best number of nodes\n",
    "# # -----------------------------------------------------------------------------------\n",
    "# vae_input_values, df_input, labels = get_input_data(df_training, gene_name)\n",
    "# tsts = [1, 2, 3, 4, 6, 8, 16, 24]\n",
    "# for i in tsts:\n",
    "#     for j in range(0, 5):\n",
    "#         config = {'loss': {'loss_type': 'mse', 'distance_metric': 'mmd', 'mmd_weight': 2.0}, \n",
    "#                   'encoding': {'layers': [{'num_nodes': 64, 'activation_fn': 'selu'}, \n",
    "#                                           {'num_nodes': 32, 'activation_fn': 'relu'}]}, \n",
    "#                   'decoding': {'layers': [{'num_nodes': 32, 'activation_fn': 'relu'},\n",
    "#                                           {'num_nodes': 64, 'activation_fn': 'selu'}]}, \n",
    "#                   'latent': {'num_nodes': i}, 'optimiser': {'params': {}, 'name': 'adam'}}\n",
    "\n",
    "#         vae_mse = VAE(vae_input_values, vae_input_values, labels, config, f'vae_num-nodes_{i}_rep_{j}')\n",
    "#         vae_mse.encode('default', epochs=100, batch_size=50,\n",
    "#                        logging_dir=logging_dir, logfile=f'run_num-nodes_{i}_rep_{j}.csv')\n",
    "\n",
    "# # -----------------------------------------------------------------------------------\n",
    "# #                         Plot results\n",
    "# # -----------------------------------------------------------------------------------\n",
    "\n",
    "# # Read in each of the csv's and then add a line for each one\n",
    "# tsts = [1, 2, 3, 4, 6, 8, 16, 24]\n",
    "# fig, ax = plt.subplots()\n",
    "# labels = []\n",
    "# for i, t in enumerate(tsts):\n",
    "#     c = sci_colour[i]\n",
    "#     print(c, i)\n",
    "#     for j in range(0, 5):\n",
    "#         d = pd.read_csv(f'{logging_dir}run_num-nodes_{t}_rep_{j}.csv')\n",
    "#         print(f'{logging_dir}run_num-nodes_{t}_rep_{j}.csv')\n",
    "#         ax.plot(d['epoch'].values, d['loss'].values, c=c, alpha=0.5)\n",
    "#         label=f'{t} node(s)'\n",
    "#         if label not in labels:\n",
    "#             ax.plot(d['epoch'].values, d['val_loss'].values, c=c, alpha=0.8, label=label)\n",
    "#             labels.append(label)\n",
    "#         else:\n",
    "#             ax.plot(d['epoch'].values, d['val_loss'].values, c=c, alpha=0.8)\n",
    "        \n",
    "# ax.legend(loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "# axes = plt.gca()\n",
    "# axes.set_ylim([0, 5])\n",
    "# save_fig(f'node_num')\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -----------------------------------------------------------------------------------\n",
    "#                         Based on the above, set the number of nodes\n",
    "#                         1) Build non-linear VAE\n",
    "#                         2) Build linear VAE\n",
    "# -----------------------------------------------------------------------------------\n",
    "num_nodes = 6\n",
    "vae_input_values, df_input, labels = get_input_data(df_training, gene_name)\n",
    "\n",
    "config = {'loss': {'loss_type': 'mse', 'distance_metric': 'mmd', 'mmd_weight': 1.0, \n",
    "                   'mmcd_method': 'k'}, \n",
    "                  'encoding': {'layers': [{'num_nodes': 64, 'activation_fn': 'selu'}, \n",
    "                                          {'num_nodes': 32, 'activation_fn': 'relu'}\n",
    "                                          ]}, \n",
    "                  'decoding': {'layers': [\n",
    "                                          {'num_nodes': 32, 'activation_fn': 'relu'},\n",
    "                                          {'num_nodes': 64, 'activation_fn': 'selu'}]}, \n",
    "                  'latent': {'num_nodes': num_nodes}, 'optimiser': {'params': {}, 'name': 'adam'}}\n",
    "\n",
    "vae_mse = VAE(vae_input_values, vae_input_values, labels, config, f'final')\n",
    "if not from_saved:\n",
    "    vae_mse.encode('default', epochs=250, batch_size=50,\n",
    "                   logging_dir=\"final_vae_logs\", logfile=f'final.csv')\n",
    "    vae_mse.save() # save the VAE\n",
    "    vae_mse.u.dp([\"Saved VAE to current directory.\"])\n",
    "else:\n",
    "    vae_mse.u.dp([\"Trying to load saved VAE...\"])\n",
    "    vae_mse.load()\n",
    "    vae_mse.u.dp([\"Loaded saved VAE :) \"])\n",
    "\n",
    "df = df_training.copy()\n",
    "\n",
    "# Encode the values using the VAE\n",
    "scaler = MinMaxScaler(copy=True)\n",
    "scaled_vals = scaler.fit_transform(vae_input_values)\n",
    "\n",
    "vae_data = vae_mse.encode_new_data(scaled_vals)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4) Setup data for visualisations\n",
    "\n",
    "Here we setup the data and summarise key features, for example calcualting the median values. \n",
    "\n",
    "We do this so we can easily access these later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -----------------------------------------------------------------------------------\n",
    "#                         Data setup for Vis\n",
    "# -----------------------------------------------------------------------------------\n",
    "\n",
    "hist_metric = 'signal'\n",
    "input_cols = list(df_input.columns)\n",
    "\n",
    "\n",
    "def get_median(df, mark, time=\"\", tissue=\"brain\"):\n",
    "    cols = []\n",
    "    for c in df.columns:\n",
    "        if tissue in c and mark in c and hist_metric in c and time in c and 'median' not in c:\n",
    "            cols.append(c)\n",
    "    # get nan median\n",
    "    vals = np.nanmedian(df[cols].values, axis=1)\n",
    "    \n",
    "    return np.nan_to_num(vals)\n",
    "\n",
    "marks = ['H3K27me3', 'H3K4me1', 'H3K4me2', 'H3K4me3', 'H3K27ac', 'H3K9me3', 'H3K36me3', 'H3K9ac']\n",
    "\n",
    "# Get the median over all time points and all brain tissues\n",
    "for m in marks:\n",
    "    df[f'{m}'] = get_median(df, m, time=\"\", tissue=\"brain\")\n",
    "    \n",
    "times = ['10.5', '11.5', '12.5', '13.5', '14.5', '15.5', '16.5']\n",
    "for m in marks:\n",
    "    for e in times:\n",
    "        df[f'median_brain_{e}_signal_{m}'] = get_median(df, m, time=e, tissue=\"brain\")\n",
    "    \n",
    "tissues = ['hindbrain', 'midbrain', 'forebrain']\n",
    "for m in marks:\n",
    "    for e in times:\n",
    "        for t in tissues:\n",
    "            df[f'median_{t}_{e}_signal_{m}'] = get_median(df, m, time=e, tissue=t)\n",
    "\n",
    "\n",
    "df['mean_wt'] = np.mean(df[[c for c in df.columns if 'wt' in c]].values, axis=1)\n",
    "df['mean_ko'] = np.mean(df[[c for c in df.columns if 'ko' in c]].values, axis=1)\n",
    "df['mean_ko-wt'] = df['mean_ko'].values - df['mean_wt'].values\n",
    "\n",
    "# Add in the VAE columns \n",
    "for v in range(0, len(vae_data[0])):\n",
    "    df[f'VAE{v}'] = vae_data[:, v]\n",
    "\n",
    "# Rename/add some columns just to make figs easier\n",
    "df['Log2FC FB'] = df['log2FoldChange_fb'].values\n",
    "df['Log2FC MB'] = df['log2FoldChange_mb'].values\n",
    "df['Log2FC HB'] = df['log2FoldChange_hb'].values\n",
    "df['Log2FC SC'] = df['log2FoldChange_sc'].values\n",
    "df['Log2FC A11'] = df['log2FoldChange_a11'].values\n",
    "df['Log2FC A13'] = df['log2FoldChange_a13'].values\n",
    "df['Log2FC A15'] = df['log2FoldChange_a15'].values\n",
    "df['Log2FC A18'] = df['log2FoldChange_a18'].values\n",
    "df['Log2FC P11'] = df['log2FoldChange_p11'].values\n",
    "df['Log2FC P13'] = df['log2FoldChange_p13'].values\n",
    "df['Log2FC P15'] = df['log2FoldChange_p15'].values\n",
    "df['Log2FC P18'] = df['log2FoldChange_p18'].values\n",
    "\n",
    "all_cols = [\n",
    "    'H3K9me3',\n",
    "    'H3K4me1', \n",
    "    'H3K4me2', \n",
    "    'H3K4me3', \n",
    "    'H3K27ac',\n",
    "    'H3K36me3', \n",
    "    'H3K9ac',\n",
    "    'mean_wt',\n",
    "    'mean_ko',\n",
    "    'mean_ko-wt',\n",
    "    'H3K27me3',\n",
    "    'Log2FC FB',\n",
    "    'Log2FC MB',\n",
    "    'Log2FC HB',\n",
    "    'Log2FC SC',\n",
    "    'VAE0',\n",
    "    'VAE1',\n",
    "    'VAE2',\n",
    "    'Log2FC A11',\n",
    "    'Log2FC A13',\n",
    "    'Log2FC A15',\n",
    "    'Log2FC A18',\n",
    "    'Log2FC P11',\n",
    "    'Log2FC P13',\n",
    "    'Log2FC P15',\n",
    "    'Log2FC P18'\n",
    "]\n",
    "unobserved_cols = [\n",
    "    'H3K9me3',\n",
    "    'H3K4me1', \n",
    "    'H3K4me2', \n",
    "    'H3K4me3', \n",
    "    'H3K27ac',\n",
    "    'H3K36me3', \n",
    "    'H3K9ac',\n",
    "    'mean_wt',\n",
    "    'mean_ko',\n",
    "    'mean_ko-wt'\n",
    "]\n",
    "\n",
    "observed_cols = [\n",
    "    'H3K4me1', \n",
    "    'H3K4me2', \n",
    "    'H3K4me3', \n",
    "    'H3K27ac',\n",
    "    'H3K36me3', \n",
    "    'H3K27me3',\n",
    "    'Log2FC FB',\n",
    "    'Log2FC MB',\n",
    "    'Log2FC HB',\n",
    "    'Log2FC SC',\n",
    "    'VAE0',\n",
    "    'VAE1',\n",
    "    'VAE2'\n",
    "]\n",
    "\n",
    "hist_cols = [\n",
    "    'H3K9me3',\n",
    "    'H3K4me1', \n",
    "    'H3K4me2', \n",
    "    'H3K4me3', \n",
    "    'H3K27ac',\n",
    "    'H3K36me3', \n",
    "    'H3K27me3',\n",
    "    'H3K9ac'\n",
    "]\n",
    "\n",
    "div_cols = [\n",
    "    'Log2FC FB',\n",
    "    'Log2FC MB',\n",
    "    'Log2FC HB',\n",
    "    'Log2FC SC',\n",
    "    'Log2FC A11',\n",
    "    'Log2FC A13',\n",
    "    'Log2FC A15',\n",
    "    'Log2FC A18',\n",
    "    'Log2FC P11',\n",
    "    'Log2FC P13',\n",
    "    'Log2FC P15',\n",
    "    'Log2FC P18',\n",
    "]\n",
    "\n",
    "raw_cols = [\n",
    "    'mean_wt',\n",
    "    'mean_ko'\n",
    "]\n",
    "vae_cols = observed_cols\n",
    "df_unobserved = df[unobserved_cols]\n",
    "df_observed = df[observed_cols]\n",
    "observed_dict = {}\n",
    "\n",
    "for c in observed_cols:\n",
    "    observed_dict[c] = df_observed[c].values\n",
    "unobserved_dict = {}\n",
    "for c in unobserved_cols:\n",
    "    unobserved_dict[c] = df_unobserved[c].values\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4) Plot the correlations for the latent nodes with data features\n",
    "\n",
    "Here we want to plot how correlated each node is with the mean values of different epigenetic marks & also \n",
    "the means of input features in the dataset.  \n",
    "\n",
    "We also plot the latent space coloured by each of the features (both observed and unobserved).  \n",
    "\n",
    "Lastly, we plot heatmaps of the input, latent and decoding results for key marker genes so that we can work out whether our model is logical. These same genes are also projected onto the latent space where we can see that the different classes of markers (e.g. Hox genes vs brain genes) are easily discernable.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -----------------------------------------------------------------------------------\n",
    "#                         Plot scatter of the features\n",
    "# -----------------------------------------------------------------------------------\n",
    "all_vae_data = np.copy(vae_data)\n",
    "vae_data = all_vae_data[:, 0:3]\n",
    "vae_vis = VVis(vae_mse, u, None)\n",
    "div_cmap = 'seismic'\n",
    "\n",
    "feature_obs_ax = vae_vis.plot_feature_scatters(df, \"\", columns=hist_cols, vae_data=vae_data, \n",
    "                                               show_plt=False, output_dir=fig_dir, fig_type=\"svg\", \n",
    "                                               save_fig=False, title=f'{experiment_name}',\n",
    "                                               angle_plot=315, cmap=hist_cmap, vmin=0, vmax=10)\n",
    "\n",
    "feature_obs_ax = vae_vis.plot_feature_scatters(df, \"\", columns=div_cols, vae_data=vae_data, \n",
    "                                               show_plt=False, output_dir=fig_dir, fig_type=\"svg\", \n",
    "                                               save_fig=False, title=f'{experiment_name}',\n",
    "                                               angle_plot=315, cmap=div_cmap, vmin=-10, vmax=10)\n",
    "\n",
    "feature_obs_ax = vae_vis.plot_feature_scatters(df, \"\", columns=raw_cols, vae_data=vae_data, \n",
    "                                               show_plt=False, output_dir=fig_dir, fig_type=\"svg\", \n",
    "                                               save_fig=False, title=f'{experiment_name}',\n",
    "                                               angle_plot=315, cmap=rna_cmap, vmin=0, vmax=10)\n",
    "\n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Plot histograms\n",
    "# -----------------------------------------------------------------------------------\n",
    "vae_vis.plot_node_hists(show_plt=True, save_fig=False)\n",
    "vae_vis.plot_input_distribution(df_observed, show_plt=True, save_fig=True, output_dir=fig_dir)\n",
    "vae_vis.plot_input_distribution(df_unobserved, show_plt=True, output_dir=fig_dir, save_fig=True)\n",
    "\n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Plot correlations\n",
    "# -----------------------------------------------------------------------------------\n",
    "vae_vis.plot_node_correlation(show_plt=True, save_fig=False)\n",
    "\n",
    "vae_vis.plot_node_feature_correlation(df_observed, '', columns=observed_cols, vae_data=vae_data, \n",
    "                                  show_plt=True, output_dir=fig_dir, save_fig=True, \n",
    "                                      title=f'Node vs observed features {experiment_name}', cmap=div_cmap)\n",
    "\n",
    "vae_vis.plot_feature_correlation(df_observed, '', columns=observed_cols, show_plt=True, cmap=div_cmap,\n",
    "                                 title=f'Feature vs observed features {experiment_name}',\n",
    "                             output_dir=fig_dir, save_fig=True)\n",
    "\n",
    "vae_vis.plot_node_feature_correlation(df_unobserved, '', columns=unobserved_cols, vae_data=vae_data, \n",
    "                                      cmap=div_cmap,\n",
    "                                      show_plt=True, output_dir=fig_dir, save_fig=True,\n",
    "                                      title=f'Node vs unobserved features {experiment_name}')\n",
    "\n",
    "vae_vis.plot_feature_correlation(df_unobserved, '', columns=unobserved_cols, show_plt=True, cmap=div_cmap,\n",
    "                             output_dir=fig_dir, save_fig=True, title=f'feature un-observed features {experiment_name}')\n",
    "\n",
    "vae_vis.plot_feature_correlation(df, '', columns=unobserved_cols + observed_cols, show_plt=True, cmap=div_cmap,\n",
    "                             output_dir=fig_dir, save_fig=True, title=f'features all {experiment_name}')\n",
    "        \n",
    "vae_vis.plot_feature_correlation(df, '', columns=vae_cols, show_plt=True, cmap=div_cmap,\n",
    "                             output_dir=fig_dir, save_fig=True, title=f'features VAE {experiment_name}')\n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Plot heatmaps of marker genes\n",
    "# -----------------------------------------------------------------------------------\n",
    "fb_genes = ['Foxg1']\n",
    "mb_genes = ['En2']\n",
    "hb_genes = ['Sox3']\n",
    "sc_genes = ['Hoxc9']\n",
    "input_cols = [c for c in df_input.columns if c != gene_name]\n",
    "\n",
    "merged_cols = [c for c in df.columns if \"merged\" in c]\n",
    "# Plot the input space\n",
    "plot_gene_heatmap(df, fb_genes, mb_genes, hb_genes, sc_genes, \n",
    "                  merged_cols, vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', method='input', cmap=hist_cmap)\n",
    "\n",
    "# Plot the decoding space\n",
    "plot_gene_heatmap(df_input, fb_genes, mb_genes, hb_genes, sc_genes, \n",
    "                  input_cols, all_vae_data, vae_mse, \n",
    "                  gene_name, mark='H3K27me3', method='decoding', cmap=div_cmap)\n",
    "\n",
    "# Plot the latent space\n",
    "plot_gene_heatmap(df, fb_genes, mb_genes, hb_genes, sc_genes, \n",
    "                  ['VAE0', 'VAE1', 'VAE2'], all_vae_data, vae_mse, \n",
    "                  gene_name, mark='H3K27me3', method='latent', cmap=hist_cmap)\n",
    "\n",
    "# Also do the main markers\n",
    "fb_genes = ['Emx1', 'Eomes', 'Tbr1', 'Foxg1', 'Lhx6']\n",
    "mb_genes = ['En1', 'En2', 'Lmx1a', 'Bhlhe23', 'Sall4']\n",
    "hb_genes = ['Phox2b', 'Krox20', 'Fev', 'Hoxb1',  'Hoxd3']\n",
    "sc_genes = ['Hoxd8', 'Hoxd9', 'Hoxd10', 'Hoxd11','Hoxa7', 'Hoxa9', 'Hoxa10',\n",
    "              'Hoxb9', 'Hoxb13',  'Hoxc8', 'Hoxc9', 'Hoxc10', 'Hoxc11', 'Hoxc12', 'Hoxc13']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualisations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ko_cols = [c for c in df_input.columns if 'ko' in c]\n",
    "wt_cols = [c for c in df_input.columns if 'wt' in c]\n",
    "# get IDs of the KO cols \n",
    "ko_col_ids = []\n",
    "for i, c in enumerate(df_input.columns):\n",
    "    if 'ko' in c:\n",
    "        ko_col_ids.append(i)\n",
    "\n",
    "bin_cmap = 'bone'\n",
    "# Plot the input space WT and KO\n",
    "merged_wt = [c for c in merged_cols if 'wt' in c]\n",
    "merged_ko = [c for c in merged_cols if 'ko' in c]\n",
    "mb_genes = ['En1', 'En2', 'Lmx1a', 'Otx2', 'Sall4']\n",
    "\n",
    "plot_gene_heatmap(df, fb_genes, mb_genes, hb_genes, sc_genes, \n",
    "                  merged_wt, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=True, cmap=bin_cmap,\n",
    "                  method='input', title=f'All markers input WT {experiment_name}')\n",
    "\n",
    "plot_gene_heatmap(df, fb_genes, mb_genes, hb_genes, sc_genes, \n",
    "                  merged_ko, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=True,  cmap=bin_cmap,\n",
    "                  method='input', title=f'All markers input KO {experiment_name}')\n",
    "\n",
    "# Plot the decoding space for KO and then WT\n",
    "plot_gene_heatmap(df_input, fb_genes, mb_genes, hb_genes, sc_genes, \n",
    "                  wt_cols, all_vae_data, vae_mse, \n",
    "                  gene_name, merge_reps=True, mark='H3K27me3', cmap=bin_cmap,\n",
    "                  method='decoding', title=f'All markers decoding WT {experiment_name}')\n",
    "\n",
    "\n",
    "# Plot the latent space\n",
    "plot_gene_heatmap(df, fb_genes, mb_genes, hb_genes, sc_genes, \n",
    "                  ['VAE0', 'VAE1', 'VAE2'], vae_data, vae_mse,\n",
    "                  gene_name, \n",
    "                   cmap=div_cmap,\n",
    "                  mark='H3K27me3', method='latent', title=f'All markers latent {experiment_name}')\n",
    "\n",
    "\n",
    "progenitors = ['Sox2', 'Sox1', 'Sox3', 'Hes1', 'Hes5']\n",
    "neurons = ['Snap25', 'Syt1', 'Slc32a1','Slc17a6', 'Syn1']\n",
    "glia = ['Cspg4', 'Aqp4', 'Slc6a11', 'Olig1', 'Igfbp3']\n",
    "\n",
    "\n",
    "# Plot the input space\n",
    "plot_gene_heatmap(df, progenitors, neurons, glia, [], \n",
    "                  merged_wt, vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', method='input', cmap=bin_cmap,\n",
    "                  merge_reps=True, title=f'Time markers input WT {experiment_name}')\n",
    "plot_gene_heatmap(df, progenitors, neurons, glia, [], \n",
    "                  merged_ko, vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', method='input', cmap=bin_cmap,\n",
    "                  merge_reps=True, title=f'Time markers input KO {experiment_name}')\n",
    "# Plot the decoding space\n",
    "plot_gene_heatmap(df_input, progenitors, neurons, glia, [], \n",
    "                  wt_cols, all_vae_data, vae_mse, \n",
    "                  gene_name, mark='H3K27me3', method='decoding', cmap=bin_cmap,\n",
    "                  merge_reps=True, title=f'Time markers decoding WT {experiment_name}')\n",
    "# plot_gene_heatmap(df_input, progenitors, neurons, glia, [], \n",
    "#                   ko_cols, vae_data, vae_mse, \n",
    "#                   gene_name, mark='H3K27me3', method='decoding', cmap=bin_cmap,\n",
    "#                   merge_reps=True, title=f'Time markers decoding KO {experiment_name}', \n",
    "#                   input_col_ids=ko_col_ids)\n",
    "\n",
    "# Plot the latent space\n",
    "\n",
    "plot_gene_heatmap(df, progenitors, neurons, glia, [], \n",
    "                  ['VAE0', 'VAE1', 'VAE2'], all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', method='latent', \n",
    "                   cmap=div_cmap,  title=f'Time markers latent {experiment_name}')\n",
    "                  \n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Plot the genes on a scatter plot\n",
    "# -----------------------------------------------------------------------------------\n",
    "gene_markers_ns = [['Slc32a1', 'Gad1', 'Gad2', 'Slc6a1'],\n",
    "                    ['Slc17a6', 'VGLUT2', 'Slc17a7', 'VGLUT1', 'Slc17a8', 'VGLUT3', 'Slc1a1', 'Slc1a2', 'Slc1a6'],\n",
    "                    ['Chat', 'Slc5a7', 'Slc18a3', 'Ache'],\n",
    "                    ['Tubb3', 'Snap25', 'Syt1'],\n",
    "                    ['Gfap', 'Olig2'],\n",
    "                    ['Sox2', 'Hes1', 'Hes5', 'Vim']]\n",
    "\n",
    "marker_labels_ns = ['NS GABAergic', 'NS Glutamatergic', 'NS Cholinergic', 'NS Neurons', 'NS Glia', 'NS Progenitors']\n",
    "\n",
    "vae_vis.plot_values_on_scatters(df_input, gene_name, marker_labels_ns,\n",
    "                            gene_markers_ns, output_dir=fig_dir, plt_bg=False,vae_data=vae_data,\n",
    "                            title=f'Markers NS {experiment_name}', fig_type=\"pdf\",\n",
    "                            show_plt=True, save_fig=True)\n",
    "\n",
    "gene_markers_sep = [['Emx1', 'Eomes', 'Tbr1', 'Foxg1', 'Lhx6'], \n",
    "                    ['En1', 'En2', 'Lmx1a', 'Bhlhe23', 'Sall4'], \n",
    "                    ['Hoxb1', 'Krox20', 'Fev', 'Hoxd3', 'Phox2b'],\n",
    "                    ['Hoxd8', 'Hoxd9', 'Hoxd10', 'Hoxd11', 'Hoxd12', 'Hoxd13', 'Hoxa7', 'Hoxa9', 'Hoxa10', 'Hoxa11', \n",
    "                    'Hoxa13', 'Hoxb9', 'Hoxb13',  'Hoxc8', 'Hoxc9', 'Hoxc10', 'Hoxc11', 'Hoxc12', 'Hoxc13'],\n",
    "                    ['Ccna1', 'Ccna2', 'Ccnd1', 'Ccnd2', 'Ccnd3', 'Ccne1', 'Ccne2', 'Cdc25a', \n",
    "                    'Cdc25b', 'Cdc25c', 'E2f1', 'E2f2', 'E2f3', 'Mcm10', 'Mcm5', 'Mcm3', 'Mcm2', 'Cip2a'],\n",
    "                    ['Cdkn1a', 'Cdkn1b', 'Cdkn1c', 'Cdkn2a', 'Cdkn2b', 'Cdkn2c', 'Cdkn2d'],\n",
    "                    ['Sox2', 'Sox1', 'Sox3', 'Hes1', 'Hes5'],\n",
    "                    ['Snap25', 'Syt1', 'Slc32a1','Slc17a6', 'Syn1'],\n",
    "                    ['Cspg4', 'Aqp4', 'Slc6a11', 'Olig1', 'Igfbp3'],\n",
    "                    ['Foxg1'], \n",
    "                    ['En2'], \n",
    "                    ['Phox2b'],\n",
    "                    ['Hoxc9'],\n",
    "                    ['Sox3']\n",
    "                    ]\n",
    "\n",
    "marker_labels_sep = ['forebrain', 'midbrain', 'hindbrain',  'spinalcord', \n",
    "                     'Proliferation', 'Negative regulators of Cell Cycle', \n",
    "                    'progenitors', 'neurons', 'glia',\n",
    "                     'Foxg1', 'En2', 'Phox2b', 'Hoxc9', 'Sox3'\n",
    "                    ]\n",
    "color_map = {}\n",
    "i = 4\n",
    "for c in marker_labels_sep:\n",
    "    if 'brain' in c or 'spinal' in c:\n",
    "        if 'spinal' in c:\n",
    "            color_map[c] = sc_colour\n",
    "        else:\n",
    "            color_map[c] = get_tissue_colour(c.lower())\n",
    "    else:\n",
    "        color_map[c] = sci_colour[i]\n",
    "    i += 1\n",
    "vae_vis.plot_values_on_scatters(df_input, gene_name, marker_labels_sep,\n",
    "                            gene_markers_sep, output_dir=fig_dir, plt_bg=False,vae_data=vae_data, fig_type=\"svg\",\n",
    "                            title=f'Markers significant {experiment_name}',\n",
    "                            show_plt=False, save_fig=False, color_map=color_map, angle_plot=15)\n",
    "\n",
    "col_id = len(wt_cols)\n",
    "ko_cols_i = [i for i, c in enumerate(df_input.columns) if 'ko' in c and 'merged' not in c]\n",
    "\n",
    "\n",
    "plot_gene_heatmap(df_input, ['Sox3'], ['Foxg1'], ['En2'], ['Hoxc9'], \n",
    "                  df_input.columns, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=False,  cmap=rna_cmap, input_col_ids=ko_cols_i,\n",
    "                  method='decoding', title=f'RNA KO DEC {experiment_name}')\n",
    "col_id = 0\n",
    "wt_cols_i = [i for i, c in enumerate(df_input.columns) if 'wt' in c and 'merged' not in c]\n",
    "\n",
    "plot_gene_heatmap(df_input, ['Sox3'], ['Foxg1'], ['En2'], ['Hoxc9'], \n",
    "                  df_input.columns, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=False,  cmap=rna_cmap, input_col_ids=wt_cols_i,\n",
    "                  method='decoding', title=f'RNA WT DEC {experiment_name}')\n",
    "log2_cols = [c for c in df_input.columns if 'log2Fold' in c]\n",
    "log2_cols_i = [i for i, c in enumerate(df_input.columns) if 'log2Fold' in c]\n",
    "\n",
    "plot_gene_heatmap(df_input, ['Sox3'], ['Foxg1'], ['En2'], ['Hoxc9'], \n",
    "                  df_input.columns, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=False,  cmap=div_cmap, input_col_ids=log2_cols_i,\n",
    "                  method='decoding', title=f'log2FC DEC {experiment_name}')\n",
    "\n",
    "h3k_cols = [c for c in df_input.columns if 'H3K27me3' in c and 'signal' in c and 'median' not in c and 'brain' in c]\n",
    "h3k_cols_i = [i for i, c in enumerate(df_input.columns) if 'H3K27me3' in c and 'signal' in c and 'median' not in c and 'brain' in c]\n",
    "\n",
    "plot_gene_heatmap(df_input, ['Sox3'], ['Foxg1'], ['En2'], ['Hoxc9'], \n",
    "                  df_input.columns, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=False,  cmap=hist_cmap, input_col_ids=h3k_cols_i,\n",
    "                  method='decoding', title=f'H3K WT DEC {experiment_name}')\n",
    "\n",
    "# Ploot all the raw H3K values on teh scatter plot\n",
    "h3k_cols = [c for c in df.columns if 'H3K27me3' in c and 'signal' in c and 'median' not in c and 'brain' in c]\n",
    "tmp_df = df.fillna(0)\n",
    "feature_obs_ax = vae_vis.plot_feature_scatters(tmp_df, \"\", columns=h3k_cols, vae_data=vae_data, \n",
    "                                               show_plt=False, output_dir=fig_dir, fig_type=\"svg\", \n",
    "                                               save_fig=False, title=f'{experiment_name}',\n",
    "                                               angle_plot=315, cmap=hist_cmap, vmin=0, vmax=10)\n",
    "\n",
    "\n",
    "# Plot  genes on heatmap\n",
    "plot_gene_heatmap(df, ['Sox3'], ['Foxg1'], ['En2'], ['Hoxc9'], \n",
    "                  ko_cols, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=True,  cmap=rna_cmap,\n",
    "                  method='input', title=f'RNA KO {experiment_name}')\n",
    "plot_gene_heatmap(df, ['Sox3'], ['Foxg1'], ['En2'], ['Hoxc9'], \n",
    "                  wt_cols, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=True,  cmap=rna_cmap,\n",
    "                  method='input', title=f'RNA WT {experiment_name}')\n",
    "h3k_cols = [c for c in df.columns if 'H3K27me3' in c and 'signal' in c and 'median' not in c and 'brain' in c]\n",
    "print(h3k_cols)\n",
    "plot_gene_heatmap(df, ['Sox3'], ['Foxg1'], ['En2'], ['Hoxc9'], \n",
    "                  h3k_cols, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=True,  cmap=hist_cmap,\n",
    "                  method='input', title=f'H3K WT {experiment_name}')\n",
    "log2_cols = [c for c in df.columns if 'log2Fold' in c]\n",
    "print(log2_cols)\n",
    "plot_gene_heatmap(df, ['Sox3'], ['Foxg1'], ['En2'], ['Hoxc9'], \n",
    "                  log2_cols, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=True,  cmap=div_cmap,\n",
    "                  method='input', title=f'log2FC {experiment_name}')\n",
    "\n",
    "ko_cols = [c for c in df_all if 'ko' in c and 'merged' in c]\n",
    "plot_gene_heatmap(df_all, ['Sox1'], ['Sox2'], ['Sox3'], [], \n",
    "                  ko_cols, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=True,  cmap=rna_cmap,\n",
    "                  method='input', title=f'Sox RNA KO {experiment_name}')\n",
    "wt_cols = [c for c in df_all if 'wt' in c and 'merged' in c]\n",
    "\n",
    "plot_gene_heatmap(df_all, ['Sox1'], ['Sox2'], ['Sox3'], [],\n",
    "                  wt_cols, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=True,  cmap=rna_cmap,\n",
    "                  method='input', title=f'Sox RNA WT {experiment_name}')\n",
    "\n",
    "plot_gene_heatmap(df, ['Sox3'], ['Foxg1'], ['En2'], ['Hoxc9'], \n",
    "                  ['VAE0', 'VAE1', 'VAE2'], all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=False,  cmap=bin_cmap,\n",
    "                  method='input', title=f'Latent {experiment_name}')\n",
    "\n",
    "# Plot feature correlation\n",
    "vae_vis.plot_feature_correlation(df_observed, '', columns=observed_cols, show_plt=True, cmap=\"RdBu_r\",\n",
    "                                 title=f'Feature vs observed features {experiment_name}',\n",
    "                             output_dir=fig_dir, save_fig=True)\n",
    "\n",
    "# Plot feature correlations for heatmap \n",
    "vae_vis.plot_feature_correlation(df, '', columns=vae_cols, show_plt=True, cmap='RdBu_r',\n",
    "                             output_dir=fig_dir, save_fig=True, title=f'features VAE {experiment_name}')\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Plot the heatmaps of the Hox and other tisue specific genes\n",
    "\"\"\"\n",
    "plot_gene_heatmap(df_all, ['Pax2'], ['Hoxb9'], ['Hoxc9'], ['Hoxd9'],\n",
    "                  wt_cols, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=False,  cmap=rna_cmap,\n",
    "                  method='input', title=f'Sox pax2 RNA WT {experiment_name}')\n",
    "\n",
    "plot_gene_heatmap(df_all, ['Pax2'], ['Hoxb9'], ['Hoxc9'], ['Hoxd9'],\n",
    "                  ko_cols, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=False,  cmap=rna_cmap,\n",
    "                  method='input', title=f'pax2 RNA KO {experiment_name}')\n",
    "\n",
    "mb_genes = ['En1', 'En2', 'Lmx1a', 'Bhlhe23', 'Sall4']\n",
    "hb_genes = ['Phox2b', 'Krox20', 'Fev', 'Hoxb1',  'Hoxd3']\n",
    "mb_genes = ['En1', 'En2', 'Lmx1a', 'Bhlhe23', 'Sall4', 'Pax7', 'Lhx9', 'Evx1', 'Sox14', \n",
    "           'Gata3', 'Dmbx1', 'Tal1', 'Pou4f3', 'Pouf4f2', 'Gata3', 'Barhl1', 'Pax5', 'Dmbx1',\n",
    "           'Pou4f1', 'Irx4', 'Pax7', 'Pax1', 'Pax8', 'Ebf2', 'Patx3', 'Onecut1', 'Otx2', \n",
    "           'Lhx5', 'Irx3', 'Irx5', 'Ebf3', 'Foxb1', 'Irx2', 'Shox2', 'Foxa2', 'Irx1', 'Tcfap2b']\n",
    "\n",
    "plot_gene_heatmap(df_all, ['Otx2'], ['En2'], ['Phox2b'], ['Hoxd3'],\n",
    "                  wt_cols, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=False,  cmap=rna_cmap,\n",
    "                  method='input', title=f'Mid RNA WT {experiment_name}')\n",
    "\n",
    "\n",
    "plot_gene_heatmap(df_all, ['Otx2'], ['En2'], ['Phox2b'], ['Hoxd3'],\n",
    "                  ko_cols, all_vae_data, vae_mse,\n",
    "                  gene_name, mark='H3K27me3', merge_reps=False,  cmap=rna_cmap,\n",
    "                  method='input', title=f'Mid RNA KO {experiment_name}')\n",
    "\n",
    "\n",
    "# Plot gene names on scatter\n",
    "gene_markers_sep = [['Cdkn1a'], ['Cdkn2a'], ['Cdkn2b'], ['Cip2a'], ['Cdc25c'], ['Mcm10'],\n",
    "                    ['E2f1'], ['Ccna2'], ['Ccnd1'], ['Aqp4'], ['Snap25'], ['Hes5']\n",
    "                    ]\n",
    "\n",
    "marker_labels_sep = ['Cdkn1a', 'Cdkn2a', 'Cdkn2b',  'Cip2a', \n",
    "                     'Cdc25c', 'Mcm10', \n",
    "                    'E2f1', 'Ccna2', 'Ccnd1',\n",
    "                     'Aqp4', 'Snap25', 'Hes5'\n",
    "                    ]\n",
    "color_map = {}\n",
    "i = 4\n",
    "for c in marker_labels_sep:\n",
    "    if 'brain' in c or 'spinal' in c:\n",
    "        if 'spinal' in c:\n",
    "            color_map[c] = sc_colour\n",
    "        else:\n",
    "            color_map[c] = get_tissue_colour(c.lower())\n",
    "    else:\n",
    "        color_map[c] = sci_colour[i]\n",
    "    i += 1\n",
    "vae_vis.plot_values_on_scatters(df_input, gene_name, marker_labels_sep,\n",
    "                            gene_markers_sep, output_dir=fig_dir, plt_bg=False,vae_data=vae_data, fig_type=\"svg\",\n",
    "                            title=f'Markers specific {experiment_name}',\n",
    "                            show_plt=False, save_fig=False, color_map=color_map, angle_plot=315)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Print out the gene lists. \"\"\"\n",
    "\n",
    "gene_markers_sep = [['Emx1', 'Eomes', 'Tbr1', 'Foxg1', 'Lhx6'], \n",
    "                    ['En1', 'En2', 'Lmx1a', 'Bhlhe23', 'Sall4'], \n",
    "                    ['Hoxb1', 'Krox20', 'Fev', 'Hoxd3', 'Phox2b'],\n",
    "                    ['Hoxd8', 'Hoxd9', 'Hoxd10', 'Hoxd11', 'Hoxd12', 'Hoxd13', 'Hoxa7', 'Hoxa9', 'Hoxa10', 'Hoxa11', \n",
    "                    'Hoxa13', 'Hoxb9', 'Hoxb13',  'Hoxc8', 'Hoxc9', 'Hoxc10', 'Hoxc11', 'Hoxc12', 'Hoxc13'],\n",
    "                    ['Ccna1', 'Ccna2', 'Ccnd1', 'Ccnd2', 'Ccnd3', 'Ccne1', 'Ccne2', 'Cdc25a', \n",
    "                    'Cdc25b', 'Cdc25c', 'E2f1', 'E2f2', 'E2f3', 'Mcm10', 'Mcm5', 'Mcm3', 'Mcm2', 'Cip2a'],\n",
    "                    ['Cdkn1a', 'Cdkn1b', 'Cdkn1c', 'Cdkn2a', 'Cdkn2b', 'Cdkn2c', 'Cdkn2d'],\n",
    "                    ['Sox2', 'Sox1', 'Sox3', 'Hes1', 'Hes5'],\n",
    "                    ['Snap25', 'Syt1', 'Slc32a1','Slc17a6', 'Syn1'],\n",
    "                    ['Cspg4', 'Aqp4', 'Slc6a11', 'Olig1', 'Igfbp3'],\n",
    "                    ['Foxg1'], \n",
    "                    ['En2'], \n",
    "                    ['Phox2b'],\n",
    "                    ['Hoxc9'],\n",
    "                    ['Sox3']\n",
    "                    ]\n",
    "\n",
    "marker_labels_sep = ['forebrain', 'midbrain', 'hindbrain',  'spinalcord', \n",
    "                     'Proliferation', 'Negative regulators of Cell Cycle', \n",
    "                    'progenitors', 'neurons', 'glia',\n",
    "                     'Foxg1', 'En2', 'Phox2b', 'Hoxc9', 'Sox3'\n",
    "                    ]\n",
    "# Print genes that were actually in our dataset \n",
    "cx = [g for g in df[gene_name].values if g in ['Hoxd8', 'Hoxd9', 'Hoxd10', 'Hoxd11', 'Hoxd12', 'Hoxd13', 'Hoxa7', 'Hoxa9', 'Hoxa10', 'Hoxa11', \n",
    "                    'Hoxa13', 'Hoxb9', 'Hoxb13',  'Hoxc8', 'Hoxc9', 'Hoxc10', 'Hoxc11', 'Hoxc12', 'Hoxc13']]\n",
    "print(', '.join(cx))\n",
    "\n",
    "len(cx)\n",
    "cx = [g for g in df[gene_name].values if g in ['Sox2', 'Sox1', 'Sox3', 'Hes1', 'Hes5']]\n",
    "print(', '.join(cx))\n",
    "cx = [g for g in df[gene_name].values if g in ['Snap25', 'Syt1', 'Slc32a1','Slc17a6', 'Syn1']]\n",
    "print(', '.join(cx))\n",
    "cx = [g for g in df[gene_name].values if g in['Cspg4', 'Aqp4', 'Slc6a11', 'Olig1', 'Igfbp3']]\n",
    "print(', '.join(cx))\n",
    "prolif = ['Ccna1', 'Ccna2', 'Ccnd1', 'Ccnd2', 'Ccnd3', 'Ccne1', 'Ccne2', 'Cdc25a', 'Cdc25b', 'Cdc25c', 'E2f1', 'E2f2', 'E2f3', 'Mcm10', 'Mcm5', 'Mcm3', 'Mcm2', 'Cip2a']\n",
    "anti_prolif = ['Cdkn1a', 'Cdkn1b', 'Cdkn1c', 'Cdkn2a', 'Cdkn2b', 'Cdkn2c', 'Cdkn2d']\n",
    "pl = [g for g in df[gene_name].values if g in prolif]\n",
    "apl = [g for g in df[gene_name].values if g in anti_prolif]\n",
    "print(', '.join(pl))\n",
    "print(', '.join(apl))\n",
    "\n",
    "\n",
    "# Reset the vae data \n",
    "vae_data = all_vae_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5) Quantify separability of latent space\n",
    "\n",
    "In the next section we compare the distances in the latent space to those using other summaries of the data. \n",
    "\n",
    "The distances are computed as the euclidean distance to the centroid. Where the centroid is either the center of the cluster of genes of interest OR the center of the two groups. We expect the referece to have a greater spread and that our genes should be closer to the centroid of the gene set they belong to.\n",
    "\n",
    "Datasets tested:\n",
    "1) VAE deep (3 layer)\n",
    "2) VAE shallow (1 layer)\n",
    "3) PCA (3 PCs)\n",
    "4) RNAseq data (normalised TMM 64 features)\n",
    "5) VAE input data (normalised TMM and normalised H3K27me3 and LogFC, 100 features)\n",
    "6) logFC forebrain and H3K27me3 signal.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6) Save ranks for GSEA\n",
    "\n",
    "We perform GSEA on the ranks of the VAE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Save genes sorted by latent node (GSEA)\n",
    "# -----------------------------------------------------------------------------------\n",
    "for n in range(0, num_nodes):\n",
    "    with open(f'{ora_dir}vae{n}_genes_{experiment_name}_{date}.csv', 'w+') as f:\n",
    "        f.write(gene_id + ',value\\n')\n",
    "        desc_sorted = (-vae_data[:,n]).argsort()  # Sort the genes by descending order\n",
    "        genes = df[gene_id].values[desc_sorted]\n",
    "        vae_data_sorted = vae_data[:,n][desc_sorted]\n",
    "        i = 0\n",
    "        for g in genes:\n",
    "            f.write(f'{g},{vae_data_sorted[i]}\\n')\n",
    "            i += 1\n",
    "            \n",
    "\n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Save genes sorted by latent node (GSEA)\n",
    "# -----------------------------------------------------------------------------------\n",
    "for n in range(0, num_nodes):\n",
    "    with open(f'{ora_dir}vae{n}_gene-names_{experiment_name}_{date}.csv', 'w+') as f:\n",
    "        f.write(gene_name + ',value\\n')\n",
    "        desc_sorted = (-vae_data[:,n]).argsort()  # Sort the genes by descending order\n",
    "        genes = df[gene_name].values[desc_sorted]\n",
    "        vae_data_sorted = vae_data[:,n][desc_sorted]\n",
    "        i = 0\n",
    "        for g in genes:\n",
    "            f.write(f'{g},{vae_data_sorted[i]}\\n')\n",
    "            i += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7) Compute clusters driving VAE \n",
    "\n",
    "In the next section, we use information about the VAE nodes to compute clusters regarding the latent space. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_q1_q3(node_idx):\n",
    "    iqr_val = 0.95 * stats.iqr(vae_data[:, node_idx])\n",
    "    q1_genes_idxs = np.where(vae_data[:, node_idx] <= -1.25)[0] # (np.mean(vae_data[:, node_idx]) - iqr_val))[0]\n",
    "    q3_genes_idxs = np.where(vae_data[:, node_idx] >= 1.25)[0] #(np.mean(vae_data[:, node_idx]) + iqr_val))[0]\n",
    "    return q1_genes_idxs, q3_genes_idxs\n",
    "\n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Find the genes driving the latent space\n",
    "# -----------------------------------------------------------------------------------\n",
    "\n",
    "n_clusters = 6\n",
    "# Get the indexs for each of the q1 and q3 for each node\n",
    "q1_n0_idxs, q3_n0_idxs = get_q1_q3(0)\n",
    "q1_n1_idxs, q3_n1_idxs = get_q1_q3(1)\n",
    "q1_n2_idxs, q3_n2_idxs = get_q1_q3(2)\n",
    "\n",
    "# Create a list of the indexs\n",
    "vae_set_idxs = [\n",
    "    q1_n0_idxs, q3_n0_idxs, \n",
    "    q1_n1_idxs, q3_n1_idxs, \n",
    "    q1_n2_idxs, q3_n2_idxs\n",
    "]\n",
    "\n",
    "vae_set_labels = ['Set 1', 'Set 2', 'Set 3', 'Set 4', 'Set 5', 'Set 6']\n",
    "\n",
    "# Get all the driver and non-driver idxs\n",
    "non_driver_idxs = []\n",
    "driver_idxs = []\n",
    "for i, ids in enumerate(vae_set_idxs):\n",
    "    vae_mse.u.dp([vae_set_labels[i], len(ids)])\n",
    "    driver_idxs += list(ids)\n",
    "    \n",
    "for i in range(0, len(df)):\n",
    "    if i not in driver_idxs:\n",
    "        non_driver_idxs.append(i)\n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Plot a VENN of how the genes in teh clusters overlap\n",
    "# -----------------------------------------------------------------------------------\n",
    "\n",
    "gene_sets = [[], [], [], [], [], []]\n",
    "\n",
    "for i, g in enumerate(df[gene_name].values):\n",
    "    ci = 0\n",
    "    for gene_set in vae_set_idxs:\n",
    "        for j in gene_set:\n",
    "            if i == j:\n",
    "                gene_sets[ci].append(g)\n",
    "        ci += 1\n",
    "\n",
    "set_sets = []\n",
    "for g in gene_sets:  \n",
    "    if len(g) > 1:\n",
    "        set_sets.append(set(g))\n",
    "\n",
    "set_labels = venn.get_labels(set_sets)\n",
    "vae_set_labels_l = [s.replace('Set', 'Group') for s in vae_set_labels]\n",
    "vae_set_genes_len_dict = {}\n",
    "vae_set_genes_dict = {}\n",
    "for i, l in enumerate(vae_set_labels):\n",
    "    vae_set_genes_dict[l] = gene_sets[i]\n",
    "    vae_set_genes_len_dict[l] = len(gene_sets[i])\n",
    "\n",
    "venn.venn6(set_labels, names=vae_set_labels_l)\n",
    "save_fig('gene_sets.svg')\n",
    "plt.show()\n",
    "\n",
    "\n",
    "vae_set_genes_as_list = []\n",
    "for set_name in vae_set_labels:\n",
    "    vae_set_genes_as_list.append(vae_set_genes_dict.get(set_name))\n",
    "\n",
    "sorted_sets_by_size = {k: v for k, v in sorted(vae_set_genes_len_dict.items(), key=lambda item: item[1])}\n",
    "gene_set_labels_unique = []\n",
    "for g in df[gene_name].values:\n",
    "    found = False\n",
    "    for label, values in sorted_sets_by_size.items():\n",
    "        gene_group = vae_set_genes_dict[label]\n",
    "        for g_l in gene_group:\n",
    "            if g_l == g:\n",
    "                gene_set_labels_unique.append(label)\n",
    "                found = True\n",
    "                break\n",
    "        if found:\n",
    "            break\n",
    "    if not found:\n",
    "        gene_set_labels_unique.append(None)\n",
    "df['GeneDriverSet'] =  gene_set_labels_unique\n",
    "\n",
    "# We also want to add each of the gene sets\n",
    "gene_labels = [[], [], [], [], [], []]\n",
    "for g in df[gene_name].values:\n",
    "    for i, label in enumerate(vae_set_labels):\n",
    "        found = False\n",
    "        gene_group = vae_set_genes_dict[label]\n",
    "        for g_l in gene_group:\n",
    "            if g == g_l:\n",
    "                gene_labels[i].append(True)\n",
    "                found = True\n",
    "                break\n",
    "        if not found:\n",
    "            gene_labels[i].append(False)\n",
    "for i, label in enumerate(vae_set_labels):\n",
    "    df[label] = gene_labels[i]\n",
    "    \n",
    "            \n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Plot the genes on a scatter plot\n",
    "# -----------------------------------------------------------------------------------\n",
    "vae_vis.plot_values_on_scatters(df, gene_name, vae_set_labels,\n",
    "                            vae_set_genes_as_list, vae_data=vae_data, output_dir=fig_dir, plt_bg=True,\n",
    "                            title=f'VAE gene driver sets {experiment_name}',\n",
    "                            show_plt=True, save_fig=True)\n",
    "\n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Plot histograms of the VAE nodes with cutoffs used above\n",
    "# -----------------------------------------------------------------------------------\n",
    "for n in range(0, num_nodes):\n",
    "    x = vae_data[:,n]\n",
    "    iqr_val = 1.0 * stats.iqr(x)\n",
    "    plt.hist(vae_data[:,n], bins=20, color='grey')\n",
    "    plt.ylim(0, 380)\n",
    "    plt.axvline(x.mean() - iqr_val, color='k', linestyle='dashed', linewidth=1)\n",
    "    plt.axvline(x.mean() + iqr_val, color='r', linestyle='dashed', linewidth=1)\n",
    "    save_fig(f'hist-{n}')\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hox_genes = sc_genes\n",
    "forebrain_genes = fb_genes\n",
    "h3cols = []\n",
    "fb_rna_cols = []\n",
    "sc_rna_cols = []\n",
    "\n",
    "# Do it in this way so they are correctly ordered\n",
    "for c in df.columns:\n",
    "    if 'H3K27me' in c and 'signal' in c:\n",
    "        h3cols.append(c)\n",
    "    elif 'wt' in c:\n",
    "        if 'fb' in c:\n",
    "            fb_rna_cols.append(c)\n",
    "        if 'sc' in c:\n",
    "            sc_rna_cols.append(c)\n",
    "for c in df.columns:\n",
    "    if 'ko' in c:\n",
    "        if 'fb' in c:\n",
    "            fb_rna_cols.append(c)\n",
    "        if 'sc' in c:\n",
    "            sc_rna_cols.append(c)\n",
    "\n",
    "def plot_gene_boxplot(df, title, cluster_id, cols, idxs):\n",
    "    idxs = idxs[cluster_id]\n",
    "    boxplot = Boxplot(df, gene_name, cols[0])\n",
    "    box_df = boxplot.format_data_for_boxplot(df, cols, gene_name, df[gene_name].values[idxs])\n",
    "    is_hox = []\n",
    "    sns.set_style(\"white\")\n",
    "    for c in box_df['Samples'].values:\n",
    "        if 'Hox' in c:\n",
    "            is_hox.append('Hox')\n",
    "        else:\n",
    "            is_hox.append('Hox like')\n",
    "    box_df['is_hox'] = is_hox\n",
    "\n",
    "    boxplot = Boxplot(box_df, \"Conditions\", \"Values\", add_stats=False, add_dots=False, figsize=(1,0.5),\n",
    "                       order=cols)\n",
    "    boxplot.palette = sci_colour\n",
    "\n",
    "    ax = boxplot.plot()\n",
    "    plt.title(title)\n",
    "    ax.set_ylim(0, 12)\n",
    "    sns.set_style(\"white\")\n",
    "    c = 0\n",
    "    for b in ax.artists:\n",
    "        if c < len(cols)/2:\n",
    "            b.set_facecolor(wt_colour)\n",
    "        else:\n",
    "            b.set_facecolor(ko_colour)\n",
    "        c += 1\n",
    "    save_fig(f'{title}')\n",
    "    \n",
    "    plt.show()\n",
    "\n",
    "avgs_fb = [['wt11fb1',\n",
    "         'wt11fb2'],\n",
    "         ['wt13fb1',\n",
    "         'wt13fb2'],\n",
    "         ['wt15fb1',\n",
    "         'wt15fb2'],\n",
    "         ['wt18fb1',\n",
    "         'wt18fb2'],\n",
    "         ['ko11fb1',\n",
    "         'ko11fb2'],\n",
    "         ['ko13fb1',\n",
    "         'ko13fb2'],\n",
    "         ['ko15fb1',\n",
    "         'ko15fb2'],\n",
    "         ['ko18fb1',\n",
    "         'ko18fb2']]\n",
    "\n",
    "avgs_sc = [['wt11sc1',\n",
    "         'wt11sc2'],\n",
    "         ['wt13sc1',\n",
    "         'wt13sc2'],\n",
    "         ['wt15sc1',\n",
    "         'wt15sc2'],\n",
    "         ['wt18sc1',\n",
    "         'wt18sc2'],\n",
    "         ['ko11sc1',\n",
    "         'ko11sc2'],\n",
    "         ['ko13sc1',\n",
    "         'ko13sc2'],\n",
    "         ['ko15sc1',\n",
    "         'ko15sc2'],\n",
    "         ['ko18sc1',\n",
    "         'ko18sc2']]\n",
    "\n",
    "avgs_hb = [['wt11hb1',\n",
    "         'wt11hb2'],\n",
    "         ['wt13hb1',\n",
    "         'wt13hb2'],\n",
    "         ['wt15hb1',\n",
    "         'wt15hb2'],\n",
    "         ['wt18hb1',\n",
    "         'wt18hb2'],\n",
    "         ['ko11hb1',\n",
    "         'ko11hb2'],\n",
    "         ['ko13hb1',\n",
    "         'ko13hb2'],\n",
    "         ['ko15hb1',\n",
    "         'ko15hb2'],\n",
    "         ['ko18hb1',\n",
    "         'ko18hb2']]\n",
    "\n",
    "avgs_mb = [['wt11mb1',\n",
    "         'wt11mb2'],\n",
    "         ['wt13mb1',\n",
    "         'wt13mb2'],\n",
    "         ['wt15mb1',\n",
    "         'wt15mb2'],\n",
    "         ['wt18mb1',\n",
    "         'wt18mb2'],\n",
    "         ['ko11mb1',\n",
    "         'ko11mb2'],\n",
    "         ['ko13mb1',\n",
    "         'ko13mb2'],\n",
    "         ['ko15mb1',\n",
    "         'ko15mb2'],\n",
    "         ['ko18mb1',\n",
    "         'ko18mb2']]\n",
    "avgs_df = pd.DataFrame()\n",
    "avgs_df[gene_name] = df[gene_name].values\n",
    "fb_cols = []\n",
    "for f in avgs_fb:\n",
    "    new_col = f'{f[0][:2]} {f[0][2:4]} FB'\n",
    "    avgs_df[new_col] = (df[f[0]].values + df[f[1]].values) / 2.0\n",
    "    fb_cols.append(new_col)\n",
    "    \n",
    "sc_cols = []\n",
    "for f in avgs_sc:\n",
    "    new_col = f'{f[0][:2]} {f[0][2:4]} SC'\n",
    "    avgs_df[new_col] = (df[f[0]].values + df[f[1]].values) / 2.0\n",
    "    sc_cols.append(new_col)\n",
    "    \n",
    "mb_cols = []\n",
    "for f in avgs_mb:\n",
    "    new_col = f'{f[0][:2]} {f[0][2:4]} MB'\n",
    "    avgs_df[new_col] = (df[f[0]].values + df[f[1]].values) / 2.0\n",
    "    mb_cols.append(new_col)\n",
    "    \n",
    "hb_cols = []\n",
    "for f in avgs_hb:\n",
    "    new_col = f'{f[0][:2]} {f[0][2:4]} HB'\n",
    "    avgs_df[new_col] = (df[f[0]].values + df[f[1]].values) / 2.0\n",
    "    hb_cols.append(new_col)\n",
    "    \n",
    "for i, gs in enumerate(vae_set_idxs):\n",
    "    plot_gene_boxplot(avgs_df, f'Cluster {i + 1} FB', i, fb_cols, vae_set_idxs)\n",
    "    plot_gene_boxplot(avgs_df, f'Cluster {i + 1} MB', i, mb_cols, vae_set_idxs)\n",
    "    plot_gene_boxplot(avgs_df, f'Cluster {i + 1} HB', i, hb_cols, vae_set_idxs)\n",
    "    plot_gene_boxplot(avgs_df, f'Cluster {i + 1} SC', i, sc_cols, vae_set_idxs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot the heatmaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_vae_rna_cols(vis_df, cols, num_nodes, fb_logfc, mb_logfc, hb_logfc, sc_logfc, \n",
    "                      h3k27me3_fb_e11, h3k27me3_fb_e18, h3k27me3_hb_e11, h3k27me3_hb_e18, title='Heatmap', \n",
    "                      num_values=10):\n",
    "    vae_cols = [f'VAE{i}' for i in range(0, num_nodes)]\n",
    "\n",
    "    sns.set_style(\"ticks\")\n",
    "    for r_i, r_c in enumerate(cols):\n",
    "        for direction in ['up', 'down']:\n",
    "            # First add the bottom ones\n",
    "            fig, ax = plt.subplots()\n",
    "            sns.set(rc={'figure.figsize': (3.5, 2), 'font.family': 'sans-serif',\n",
    "            'font.sans-serif': 'Arial', 'font.size': 6}, style=\"ticks\")\n",
    "            rcm_df_sorted = vis_df.sort_values(by=[r_c])\n",
    "            if direction == 'up':\n",
    "                rcm_df_tails = rcm_df_sorted.nlargest(num_values, r_c)\n",
    "            else:\n",
    "                rcm_df_tails = rcm_df_sorted.nsmallest(num_values, r_c)\n",
    "            # Add the smallest and the largest\n",
    "            # first heatmap\n",
    "            N = len(rcm_df_tails)\n",
    "            ax.set_xlim(-0.5, 7.5)\n",
    "            ax.set_xticks([0, 1, 2, 3, 4, 5, 6, 7])\n",
    "            ax.set_xticklabels(['FB', 'MB', 'HB', 'SC', \n",
    "                                'FB e11.5', 'FB e18.5', 'HB e11.5', 'HB e18'])\n",
    "            ax.set_yticks(range(N))\n",
    "            ax.set_yticklabels(list(rcm_df_tails[gene_name].values))\n",
    "            ax.set_ylabel('Genes')\n",
    "\n",
    "            max_hk3 = 35\n",
    "            max_rna = 10\n",
    "            min_rna = -10\n",
    "            cmap_rna = 'seismic'\n",
    "            cmap_hk3 = 'Greens'\n",
    "            im1 = ax.imshow(np.vstack([rcm_df_tails[fb_logfc], rcm_df_tails[fb_logfc]]).T, \n",
    "                            aspect='auto',\n",
    "                            extent=[-0.5, 0.5, -0.5, N - 0.5], vmin=min_rna, vmax=max_rna,\n",
    "                            origin='lower', cmap=cmap_rna)\n",
    "            im1 = ax.imshow(np.vstack([rcm_df_tails[mb_logfc], rcm_df_tails[mb_logfc]]).T, \n",
    "                            aspect='auto',\n",
    "                            extent=[0.5, 1.5, -0.5, N - 0.5], vmin=min_rna, vmax=max_rna,\n",
    "                            origin='lower', cmap=cmap_rna)\n",
    "            im1 = ax.imshow(np.vstack([rcm_df_tails[hb_logfc], rcm_df_tails[hb_logfc]]).T, \n",
    "                aspect='auto',\n",
    "                extent=[1.5, 2.5, -0.5, N - 0.5], vmin=min_rna, vmax=max_rna,\n",
    "                origin='lower', cmap=cmap_rna)\n",
    "            # Now add in the H3K27me3 ones\n",
    "            im1 = ax.imshow(np.vstack([rcm_df_tails[sc_logfc], rcm_df_tails[sc_logfc]]).T, \n",
    "                aspect='auto',\n",
    "                extent=[2.5, 3.5, -0.5, N - 0.5], vmin=min_rna, vmax=max_rna,\n",
    "                origin='lower', cmap=cmap_rna)\n",
    "\n",
    "            im3 = ax.imshow(np.vstack([rcm_df_tails[h3k27me3_fb_e11], rcm_df_tails[h3k27me3_fb_e11]]).T, \n",
    "                            aspect='auto',\n",
    "                            extent=[3.5, 4.5, -0.5, N - 0.5], vmin=0, vmax=max_hk3,\n",
    "                            origin='lower', cmap=cmap_hk3)\n",
    "\n",
    "            im3 = ax.imshow(np.vstack([rcm_df_tails[h3k27me3_fb_e18], rcm_df_tails[h3k27me3_fb_e18]]).T, \n",
    "                aspect='auto',\n",
    "                extent=[4.5, 5.5, -0.5, N - 0.5], vmin=0, vmax=max_hk3,\n",
    "                origin='lower', cmap=cmap_hk3)\n",
    "\n",
    "            im3 = ax.imshow(np.vstack([rcm_df_tails[h3k27me3_hb_e11], rcm_df_tails[h3k27me3_hb_e11]]).T, \n",
    "                            aspect='auto',\n",
    "                            extent=[5.5, 6.5, -0.5, N - 0.5], vmin=0, vmax=max_hk3,\n",
    "                            origin='lower', cmap=cmap_hk3)\n",
    "\n",
    "            im3 = ax.imshow(np.vstack([rcm_df_tails[h3k27me3_hb_e18], rcm_df_tails[h3k27me3_hb_e18]]).T, \n",
    "                aspect='auto',\n",
    "                extent=[6.5, 7.5, -0.5, N - 0.5], vmin=0, vmax=max_hk3,\n",
    "                origin='lower', cmap=cmap_hk3)\n",
    "\n",
    "            cbar1 = fig.colorbar(im1, ax=ax, label='RNA logFC')\n",
    "            cbar2 = fig.colorbar(im3, ax=ax, label='H3K27me3 signal')\n",
    "\n",
    "            fig.tight_layout()\n",
    "            plt.xticks(rotation=45, horizontalalignment='right')\n",
    "            plt.title(f'{title} {r_c}')\n",
    "            save_fig(f'{title.replace(\" \", \"_\")}_{direction}_{r_c}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "h3k27me3_fb_e11 = [c for c in df.columns if '10.5' in c and 'forebrain' in c and 'H3K27me3' in c and 'signal' in c and 'median' not in c][0]\n",
    "h3k27me3_fb_e18 = [c for c in df.columns if '16.5' in c and 'forebrain' in c and 'H3K27me3' in c and 'signal' in c and 'median' not in c][0]\n",
    "\n",
    "h3k27me3_hb_e11 = [c for c in df.columns if '10.5' in c and 'hindbrain' in c and 'H3K27me3' in c and 'signal' in c and 'median' not in c][0]\n",
    "h3k27me3_hb_e18 = [c for c in df.columns if '16.5' in c and 'hindbrain' in c and 'H3K27me3' in c and 'signal' in c and 'median' not in c][0]\n",
    "\n",
    "plot_vae_rna_cols(df, [\"VAE0\", \"VAE1\", \"VAE2\"], 3,     \n",
    "                    'Log2FC FB',\n",
    "                    'Log2FC MB',\n",
    "                    'Log2FC HB',\n",
    "                    'Log2FC SC', \n",
    "                     h3k27me3_fb_e11, h3k27me3_fb_e18, h3k27me3_hb_e11, h3k27me3_hb_e18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -----------------------------------------------------------------------------------\n",
    "#                         Save background genes (ORA)\n",
    "# -----------------------------------------------------------------------------------\n",
    "n_clusters = num_nodes\n",
    "with open(f'{ora_dir}background_genes_{experiment_name}_{date}.csv', 'w+') as f:\n",
    "    genes = df_all[gene_id].values\n",
    "    f.write(gene_id + '\\n')\n",
    "    for g in genes:\n",
    "        f.write(f'{g}\\n')\n",
    "        \n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Save training genes (sig genes) (ORA)\n",
    "# -----------------------------------------------------------------------------------\n",
    "with open(f'{ora_dir}training-df_genes_{experiment_name}_{date}.csv', 'w+') as f:\n",
    "    genes = df_training[gene_id].values\n",
    "    f.write(gene_id + '\\n')\n",
    "    for g in genes:\n",
    "        f.write(f'{g}\\n')\n",
    "        \n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Save genes in each of the clusters we annotated (ORA)\n",
    "# -----------------------------------------------------------------------------------\n",
    "for i in range(0, n_clusters):\n",
    "    with open(f'{ora_dir}cluster_{i}_genes_{experiment_name}_{date}.csv', 'w+') as f:\n",
    "        genes = df[gene_id].values[vae_set_idxs[i]]\n",
    "        f.write(gene_id + '\\n')\n",
    "        for g in genes:\n",
    "            f.write(f'{g}\\n')\n",
    "    \n",
    "# -----------------------------------------------------------------------------------\n",
    "#                         Save genes sorted by latent node (GSEA) and only the top\n",
    "# -----------------------------------------------------------------------------------\n",
    "def plot_top_vae_genes(vae_data_sorted, gene_ids_sorted, gene_names_sorted, label=\"\", num=20):\n",
    "        i = 0\n",
    "        gs = []\n",
    "        for g in gene_names_sorted:\n",
    "            if g != \"0\":\n",
    "                f.write(f'{g},{vae_data_sorted[i]}\\n')\n",
    "                gs.append(g)\n",
    "            else:\n",
    "                f.write(f'{gene_ids_sorted[i]},{vae_data_sorted[i]}\\n')\n",
    "                gs.append(gene_ids_sorted[i])\n",
    "            i += 1\n",
    "            if i >= num:\n",
    "                break\n",
    "        heatmap_df = pd.DataFrame()\n",
    "        heatmap_df[gene_name] = gs\n",
    "        heatmap_df['values'] = vae_data_sorted[:num]\n",
    "        heatmap = Heatmap(heatmap_df, ['values'], gene_name, vmin=-4, vmax=4,\n",
    "                  title=f'top {num} genes', cluster_cols=False, cluster_rows=False)\n",
    "        heatmap.palette=sns.color_palette(sci_colour)\n",
    "        heatmap.plot()\n",
    "        pplot()\n",
    "        save_fig(f'vae{n}_genes-top-{num}-{label}')\n",
    "        plt.show()\n",
    "        \n",
    "for n in range(0, num_nodes):\n",
    "    with open(f'{ora_dir}vae-{n + 1}_genes-top10_{experiment_name}_{date}.csv', 'w+') as f:\n",
    "        f.write(gene_id + ',value\\n')\n",
    "        desc_sorted = (-vae_data[:,n]).argsort()  # Sort the genes by descending order\n",
    "        gene_names_sorted = df[gene_name].values[desc_sorted]\n",
    "        gene_ids_sorted = df[gene_id].values[desc_sorted]\n",
    "        vae_data_sorted = vae_data[:,n][desc_sorted]\n",
    "        plot_top_vae_genes(vae_data_sorted, gene_ids_sorted, gene_names_sorted, label=\"top\")\n",
    "        \n",
    "    with open(f'{ora_dir}vae-{n + 1}_genes-bottom10_{experiment_name}_{date}.csv', 'w+') as f:\n",
    "        f.write(gene_id + ',value\\n')\n",
    "        desc_sorted = (vae_data[:,n]).argsort()  # Sort the genes by descending order\n",
    "        gene_names_sorted = df[gene_name].values[desc_sorted]\n",
    "        vae_data_sorted = vae_data[:,n][desc_sorted]\n",
    "        gene_ids_sorted = df[gene_id].values[desc_sorted]\n",
    "        plot_top_vae_genes(vae_data_sorted, gene_ids_sorted, gene_names_sorted, label=\"bottom\")\n",
    "\n",
    "for n in range(0, num_nodes):\n",
    "    with open(f'{ora_dir}vae-{n + 1}_{experiment_name}_{date}.csv', 'w+') as f:\n",
    "        f.write(gene_id + ',value\\n')\n",
    "        desc_sorted = (vae_data[:,n]).argsort()  # Sort the genes by descending order\n",
    "        gene_names_sorted = df[gene_id].values[desc_sorted]\n",
    "        vae_data_sorted = vae_data[:,n][desc_sorted]\n",
    "        for i, g in enumerate(gene_names_sorted):\n",
    "            f.write(f'{g},{vae_data_sorted[i]}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -----------------------------------------------------------------------------------\n",
    "#                         Save genes in each of the clusters we annotated (ORA)\n",
    "# -----------------------------------------------------------------------------------\n",
    "for i in range(0, n_clusters):\n",
    "    with open(f'{ora_dir}cluster_{i}_gene-names_{experiment_name}_{date}.csv', 'w+') as f:\n",
    "        genes = df[gene_name].values[vae_set_idxs[i]]\n",
    "        rank_vae0 = vae_data[:, 0][vae_set_idxs[i]]\n",
    "        rank_vae1 = vae_data[:, 1][vae_set_idxs[i]]\n",
    "        rank_vae2 = vae_data[:, 2][vae_set_idxs[i]]\n",
    "        f.write(f'gene_name,VAE-1,VAE-2,VAE-3\\n')\n",
    "        for i, g in enumerate(genes):\n",
    "            f.write(f'{g},{rank_vae0[i]},{rank_vae1[i]},{rank_vae2[i]}\\n')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import string\n",
    "tissues = ['forebrain', 'midbrain', 'hindbrain', 'neural-tube', \n",
    "           'embryonic-facial-prominence', 'limb',\n",
    "           'heart', 'liver']\n",
    "marks = ['H3K9me3', 'H3K36me3', 'H3K27me3', 'H3K27ac', 'H3K4me1', 'H3K4me2', 'H3K4me3','H3K9ac']\n",
    "def get_median_naninc(df, mark, hist_metric=\"\", time_1=\"\", time_2=\"\", time_3=\"\", tissue=\"brain\"):\n",
    "    cols = []\n",
    "    for c in df.columns:\n",
    "        if tissue in c and mark in c and hist_metric in c and (time_1 in c or time_2 in c or time_3 in c):\n",
    "            cols.append(c)\n",
    "    # get nan median\n",
    "    vals = np.nanmedian(df[cols].values, axis=1)\n",
    "    return vals\n",
    "\n",
    "def plot_mark_heatmap(df, idxs, title, mark_cutoff=3.0):\n",
    "    mark_df = pd.DataFrame()\n",
    "    mark_values = []\n",
    "    mean_col = []\n",
    "    if idxs is None:\n",
    "        num_genes = len(df)\n",
    "    else:\n",
    "        num_genes = len(idxs) * 1.0\n",
    "    for t in [['10.5', '11.5', '12.5'], ['15.5', '15.5', '16.5']]:\n",
    "        tissue_titles = []\n",
    "        marks_title = []\n",
    "        for m in marks:\n",
    "            mark_col = []\n",
    "            for tissue in tissues:\n",
    "                median_all_data = get_median_naninc(df, m, \"signal\", t[0], t[1], t[2], tissue)\n",
    "                if idxs is None:\n",
    "                    median_genes = median_all_data\n",
    "                else:\n",
    "                    median_genes = median_all_data[idxs]     \n",
    "                has_mark = 1.0 * len(np.where(median_genes > mark_cutoff)[0])\n",
    "                mark_col.append(has_mark/num_genes) #np.nan_to_num(np.nanmean(median_genes)))\n",
    "                if string.capwords(tissue) not in tissue_titles:\n",
    "                    if tissue == 'embryonic-facial-prominence':\n",
    "                        if 'E.F.P' not in tissue_titles:\n",
    "                            tissue_titles.append('E.F.P')\n",
    "                    else:\n",
    "                        tissue_titles.append(string.capwords(tissue))\n",
    "            marks_title.append(string.capwords(m))\n",
    "            mark_df[string.capwords(m)] = mark_col\n",
    "\n",
    "        mark_df['Tissue'] = tissue_titles\n",
    "\n",
    "        heatmap = Heatmap(mark_df, marks_title, 'Tissue', vmin=0, vmax=None, cmap=hist_cmap, figsize=(2,2),\n",
    "                          title=f'{title} {\",\".join(t)}', cluster_rows=False, cluster_cols=False)\n",
    "        heatmap.plot()\n",
    "        pplot()\n",
    "        print(mark_df.head())\n",
    "        # vmin=0, vmax=1,\n",
    "        save_fig(f'mark_all_tissues_signal-{title}_{\"-\".join(t)}')\n",
    "        plt.show()\n",
    "        \n",
    "for i, gs in enumerate(vae_set_idxs):\n",
    "    plot_mark_heatmap(df, gs, f'Cluster {i}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7) Inspecting the data in context of chromHMM annotations\n",
    "\n",
    "Downloaded the pooled samples frmo http://enhancer.sdsc.edu/enhancer_export/ENCODE/chromHMM/pooled/\n",
    "for forebrain, hindbrain, and midbrian at e16.5 on 04 December 2020.\n",
    "\n",
    "\n",
    "http://enhancer.sdsc.edu/enhancer_export/ENCODE/chromHMM/readme\n",
    "\n",
    "The files in this directory contain chromHMM chromatin state calls for multiple tissues during mouse embryonic development, as described here: http://www.biorxiv.org/content/early/2017/07/21/166652.\n",
    "\n",
    "All files are in bed format. There is one file for each tissue at each developmental stage.\n",
    "The columns are 1) chromosome, 2) regions start, 3)  region end, 4) chromatin state number 5) Chromatin state label. \n",
    "There are 15 chromating states, labeled as follows:\n",
    "\n",
    "\n",
    "        State\tLabel\tDescription\n",
    "        1\tPr-A\tPromoter, Active\n",
    "        2\tPr-W\tPromoter, Weak/Inactive\n",
    "        3\tPr-B\tPromoter, Bivalent\n",
    "        4\tPr-F\tPromoter, Flanking\n",
    "        5\tEn-Sd\tEnhancer, Strong TSS-distal\n",
    "        6\tEn-Sp\tEnhancer, Strong TSS-proximal\n",
    "        7\tEn-W\tEnhancer, Weak TSS-distal\n",
    "        8\tEn-Pd\tEnhancer, Poised TSS-distal\n",
    "        9\tEn-Pp\tEnhancer, Poised TSS-proximal\n",
    "        10\tTr-S\tTranscription, Strong\n",
    "        11\tTr-P\tTranscription, Permissive\n",
    "        12\tTr-I\tTranscription, Initiation\n",
    "        13\tHc-P\tHeterochromatin, Polycomb-associated\n",
    "        14\tHc-H\tHeterochromatin, H3K9me3-associated\n",
    "        15\tNS\tNo Chromatin Signal\n",
    "        NRS\tNRS\tNo Reproducible State (ony relevant to \"replicated\" call set)\n",
    "\n",
    "\n",
    "There are four subdirectories:\n",
    "\n",
    "rep1\\\n",
    "This subdirectory contains chromatin state calls made using ChIP-seq datasets from biological replicate 1 (two biological replicates were performed for each ChIP-seq experiment).\n",
    "\n",
    "rep2\\\n",
    "This subdirectory contains chromatin state calls made using ChIP-seq datasets from biological replicate 2.\n",
    "\n",
    "replicated\\\n",
    "This subdirectory contains the intersection of the replicate 1 and replicate 2 chromatin state calls. \n",
    "Here, we require that a region is called in the same state in both replicates. If a region is not called in the same state in both replicates, it is labeled \"NRS\" for No Reproducible Signal. Note this is differenct than state 15, which is \"No Signal.\" \n",
    "\n",
    "pooled\\\n",
    "This subdirectory contains chromatin state calls on ChIP-seq data pooled from both replicates.\n",
    "We expect this will be the desired set of files for most users.\n",
    "\n",
    "\n",
    "Each subdirectory has two additional subdirectorites:\n",
    "\n",
    "archive\\\n",
    "This contains an older version of the bed files in which there are only state number, no labels. Aside from the labels, the files are identical to the working versions in the main directory.\n",
    "\n",
    "tracks\\\n",
    "This contains bigBed versions of the files for visualization on the UCSC Genome Browser.\n",
    "\n",
    "\n",
    "There is one additional subdirectory that does not conform to the templates decribed above. This is 6_mark_models\\.\n",
    "This subdirectory contains chromHMM segmentation of the e10.5 stage where we did not have all 8 marks. \n",
    "Note that the 11-state and 16-states models used for this segmentation are distinct from the 15-state (8-mark) model described above for the other stages. \n",
    "Please see manuscript and Extended Data Figure 7 for additional information.\n",
    "\n",
    "\n",
    "\n",
    "        header = \"chr,start,end,label,annotation\"\n",
    "        bed = Bed(os.path.join(self.data_dir, 'e16.5_forebrain_15_segments.bed'),\n",
    "                  header=None, overlap_method='in_promoter',\n",
    "                  output_bed_file=os.path.join(self.data_dir, 'e16.5_forebrain_15_segments_selectedPeaks.bed'),\n",
    "                  buffer_after_tss=0, buffer_before_tss=10, buffer_gene_overlap=0,\n",
    "                  gene_start=3, gene_end=4, gene_chr=2, gene_direction=5, gene_name=0,\n",
    "                  chr_idx=0, start_idx=1, end_idx=2, peak_value=4, header_extra=\"3\", sep='\\t'\n",
    "                  )\n",
    "        # Add the gene annot\n",
    "        bed.set_annotation_from_file(self.mm10_annot)\n",
    "        # Now we can run the assign values\n",
    "        bed.assign_locations_to_genes()\n",
    "        bed.save_loc_to_csv(f'{self.data_dir}test_e16.5_forebrain_15_segments.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -----------------------------------------------------------------------------------\n",
    "#                         Merge information with annotations\n",
    "# -----------------------------------------------------------------------------------\n",
    "\n",
    "chrom_dir = os.path.join(supp_dir, \"annot\")\n",
    "chromhmm_annot = pd.read_csv(f'{chrom_dir}/e16.5_forebrain_15_segments.csv')\n",
    "print(len(chromhmm_annot), len(df))\n",
    "# Make sure we only keep 1 annotation per gene\n",
    "chromhmm_annot = chromhmm_annot.groupby('external_gene_name').first()\n",
    "# Merge this with our dataframe\n",
    "df = df.merge(chromhmm_annot, on='external_gene_name', how='left', suffixes=('', '_chmm'))\n",
    "# Ensure the new length is the same as the old length\n",
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We also want to plot the annotations from gorkin et al for each of the clusters\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "def run_annot_plot(df_bg, df_fg, title=\"\"):\n",
    "    changes = []\n",
    "    order = ['Pr-A', 'Pr-W', 'Pr-B', 'Pr-F', 'En-Sd', 'En-Sp', \n",
    "         'En-W', 'En-Pd', 'En-Pp', 'Tr-S', 'Tr-P', 'Tr-I', \n",
    "          'Hc-P', 'Hc-H', 'NS']\n",
    "    total = len(df_bg) * 1.0\n",
    "    total_sig = len(df_fg) * 1.0\n",
    "    perc_all = []\n",
    "    perc_sig = []\n",
    "    \n",
    "    for o in order:\n",
    "        o_all = len(df_bg[df_bg['peak_value'] == o])\n",
    "        o_sig = len(df_fg[df_fg['peak_value'] == o])\n",
    "        perc_all.append(o_all)\n",
    "        perc_sig.append(o_sig)\n",
    "        if o_all == 0 and o_sig == 0:\n",
    "            changes.append(0)\n",
    "        else:\n",
    "            changes.append(((o_sig) - (o_all)))\n",
    "\n",
    "    odds_ratios = []\n",
    "    pvalues = []\n",
    "    for i, o in enumerate(order):\n",
    "        # Do a FET on each one\n",
    "        oddsratio, pvalue = stats.fisher_exact([[perc_sig[i], perc_all[i]], \n",
    "                                                [total_sig - perc_sig[i],\n",
    "                                                 total - perc_all[i]]])\n",
    "\n",
    "        print(o, oddsratio, pvalue,[[perc_sig[i], perc_all[i]], \n",
    "                                                [total_sig - perc_sig[i], \n",
    "                                                 total - perc_all[i]]])\n",
    "        odds_ratios.append(oddsratio)\n",
    "        pvalues.append(pvalue)\n",
    "\n",
    "    reg, padj, a, b = multipletests(pvalues, alpha=0.1, \n",
    "                                    method='fdr_bh', returnsorted=False)\n",
    "\n",
    "    p_sigs = []\n",
    "    for p in padj: \n",
    "        if p > 0.05:\n",
    "            p_sigs.append('')\n",
    "        elif p <= 0.05 and p > 0.01:\n",
    "            p_sigs.append('*')\n",
    "        elif p <= 0.01 and p > 0.001:\n",
    "            p_sigs.append('**')\n",
    "        elif p <= 0.001 and p > 0.0001:\n",
    "            p_sigs.append('***')\n",
    "        elif p <= 0.0001:\n",
    "            p_sigs.append('****')\n",
    "        else:\n",
    "            print(p)\n",
    "            p_sigs.append('')\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(1.5,1.0))\n",
    "    c = [grey] * 14\n",
    "    c[2] = \"green\"\n",
    "    c[12] = \"green\"\n",
    "    plt.bar(order, odds_ratios, color=c, linewidth=0.5, edgecolor='black')\n",
    "    rects = ax.patches\n",
    "    # Make some labels.\n",
    "    labels = [f'{p_sigs[i]}' for i in range(0, len(rects))]\n",
    "\n",
    "    for rect, label in zip(rects, labels):\n",
    "        height = rect.get_height()\n",
    "        ax.text(rect.get_x() + rect.get_width() / 2, height, label,\n",
    "                ha='center', va='bottom')\n",
    "    plt.ylim(0, 7)\n",
    "    plt.yticks(np.arange(0, 7, 2.0))\n",
    "    ax.set_title(f'Odds ratio FET for significant dataset {title} compared to all genes ChromHMM annotations')\n",
    "    ax.set_xlabel('', fontsize=6)\n",
    "    ax.set_ylabel('Odds ratio', fontsize=6)\n",
    "    ax.set_xticklabels(order, rotation=90, ha=\"center\")\n",
    "\n",
    "        \n",
    "    ax.tick_params(direction='out', length=2, width=0.5)\n",
    "    ax.spines['bottom'].set_linewidth(0.5)\n",
    "    ax.spines['top'].set_linewidth(0)\n",
    "    ax.spines['left'].set_linewidth(0.5)\n",
    "    ax.spines['right'].set_linewidth(0)\n",
    "    ax.tick_params(labelsize=6)\n",
    "    ax.tick_params(axis='x', which='major', pad=0)\n",
    "    ax.tick_params(axis='y', which='major', pad=0)\n",
    "\n",
    "    save_fig(f'OR-CM-{title}')\n",
    "    \n",
    "    plt.show()\n",
    "\n",
    "for gene_set_label in vae_set_labels:\n",
    "    run_annot_plot(df, df[df[gene_set_label] == True], title=f'Training vs {gene_set_label}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_bg = df_all.copy()  # Make a copy so we don't override the other one\n",
    "\n",
    "# Also have a look at how the marks plot over genes that are not significant\n",
    "chrom_dir = os.path.join(supp_dir, \"annot\")\n",
    "chromhmm_annot = pd.read_csv(f'{chrom_dir}/e16.5_forebrain_15_segments.csv')\n",
    "print(len(chromhmm_annot), len(df_bg))\n",
    "# Make sure we only keep 1 annotation per gene\n",
    "chromhmm_annot = chromhmm_annot.groupby('external_gene_name').first()\n",
    "# Merge this with our dataframe\n",
    "df_bg = df_bg.merge(chromhmm_annot, on='external_gene_name', how='left', suffixes=('', '_chmm'))\n",
    "# Ensure the new length is the same as the old length\n",
    "run_annot_plot(df_bg, df, title=f'Training vs All')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ) Epi genes annotation\n",
    "\n",
    "Using the information from: https://epifactors.autosome.ru/ (EpiFactors: a comprehensive database of human\n",
    "epigenetic factors and complexes, by Yulia A. Medvedeva et al) we test each cluser for enrichment for epigenetic complex's and functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "epi_genes = pd.read_csv(os.path.join(supp_dir, 'EpiGenes_main.csv'))\n",
    "df_epi = df.merge(epi_genes, left_on='external_gene_name', right_on='MGI_symbol',\n",
    "                      how='left', suffixes=('', '_eg'))\n",
    "    \n",
    "ax = sns.countplot(x=\"Function\", data=df_epi)\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=40, ha=\"right\")\n",
    "plt.tight_layout()\n",
    "save_fig(f'Epigenes-function')\n",
    "plt.show() \n",
    "ax = sns.countplot(x=\"Target\", data=df_epi)\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=40, ha=\"right\")\n",
    "plt.tight_layout()\n",
    "save_fig(f'Epigenes-target')\n",
    "plt.show() \n",
    "\n",
    "for set_label in vae_set_labels:\n",
    "    try:\n",
    "        vae_mse.u.dp([set_label])\n",
    "        ax = sns.countplot(x=\"Function\", data=df_epi[df_epi[set_label] == True])\n",
    "        ax.set_xticklabels(ax.get_xticklabels(), rotation=40, ha=\"right\")\n",
    "        plt.tight_layout()\n",
    "        save_fig(f'Epigenes-function-{set_label}')\n",
    "        plt.show() \n",
    "        ax = sns.countplot(x=\"Target\", data=df_epi[df_epi[set_label] == True])\n",
    "        ax.set_xticklabels(ax.get_xticklabels(), rotation=40, ha=\"right\")\n",
    "        plt.tight_layout()\n",
    "        save_fig(f'Epigenes-target-{set_label}')\n",
    "        plt.show() \n",
    "        # Complex_name Modification\n",
    "        ax = sns.countplot(x=\"Complex_name\", data=df_epi[df_epi[set_label] == True])\n",
    "        ax.set_xticklabels(ax.get_xticklabels(), rotation=40, ha=\"right\")\n",
    "        plt.tight_layout()\n",
    "        save_fig(f'Epigenes-Complex_name-{set_label}')\n",
    "        plt.show() \n",
    "        ax = sns.countplot(x=\"Modification\", data=df_epi[df_epi[set_label] == True])\n",
    "        ax.set_xticklabels(ax.get_xticklabels(), rotation=40, ha=\"right\")\n",
    "        plt.tight_layout()\n",
    "        save_fig(f'Epigenes-Modification-{set_label}')\n",
    "        plt.show() \n",
    "    except:\n",
    "        print(set_label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###############################################################################\n",
    "#                                                                             #\n",
    "#    This program is free software: you can redistribute it and/or modify     #\n",
    "#    it under the terms of the GNU General Public License as published by     #\n",
    "#    the Free Software Foundation, either version 3 of the License, or        #\n",
    "#    (at your option) any later version.                                      #\n",
    "#                                                                             #\n",
    "#    This program is distributed in the hope that it will be useful,          #\n",
    "#    but WITHOUT ANY WARRANTY; without even the implied warranty of           #\n",
    "#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            #\n",
    "#    GNU General Public License for more details.                             #\n",
    "#                                                                             #\n",
    "#    You should have received a copy of the GNU General Public License        #\n",
    "#    along with this program. If not, see <http://www.gnu.org/licenses/>.     #\n",
    "#                                                                             #\n",
    "###############################################################################\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "from sciviso import Vis\n",
    "\n",
    "\n",
    "class Violinplot(Vis):\n",
    "\n",
    "    def __init__(self, df: pd.DataFrame, x: object, y: object, title='', xlabel='', ylabel='', hue=None, order=None,\n",
    "                 hue_order=None, showfliers=False, add_dots=False, figsize=(1.5, 1.5), title_font_size=8,\n",
    "                 add_stats=False, stat_method='Mann-Whitney',\n",
    "                 label_font_size=6, title_font_weight=700):\n",
    "        super().__init__(df, figsize=figsize, title_font_size=title_font_size, label_font_size=label_font_size,\n",
    "                         title_font_weight=title_font_weight)\n",
    "        self.df = df\n",
    "        self.x = x\n",
    "        self.y = y\n",
    "        self.xlabel = xlabel\n",
    "        self.ylabel = ylabel\n",
    "        self.title = title\n",
    "        self.hue = hue\n",
    "        self.order = order\n",
    "        self.hue_order = hue_order\n",
    "        self.showfliers = showfliers\n",
    "        self.add_dots = add_dots\n",
    "        self.add_stats = add_stats\n",
    "        self.stat_method = stat_method\n",
    "\n",
    "    def plot(self):\n",
    "        x, y, hue, order, hue_order = self.x, self.y, self.hue, self.order, self.hue_order\n",
    "        if not isinstance(self.x, str) and not isinstance(self.y, str):\n",
    "            vis_df = pd.DataFrame()\n",
    "            vis_df['x'] = x\n",
    "            vis_df['y'] = y\n",
    "            x = 'x'\n",
    "            y = 'y'\n",
    "            if self.hue is not None:\n",
    "                vis_df['colour'] = self.hue\n",
    "                hue = 'colour'\n",
    "            if order is None:\n",
    "                order = list(set(vis_df['x'].values))\n",
    "                order.sort()\n",
    "        else:\n",
    "            vis_df = self.df\n",
    "        # set the orders\n",
    "        if hue_order is None and hue is not None:\n",
    "            hue_order = list(set(vis_df[hue].values))\n",
    "            hue_order.sort()\n",
    "        if order is None:\n",
    "            order = list(set(vis_df[x].values))\n",
    "            order.sort()\n",
    "\n",
    "        ax = sns.violinplot(data=vis_df, x=x, y=y, hue=hue, hue_order=hue_order, order=order, palette=self.palette,\n",
    "                            showfliers=self.showfliers, inner=\"quartile\")\n",
    "        if self.add_dots:\n",
    "            ax = sns.stripplot(data=vis_df, x=x, y=y, hue_order=hue_order, order=order, color='.2')\n",
    "\n",
    "        if self.add_stats:\n",
    "            # Add all pairs in the order if the box pairs is none\n",
    "\n",
    "            pairs = []\n",
    "            box_pairs = []\n",
    "            for i in order:\n",
    "                for j in order:\n",
    "                    if i != j:\n",
    "                        # Ensure we don't get duplicates\n",
    "                        pair = f'{i}{j}' if i < j else f'{j}{i}'\n",
    "                        if pair not in pairs:\n",
    "                            box_pairs.append((i, j))\n",
    "                            pairs.append(pair)\n",
    "            # Add stats annotation\n",
    "            add_stat_annotation(ax, data=vis_df, x=x, y=y, order=order,\n",
    "                                box_pairs=box_pairs,\n",
    "                                test=self.stat_method, text_format='star', loc='inside', verbose=2)\n",
    "        ax.set_xticklabels(ax.get_xticklabels(), rotation=45, horizontalalignment='right')\n",
    "        ax.tick_params(labelsize=self.label_font_size)\n",
    "        plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., fontsize=self.label_font_size)\n",
    "        self.add_labels()\n",
    "        self.set_ax_params(ax)\n",
    "        return ax"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Print out the TF relationships\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Have a look at the TFs and their targets for regulons (i.e. does grp2 have TFs that tagret group3?)\n",
    "regs = pd.read_csv(supp_dir + 'regulons_mm10_A-B.csv')\n",
    "reg_tfs = set(regs['tf'].values)\n",
    "group_labels = {1: 'Prolifferation', 2: 'Posterior', 3: 'Immune response', 4: 'Development', 5: 'Forebrain'}\n",
    "\n",
    "for i in range(1, n_clusters):\n",
    "    genes = df_training[gene_name].values[vae_set_idxs[i]]\n",
    "    # Flag the TFs in this group, we'll then test these to see if we have lots of the targets in another group\n",
    "    tfs = list(reg_tfs & set(genes)) \n",
    "    # Now we want to go through each of the other groups and record how many were targetted\n",
    "    u.dp([group_labels[i], f'\\n{len(tfs)} Tfs: ', ', '.join(tfs)])\n",
    "\n",
    "    for j in range(1, n_clusters):\n",
    "        genes_targets = df_training[gene_name].values[vae_set_idxs[j]]\n",
    "        total_targets = 0\n",
    "        for tf in tfs:\n",
    "            targets = set(regs[regs['tf'] == tf]['target'].values)\n",
    "            total_targets += len(list(targets & set(genes_targets)))\n",
    "            if len(list(targets & set(genes_targets))) > 0:\n",
    "                print(f'{tf} -->', ', '.join(list(targets & set(genes_targets))))\n",
    "        print(f'--------------- Total for {group_labels[i]} targeting {group_labels[j]}: {total_targets} -------------')\n",
    "\n",
    "        \n",
    "#regs = pd.read_csv(supp_dir + 'regulons_mm10_A-B-C-D.csv')\n",
    "regs[regs['tf'] == 'Cdkn1b'] #Tbx3, mesoderm markeds Hand1, gata2\n",
    "#df_sig = pd.read_csv(f'{input_dir}df-significant_epi-2500_20210124.csv')\n",
    "for g in regs[regs['target'] == 'Foxm1']['tf'].values:\n",
    "    if len(df_training[df_training[gene_name] == g]) > 0:\n",
    "        print(g, '\\n')\n",
    "        #print(df_sig[df_sig[gene_name] == g])\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
