{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dataset Generation for AP-axis analysis\n",
    "\n",
    "What are the genetic and regulatory mechanisms that drive the intense proliferation from the spinal-cord to the \n",
    "forebrain in bilateria?\n",
    "\n",
    "We know that the silencing of the Hox genes in the forebrain facilitates the growth of the tissue, however, do the \n",
    "Hox genes act in isolation? Or are there other genes that exhibit the same phenotype and thus, may be equally integral\n",
    "to this process?\n",
    "\n",
    "By removing the silencing mechanism (H3K27me3) we investigate which genes, act in a similar manner by integrating the \n",
    "RNAseq response to the knockout with epigenetic data.\n",
    "\n",
    "\n",
    "#### Cells appendix:\n",
    "\n",
    "    1) Imports\n",
    "    2) RNAseq merging files\n",
    "    3) Add annotation information\n",
    "    4) Save dataframe combinations to csv files for DE analysis\n",
    "    5) Run R scripts a. DeSEQ2 & b. Normalisation (in R) plot results from normalisation\n",
    "    6) Plot results from DEseq2\n",
    "    7) Make bar chart of all comparisons (i.e. how many DE genes in each comaprison)\n",
    "    8) Select log2(TMM + 1) as normalisation and build merged dataframe (raw + DEseq2 results)\n",
    "    9) Visualise the merged data top most significant genes (Heatmap and PCA) \n",
    "    10) Add histone modification data from Encode (downloaded & processed in APaxis_encodeDataDownload)\n",
    "    11) Filter the merged dataframe (i.e. for significant results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1) Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "--------------------------------------------------------\n",
    "Import RNAseq data & merge featureCounts files\n",
    "--------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "import os, sys\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import venn\n",
    "from matplotlib_venn import venn3\n",
    "\n",
    "from sciviso import *\n",
    "from sciutil import SciUtil\n",
    "\n",
    "u = SciUtil()\n",
    "\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "sys.path.append(module_path)\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "--------------------------------------------------------\n",
    "                     Global variables\n",
    "--------------------------------------------------------\n",
    "\"\"\"\n",
    "date = '20210124'\n",
    "\n",
    "gene_id = 'entrezgene_id'\n",
    "gene_name = 'external_gene_name'\n",
    "\n",
    "sci_colour = ['#483873', '#1BD8A6', '#B117B7', '#AAC7E2', '#FFC107', '#016957', '#9785C0', \n",
    "             '#D09139', '#338A03', '#FF69A1', '#5930B1', '#FFE884', '#35B567', '#1E88E5', \n",
    "             '#ACAD60', '#A2FFB4', '#B618F5', '#854A9C']\n",
    "\n",
    "hist_colour = '#483873'\n",
    "\n",
    "e11_colour = 'white'\n",
    "e13_colour = 'lightgrey'\n",
    "e15_colour = '#A7A9AC'\n",
    "e18_colour = '#58595B'\n",
    "\n",
    "fb_colour = '#ffbf80'\n",
    "mb_colour = '#ff8c1a'\n",
    "hb_colour = '#b35900'\n",
    "sc_colour = '#663300'\n",
    "grey = '#bdbdbd'\n",
    "\n",
    "h3k36me3_colour = '#CEF471'\n",
    "h3k27me3_colour = '#9F71F4'\n",
    "h3k4me3_colour = '#9F00FA'\n",
    "h3k4me2_colour = '#5930B1'\n",
    "h3k4me1_colour = '#FFE884'\n",
    "h3k27ac_colour = '#35B567'\n",
    "h3k9me3_colour = '#1E88E5'\n",
    "h3k9ac_colour = '#A2FFB4'\n",
    "           \n",
    "wt_colour = '#AADFF1'\n",
    "ko_colour = '#A53736'\n",
    "\n",
    "sns.palplot(sci_colour)\n",
    "sns.color_palette(sci_colour)\n",
    "\n",
    "project_name = 'prelim'\n",
    "\n",
    "data_dir = '../data/'\n",
    "r_dir = f'{data_dir}results/deseq2/'\n",
    "fig_dir = f'../figures/{project_name}/'\n",
    "output_dir = f'{data_dir}results/{project_name}/'\n",
    "input_dir = f'{data_dir}input/'\n",
    "rna_dir = f'{input_dir}feature_counts/'\n",
    "\n",
    "cmap = 'RdBu_r'\n",
    "u_id = 'u_id'\n",
    "sns.palplot(sci_colour)\n",
    "sns.color_palette(sci_colour)\n",
    "\n",
    "def get_time_colour(c):\n",
    "    if '11' in c or '10' in c:\n",
    "        return e11_colour\n",
    "    elif '13' in c or '12' in c:\n",
    "        return e13_colour\n",
    "    elif '15' in c or '14' in c:\n",
    "        return e15_colour\n",
    "    elif '18' in c or '16' in c:\n",
    "        return e18_colour\n",
    "    return '#FFFFFF'\n",
    "\n",
    "def get_tissue_colour(c):\n",
    "    if 'sc' in c or 'spinal' in c:\n",
    "        return sc_colour\n",
    "    elif 'hb' in c or 'hindbrain' in c:\n",
    "        return hb_colour\n",
    "    elif 'mb' in c or 'midbrain' in c:\n",
    "        return mb_colour\n",
    "    elif 'fb' in c or 'forebrain' in c:\n",
    "        return fb_colour\n",
    "    return '#FFFFFF'\n",
    "\n",
    "def get_cond_colour(c):\n",
    "    if 'ko' in c:\n",
    "        return ko_colour\n",
    "    elif 'wt' in c:\n",
    "        return wt_colour\n",
    "    return '#FFFFFF'\n",
    "\n",
    "def get_mark_colour(c):\n",
    "    if '36me3' in c:\n",
    "        return h3k36me3_colour\n",
    "    elif '27me3' in c:\n",
    "        return h3k27me3_colour\n",
    "    elif 'K4me3' in c:\n",
    "        return h3k4me3_colour\n",
    "    elif 'K4me2' in c:\n",
    "        return h3k4me2_colour\n",
    "    elif 'K4me1' in c:\n",
    "        return h3k4me1_colour\n",
    "    elif '27ac' in c:\n",
    "        return h3k27ac_colour\n",
    "    elif 'K9me3' in c:\n",
    "        return h3k9me3_colour\n",
    "    elif 'K9ac' in c:\n",
    "        return h3k9ac_colour\n",
    "    return '#FFFFFF'\n",
    "\n",
    "def pplot():\n",
    "    plt.rcParams['figure.figsize'] = [5, 4]\n",
    "    plt.rcParams['image.cmap'] = cmap\n",
    "    plt.rcParams['svg.fonttype'] = 'none'\n",
    "        \n",
    "def cplot():\n",
    "    plt.rcParams['figure.figsize'] = [6, 4]\n",
    "    sns.color_palette(sci_colour)\n",
    "    \n",
    "def save_fig(title, ending='.svg'):\n",
    "    plt.savefig(f'{fig_dir}{title.replace(\" \", \"-\")}{ending}')\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2) RNAseq merging files\n",
    "\n",
    "Here we import the results from our FeatureCounts data. We merge the files together on their entrez ID."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "files = os.listdir(rna_dir)\n",
    "\n",
    "rna_files = []\n",
    "for f in files:\n",
    "    if 'summary' not in f:\n",
    "        rna_files.append(f)\n",
    "\n",
    "# Make a df out of all expression data\n",
    "rna_files.sort()\n",
    "df = pd.DataFrame()\n",
    "        \n",
    "# Basically want to create a dummy df since we don't want the columns except the expression\n",
    "# and the entrez gene id\n",
    "tmp_df = pd.read_csv(f'{rna_dir}{rna_files[0]}', sep='\\t', header=1)\n",
    "\n",
    "df['Geneid'] = tmp_df['Geneid'].values\n",
    "df[rna_files[0][:-4]] = tmp_df[tmp_df.columns[-1]].values\n",
    "# Now we'll add all the other RNA files to this dataframe by joining on ID\n",
    "for filename in rna_files[1:]:\n",
    "    u.dp([\"Adding\", filename])\n",
    "    tmp_df = pd.DataFrame()\n",
    "    file_df = pd.read_csv(f'{rna_dir}{filename}', sep='\\t', header=1)\n",
    "    print(file_df.head())\n",
    "    u.dp([f'Length of {filename}: ', len(file_df)])\n",
    "    tmp_df['Geneid'] = file_df['Geneid'].values\n",
    "    # Note we take the log transform here \n",
    "    tmp_df[filename[:-4]] = file_df[file_df.columns[-1]].values\n",
    "    df = df.merge(tmp_df, on='Geneid', how='outer')\n",
    "\n",
    "# Save DF to csv for normalisation \n",
    "u.dp([\"Length of merged RNAseq dataframe: \", len(df)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3) Add annotation information\n",
    "\n",
    "We add annotation information to the genes: name, etc. This will also allow us to merge with our ChIP data which \n",
    "have been annotated with ensembl IDs.  \n",
    "\n",
    "We also want to save the merged dataframe to CSV files (grouped by the *tissue*) as then we will run DeSEQ2 on these \n",
    "count files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "--------------------------------------------------------\n",
    "Add annotation to RNAseq dataframe & save Tissue counts as files.\n",
    "--------------------------------------------------------\n",
    "\"\"\"\n",
    "# Lastly let's add some gene annotation information\n",
    "\n",
    "\"\"\"\n",
    "Annotations were generated using scibiomart v1.0.0 CLI\n",
    "\n",
    "scibiomart --m ENSEMBL_MART_ENSEMBL --d mmusculus_gene_ensembl --a \"ensembl_gene_id,entrezgene_id,external_gene_name,chromosome_name,start_position,end_position,strand\" --o mm10Sorted_ --s t\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "# Remove data that were not able to be annotated to gene names (i.e. these are usually the genes which are \"unknown\")\n",
    "\n",
    "annot_df = pd.read_csv(os.path.join(input_dir, 'supps', 'mm10Sorted_mmusculus_gene_ensembl-GRCm38.p6.csv'))\n",
    "gene_names = annot_df[gene_name].values\n",
    "gene_id_to_name = {}\n",
    "for i, g in enumerate(annot_df[gene_id].values):\n",
    "    if not gene_id_to_name.get(g):\n",
    "        gene_id_to_name[g] = gene_names[i]\n",
    "\n",
    "# add these to the df column\n",
    "gene_names = []\n",
    "pseudo_id = []\n",
    "df = df.rename(columns={'Geneid': 'entrezgene_id'})\n",
    "\n",
    "for g in df[gene_id].values:\n",
    "    gene_n = gene_id_to_name.get(g)\n",
    "    gene_names.append(gene_n)\n",
    "    pseudo_id.append(f'{g}-{gene_n}')\n",
    "df[gene_name] = gene_names\n",
    "df['u_id'] = pseudo_id\n",
    "\n",
    "# Organise columns a bit nicer\n",
    "cols = [gene_id, gene_name, 'u_id',\n",
    "        'ko11fb1', 'ko11fb2', 'ko13fb1', 'ko13fb2', 'ko15fb1', 'ko15fb2', 'ko18fb1', 'ko18fb2',\n",
    "        'ko11mb1', 'ko11mb2', 'ko13mb1', 'ko13mb2', 'ko15mb1', 'ko15mb2', 'ko18mb1', 'ko18mb2',\n",
    "        'ko11hb1', 'ko11hb2', 'ko13hb1', 'ko13hb2', 'ko15hb1', 'ko15hb2', 'ko18hb1', 'ko18hb2',\n",
    "        'ko11sc1', 'ko11sc2', 'ko13sc1', 'ko13sc2', 'ko15sc1', 'ko15sc2', 'ko18sc1', 'ko18sc2',\n",
    "         'wt11fb1', 'wt11fb2', 'wt13fb1', 'wt13fb2', 'wt15fb1', 'wt15fb2', 'wt18fb1', 'wt18fb2',\n",
    "        'wt11mb1', 'wt11mb2', 'wt13mb1', 'wt13mb2', 'wt15mb1', 'wt15mb2', 'wt18mb1', 'wt18mb2',\n",
    "        'wt11hb1', 'wt11hb2', 'wt13hb1', 'wt13hb2', 'wt15hb1', 'wt15hb2', 'wt18hb1', 'wt18hb2',\n",
    "        'wt11sc1', 'wt11sc2', 'wt13sc1', 'wt13sc2', 'wt15sc1', 'wt15sc2', 'wt18sc1', 'wt18sc2'\n",
    "]\n",
    "\n",
    "df = df[cols]\n",
    "u.dp([\"Number of genes in total:\", len(df)])\n",
    "# Remove the values that don't map to a gene name. On investigation these are \n",
    "# most commonly predicted genes, or non-coding RNA (which we don't expect to have messenger RNA for)\n",
    "# for example: 2610203C22Rik RIKEN cDNA 2610203C22 gene [ Mus musculus (house mouse) ]\n",
    "# There were about 6000 genes. None of which had a logFC > 2 with the majority having 0 LogFC\n",
    "df = df[df[gene_name].values != None]\n",
    "u.dp([\"Number of genes with gene names:\", len(df)])\n",
    "df.to_csv(f'{output_dir}merged_df_FEATURE_COUNTS_annot_{date}.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4) Save dataframe combinations to csv files\n",
    "\n",
    "Here we just save each combination that we want to run DE on to a CSV file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%% \n"
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Save dataframe combinations to CSV files to do DEseq2 analysis\n",
    "\"\"\"\n",
    "\n",
    "# Create DF for each tissue\n",
    "tissues = ['fb', 'mb', 'hb', 'sc']\n",
    "for t in tissues:\n",
    "    ts = [u_id]\n",
    "    for c in df.columns:\n",
    "        if t in c and '11' not in c and 'id' not in c:\n",
    "            ts.append(c)\n",
    "    t_df = df[ts]\n",
    "    t_df.to_csv(f'{r_dir}merged_df_{t}_FEATURE_COUNTS_{date}.csv', index=False)\n",
    "    \n",
    "# Create DF for each timepoint\n",
    "times = ['11', '13', '15', '18']\n",
    "for t in times:\n",
    "    ts = [u_id]\n",
    "    for c in df.columns:\n",
    "        if t in c and 'id' not in c and ('fb' in c or 'mb' in c):\n",
    "            ts.append(c)\n",
    "    t_df = df[ts]\n",
    "    t_df.to_csv(f'{r_dir}merged_df_anterior_{t}_FEATURE_COUNTS_{date}.csv', index=False)\n",
    "\n",
    "for t in times:\n",
    "    ts = [u_id]\n",
    "    for c in df.columns:\n",
    "        if t in c and 'id' not in c and ('sc' in c or 'hb' in c):\n",
    "            ts.append(c)\n",
    "    t_df = df[ts]\n",
    "    t_df.to_csv(f'{r_dir}merged_df_posterior_{t}_FEATURE_COUNTS_{date}.csv', index=False)\n",
    "\n",
    "    \n",
    "# Create df for all the data (including e11)\n",
    "columns = [gene_id]\n",
    "for c in df.columns:\n",
    "    if 'wt' in c or 'ko' in c:\n",
    "        columns.append(c)\n",
    "df[columns].to_csv(f'{r_dir}merged_df_FEATURE_COUNTS_{date}.csv', index=False)\n",
    "\n",
    "# Create combo between each tissue (i.e sc vs mb)\n",
    "def gen_csv_combo(cond1, cond2):\n",
    "    condition = ['wt', 'ko']\n",
    "    for t in condition:\n",
    "        ts = [u_id]\n",
    "        for c in df.columns:\n",
    "            if t in c and (cond1 in c or cond2 in c) and gene_id != c and gene_name != c and '11' not in c:\n",
    "                ts.append(c)\n",
    "        t_df = df[ts]\n",
    "        t_df.to_csv(f'{r_dir}merged_df_{t}_{cond1}-{cond2}_FEATURE_COUNTS_{date}.csv', index=False)\n",
    "        u.dp([\"Cond done:\", f'{t}_{cond1}-{cond2}'])\n",
    "\n",
    "def gen_csv_combo_time(cond1, cond2):\n",
    "    condition = ['wt', 'ko']\n",
    "    for t in condition:\n",
    "        ts = [u_id]\n",
    "        for c in df.columns:\n",
    "            if t in c and (cond1 in c or cond2 in c) and gene_id != c and gene_name != c and ('sc' in c or 'hb' in c):\n",
    "                ts.append(c)\n",
    "        t_df = df[ts]\n",
    "        t_df.to_csv(f'{r_dir}merged_df_posterior_{t}_{cond1}-{cond2}_FEATURE_COUNTS_{date}.csv', index=False)\n",
    "        u.dp([\"Cond done:\", f'posterior_{t}_{cond1}-{cond2}'])\n",
    "\n",
    "    for t in condition:\n",
    "        ts = [u_id]\n",
    "        for c in df.columns:\n",
    "            if t in c and (cond1 in c or cond2 in c) and gene_id != c and gene_name != c and ('fb' in c or 'mb' in c):\n",
    "                ts.append(c)\n",
    "        t_df = df[ts]\n",
    "        t_df.to_csv(f'{r_dir}merged_df_anterior_{t}_{cond1}-{cond2}_FEATURE_COUNTS_{date}.csv', index=False)\n",
    "        u.dp([\"Cond done:\", f'anterior_{t}_{cond1}-{cond2}'])\n",
    "        \n",
    "gen_csv_combo('fb', 'mb')\n",
    "gen_csv_combo('fb', 'hb')\n",
    "gen_csv_combo('fb', 'sc')\n",
    "gen_csv_combo('mb', 'sc')\n",
    "gen_csv_combo('mb', 'hb')\n",
    "gen_csv_combo('hb', 'sc')\n",
    "gen_csv_combo_time('11', '18')\n",
    "gen_csv_combo_time('11', '13')\n",
    "gen_csv_combo_time('13', '18')\n",
    "gen_csv_combo_time('15', '18')\n",
    "gen_csv_combo_time('11', '15')\n",
    "gen_csv_combo_time('13', '15')\n",
    "\n",
    "df.to_csv(f'{r_dir}merged_df_FEATURE_COUNTS_annot_{date}.csv', index=False)\n",
    "\n",
    "u.dp([\"Combinations generated!\"])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5) Run R scripts a. DeSEQ2 & b. Normalisation\n",
    "\n",
    "Here we need to step out of this notebook and run the following R markdowns:\n",
    "\n",
    "    2) Run Deseq2 (apAxis_between-cond-tissue.Rmd, apAxis_between-time.Rmd, apAxis_between-tissue.Rmd)\n",
    "    3) Normalise the data (apAxis_between-cond-tissue.Rmd)\n",
    "\n",
    "Prior to running our analysis, we need to normalise the RNAseq data. Given we're interested in integrating the changes across the genes, we use EdgeRs TMM method.\n",
    "\n",
    "### Script below for normalisation\n",
    "\n",
    "```\n",
    "Run script in file: apAxis_between-cond-tissue.Rmd\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "--------------------------------------------------------\n",
    "Display the normalisations \n",
    "--------------------------------------------------------\n",
    "\"\"\"\n",
    "%matplotlib inline\n",
    "from sciviso import Histogram, Scatterplot\n",
    "\n",
    "plot_on = True\n",
    "plt_on = True\n",
    "save_on = True\n",
    "# Show an example of each method's normalisation\n",
    "def show_scatter(df, cond_x, cond_y, x_label='', y_label='', title='', log2=False, offset=1, annotations=None): \n",
    "    a = 0.1\n",
    "    offset = 1\n",
    "    plt_df = pd.DataFrame()\n",
    "    x = cond_x\n",
    "    y = cond_y\n",
    "    if log2:\n",
    "        x = f'Log2({cond_x} + {offset})'\n",
    "        x = f'Log2({cond_y} + {offset})'\n",
    "        plt_df[x] = np.log2(df[cond_x].values + offset)\n",
    "        plt_df[y] = np.log2(df[cond_y].values + offset)\n",
    "    else:\n",
    "        plt_df[x] = df[cond_x].values \n",
    "        plt_df[y] = df[cond_y].values\n",
    "    \n",
    "    # plot histogram \n",
    "    h = Histogram(plt_df, x, title=title, fit_norm=False, plot_rug=True, xlabel=x)\n",
    "    h.plot()\n",
    "    save_fig(f'Hist_{x}_{title}', ending='.pdf')\n",
    "    plt.show()\n",
    "    # now run the scatter\n",
    "    s = Scatterplot(plt_df, x, y, title=title, xlabel=x, ylabel=y, colour=\"grey\", add_legend=False)\n",
    "    s.opacity=0.5\n",
    "    s.plot()\n",
    "    save_fig(f'Scatter_{x}-{y}_{title}', ending='.pdf')\n",
    "    plt.show()\n",
    "\n",
    "    \n",
    "deseq2 = pd.read_csv(f'{r_dir}merged_df_FEATURE_COUNTS_DEseq2Norm_{date}.csv')\n",
    "rlog = pd.read_csv(f'{r_dir}merged_df_FEATURE_COUNTS_rlog_{date}.csv')\n",
    "tmm = pd.read_csv(f'{r_dir}merged_df_FEATURE_COUNTS_tmm_{date}.csv')\n",
    "vst = pd.read_csv(f'{r_dir}merged_df_FEATURE_COUNTS_vst_{date}.csv')\n",
    "\n",
    "sc_genes = None\n",
    "show_scatter(deseq2, 'wt11fb1', 'ko11fb1', 'wt11fb1', 'ko11fb1', 'Forebrain 11 WT vs KO DEseq2', True, 1.0, sc_genes)\n",
    "show_scatter(deseq2, 'wt11fb1', 'wt11fb2', 'wt11fb1', 'wt11fb2', 'Forebrain WT vs WT DEseq2', True, 1.0, sc_genes)\n",
    "show_scatter(rlog, 'wt11fb1', 'ko11fb1', 'wt11fb1', 'ko11fb1', 'Forebrain 11 WT vs KO Rlog', False, 1.0, sc_genes)\n",
    "show_scatter(rlog, 'wt11fb1', 'wt11fb2', 'wt11fb1', 'wt11fb2', 'Forebrain WT vs WT Rlog', False, 1.0, sc_genes)\n",
    "show_scatter(tmm, 'wt11fb1', 'ko11fb1', 'wt11fb1', 'ko11fb1', 'Forebrain 11 WT vs KO TMM', True, 1.0, sc_genes)\n",
    "show_scatter(tmm, 'wt11fb1', 'wt11fb2', 'wt11fb1', 'wt11fb2', 'Forebrain WT vs WT TMM', True, 1.0, sc_genes)\n",
    "show_scatter(vst, 'wt11fb1', 'ko11fb1', 'wt11fb1', 'ko11fb1', 'Forebrain 11 WT vs KO VST', False, 1.0, sc_genes)\n",
    "show_scatter(vst, 'wt11fb1', 'wt11fb2', 'wt11fb1', 'wt11fb2', 'Forebrain WT vs WT VST', False, 1.0, sc_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6) Plot results from DEseq2\n",
    "\n",
    "Here we want to do the normal figures for a DE analysis i.e. volcano plots, heatmaps, & Venn diagrams etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from adjustText import adjust_text\n",
    "\n",
    "from sciviso import Vis\n",
    "\n",
    "\n",
    "class Volcanoplot(Vis):\n",
    "\n",
    "    def __init__(self, df: pd.DataFrame, log_fc: str, p_val: str, label_column: str, title='',\n",
    "                 xlabel='', ylabel='', invert=False, p_val_cutoff=0.05,\n",
    "                 log_fc_cuttoff=2, label_big_sig=False, colours=None, offset=None,\n",
    "                 text_colours=None, values_to_label=None, max_labels=20, values_colours=None,\n",
    "                 figsize=(2, 2), title_font_size=8, label_font_size=6, title_font_weight=700):\n",
    "        super().__init__(df, figsize=figsize, title_font_size=title_font_size, label_font_size=label_font_size,\n",
    "                         title_font_weight=title_font_weight)\n",
    "        super().__init__(df)\n",
    "        self.log_fc = log_fc\n",
    "        self.p_val = p_val\n",
    "        self.p_val_cutoff = p_val_cutoff\n",
    "        self.log_fc_cuttoff = log_fc_cuttoff\n",
    "        self.values_to_label = values_to_label\n",
    "        self.label_big_sig = label_big_sig\n",
    "        self.invert = invert\n",
    "        self.label_column = label_column\n",
    "        self.offset = offset\n",
    "        self.label = 'volcanoplot'\n",
    "        self.colours = {'ns_small-neg-logFC': 'lightgrey',\n",
    "                        'ns_small-pos-logFC': 'lightgrey',\n",
    "                        'ns_big-neg-logFC': 'grey',\n",
    "                        'ns_big-pos-logFC': 'grey',\n",
    "                        'sig_small-neg-logFC': '#ccddff',\n",
    "                        'sig_small-pos-logFC': '#ffcccc',\n",
    "                        'sig_big-neg-logFC': '#6699ff',\n",
    "                        'sig_big-pos-logFC': '#ff8080'} if colours is None else colours\n",
    "        self.xlabel = xlabel\n",
    "        self.ylabel = ylabel\n",
    "        self.title = title\n",
    "        self.max_labels = max_labels\n",
    "        self.figsize = figsize\n",
    "        self.values_colours = values_colours or {}\n",
    "        self.text_colours = text_colours or {}\n",
    "\n",
    "    def add_scatter_and_annotate(self, fig: plt, x_all: np.array, y_all: np.array,\n",
    "                                 colour: str, idxs: np.array, annotate=False, s=20):\n",
    "        x = x_all[idxs]\n",
    "        y = y_all[idxs]\n",
    "        ax = fig.scatter(x, y, c=colour, alpha=self.opacity, s=s, vmin=-10.0, vmax=10.0)\n",
    "\n",
    "        # Check if we want to annotate any of these with their gene IDs\n",
    "\n",
    "        if self.values_to_label is not None:\n",
    "            texts = []\n",
    "            labels = self.df[self.label_column].values[idxs]\n",
    "            for i, name in enumerate(labels):\n",
    "                if name in self.values_to_label:\n",
    "                    lbl_bg = self.values_colours.get(name)\n",
    "                    color = self.text_colours.get(name)\n",
    "                    texts.append(fig.text(x[i], y[i], name, color=color, fontsize=6,\n",
    "                                          bbox=dict(fc=lbl_bg, alpha=0.8, boxstyle='round,pad=0.25', lw=0)))\n",
    "            adjust_text(texts, force_text=2.0)\n",
    "        # Check if the user wants these labeled\n",
    "        if self.label_big_sig and annotate:\n",
    "            # If they do have a limit on the number of ones we show (i.e. we don't want 10000 gene names...)\n",
    "            max_values = -1 * self.max_labels\n",
    "            if len(y) < self.max_labels:\n",
    "                max_values = -1 * (len(y) - 1)\n",
    "            most_sig_idxs = np.argpartition(y, max_values)[max_values:]\n",
    "            labels = self.df[self.label_column].values[idxs][most_sig_idxs]\n",
    "            x = x[most_sig_idxs]\n",
    "            y = y[most_sig_idxs]\n",
    "            # We only label the ones with the max log fc\n",
    "            for i, name in enumerate(labels):\n",
    "                fig.annotate(name, (x[i], y[i]),\n",
    "                             xytext=(0, 10),\n",
    "                             textcoords='offset points', ha='center', va='bottom',\n",
    "                             bbox=dict(boxstyle='round,pad=0',\n",
    "                                       fc=None, alpha=0.2)\n",
    "                             )\n",
    "        return ax\n",
    "\n",
    "    def plot(self):\n",
    "        \"\"\"\n",
    "        For annotation styling see: https://matplotlib.org/api/pyplot_api.html#matplotlib.pyplot.annotate\n",
    "        Returns\n",
    "        -------\n",
    "\n",
    "        \"\"\"\n",
    "        # if offset is not given, make the offset the smallest value in the dataset\n",
    "        if not self.offset:\n",
    "            vals = self.df[self.p_val].values\n",
    "            self.offset = np.min(vals[np.nonzero(vals)])\n",
    "            self.u.warn_p(['No offset was provided, setting offset to be smallest value recorded in dataset: ',\n",
    "                           self.offset])\n",
    "\n",
    "        # x axis has log_fc, first only plot the values < cutoff\n",
    "        x = self.df[self.log_fc].values\n",
    "        y = -1 * np.log10(self.df[self.p_val].values + self.offset)\n",
    "\n",
    "        log_fc_np = self.df[self.log_fc].values\n",
    "        p_val_np = self.df[self.p_val].values\n",
    "\n",
    "        if self.invert:\n",
    "            x = -1 * np.log10(self.df[self.p_val].values + self.offset)\n",
    "            y = self.df[self.log_fc].values\n",
    "        sig_small_pos_logfc = np.where((p_val_np <= self.p_val_cutoff) & (np.abs(log_fc_np) < self.log_fc_cuttoff)\n",
    "                                       & (log_fc_np > 0))\n",
    "        sig_big_pos_logfc = np.where((p_val_np <= self.p_val_cutoff) & (np.abs(log_fc_np) >= self.log_fc_cuttoff)\n",
    "                                     & (log_fc_np > 0))\n",
    "\n",
    "        sig_small_neg_logfc = np.where((p_val_np <= self.p_val_cutoff) & (np.abs(log_fc_np) < self.log_fc_cuttoff)\n",
    "                                       & (log_fc_np <= 0))\n",
    "        sig_big_neg_logfc = np.where((p_val_np <= self.p_val_cutoff) & (np.abs(log_fc_np) >= self.log_fc_cuttoff)\n",
    "                                     & (log_fc_np <= 0))\n",
    "\n",
    "        # Plot the points\n",
    "        fig, ax = plt.subplots(figsize=self.figsize)\n",
    "        self.add_scatter_and_annotate(ax, x, y, self.colours['sig_small-pos-logFC'], sig_small_pos_logfc, s=10)\n",
    "        self.add_scatter_and_annotate(ax, x, y, self.colours['sig_big-pos-logFC'], sig_big_pos_logfc, \n",
    "                                      annotate=True, s=20)\n",
    "\n",
    "        # Negative\n",
    "        self.add_scatter_and_annotate(ax, x, y, self.colours['sig_small-neg-logFC'], sig_small_neg_logfc, s=10)\n",
    "        self.add_scatter_and_annotate(ax, x, y, self.colours['sig_big-neg-logFC'], sig_big_neg_logfc, \n",
    "                                      annotate=True, s=20)\n",
    "        #self.add_labels()\n",
    "        ax.tick_params(labelsize=6)\n",
    "        ax.tick_params(direction='out', length=2, width=0.5)\n",
    "        ax.spines['bottom'].set_linewidth(0.5)\n",
    "        ax.spines['top'].set_linewidth(0)\n",
    "        ax.spines['left'].set_linewidth(0.5)\n",
    "        ax.spines['right'].set_linewidth(0)\n",
    "        ax.tick_params(labelsize=6)\n",
    "        ax.tick_params(axis='x', which='major', pad=0)\n",
    "        ax.tick_params(axis='y', which='major', pad=0)\n",
    "        return ax\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Read in results from DEseq2 and format DFs nicer\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "df_dict = {}\n",
    "deseq2_files = os.listdir(r_dir)\n",
    "for f in deseq2_files:\n",
    "    if 'DEseq2' in f:\n",
    "        try:\n",
    "            de_df = pd.read_csv(os.path.join(r_dir, f))\n",
    "            de_df = de_df.rename(columns={de_df.columns[0]: 'u_id'})\n",
    "            gene_names = [s.split('-')[1] for s in list(de_df['u_id'].values)]\n",
    "            gene_ids = [s.split('-')[0] for s in list(de_df['u_id'].values)]\n",
    "            de_df['padj'] = de_df['padj'].fillna(1) # Replace Nan p values with 1.0s\n",
    "            # Replace Nans with 0's for other values\n",
    "            de_df = de_df.replace(np.nan, 0)\n",
    "            de_df[gene_id] = gene_ids\n",
    "            de_df[gene_name] = gene_names\n",
    "            df_dict[f] = de_df\n",
    "        except:\n",
    "            print(f)\n",
    "            \n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Set up markers to display on Volcano plot\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "fb_genes = ['Emx1', 'Eomes', 'Tbr1', 'Foxg1', 'Lhx6']\n",
    "mb_genes = ['En1', 'En2', 'Lmx1a', 'Bhlhe23', 'Sall4']\n",
    "hb_genes = ['Phox2b', 'Krox20', 'Fev', 'Hoxb1',  'Hoxd3']\n",
    "sc_genes = ['Hoxd8', 'Hoxd9', 'Hoxd10', 'Hoxd11','Hoxa7', 'Hoxa9', 'Hoxa10',\n",
    "            'Hoxb9', 'Hoxb13',  'Hoxc8', 'Hoxc9', 'Hoxc10', 'Hoxc11', 'Hoxc12', 'Hoxc13']\n",
    "progenitors = ['Sox2', 'Sox1', 'Sox3', 'Hes1', 'Hes5']\n",
    "neurons = ['Tubb3', 'Snap25', 'Syt1', 'Slc32a1','Slc17a6']\n",
    "glia = ['Pdgfra', 'Cspg4', 'Aqp4', 'Egfr', 'Slc6a11']\n",
    "\n",
    "vals_to_label = fb_genes + mb_genes + hb_genes + sc_genes + progenitors + neurons + glia\n",
    "lbl_colours = {}\n",
    "text_colours = {}\n",
    "\n",
    "for g in fb_genes:\n",
    "    lbl_colours[g] = fb_colour\n",
    "    text_colours[g] = 'black'\n",
    "for g in mb_genes:\n",
    "    lbl_colours[g] = mb_colour\n",
    "    text_colours[g] = 'black'\n",
    "for g in hb_genes:\n",
    "    lbl_colours[g] =  hb_colour\n",
    "    text_colours[g] = 'white'\n",
    "for g in sc_genes:\n",
    "    lbl_colours[g] = sc_colour\n",
    "    text_colours[g] = 'white'\n",
    "for g in progenitors:\n",
    "    lbl_colours[g] = e11_colour\n",
    "    text_colours[g] = 'black'\n",
    "for g in neurons:\n",
    "    lbl_colours[g] = e13_colour\n",
    "    text_colours[g] = 'black'\n",
    "for g in glia:\n",
    "    lbl_colours[g] = e15_colour \n",
    "    text_colours[g] = 'black'\n",
    "    \n",
    "def plot_volcano(df, title, save=True, show=True):\n",
    "    volcanoplot = Volcanoplot(df, 'log2FoldChange', 'padj', gene_name, \n",
    "                              title, 'Log 2 Fold change', '-log10(p adj)', \n",
    "                              p_val_cutoff=0.05,\n",
    "                              label_big_sig=False, log_fc_cuttoff=1.5, \n",
    "                              values_to_label=vals_to_label, figsize=(2,2),\n",
    "                              values_colours=lbl_colours, text_colours=text_colours)\n",
    "    sns.set_style(\"ticks\")\n",
    "    volcanoplot.plot()\n",
    "    if save:\n",
    "        save_fig(f'Volcano{title.replace(\" \", \"\")}', ending='.pdf')\n",
    "    if show:\n",
    "        plt.show()\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Plot volcanos\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "plot_volcano(df_dict[f'DEseq2_CNS_fb_{date}.csv'], \"FB EedcKO vs WT\")\n",
    "plot_volcano(df_dict[f'DEseq2_CNS_mb_{date}.csv'], \"MB EedcKO vs WT\")\n",
    "plot_volcano(df_dict[f'DEseq2_CNS_hb_{date}.csv'], \"HB EedcKO vs WT\")\n",
    "plot_volcano(df_dict[f'DEseq2_CNS_sc_{date}.csv'], \"SC EedcKO vs WT\")\n",
    "plot_volcano(df_dict[f'DEseq2_CNS_anterior_11_{date}.csv'], \"Anterior e11.5 EedcKO vs WT\")\n",
    "plot_volcano(df_dict[f'DEseq2_CNS_anterior_13_{date}.csv'], \"Anterior e13.5 EedcKO vs WT\")\n",
    "plot_volcano(df_dict[f'DEseq2_CNS_anterior_15_{date}.csv'], \"Anterior e15.5 EedcKO vs WT\")\n",
    "plot_volcano(df_dict[f'DEseq2_CNS_anterior_18_{date}.csv'], \"Anterior e18.5 EedcKO vs WT\")\n",
    "plot_volcano(df_dict[f'DEseq2_CNS_posterior_11_{date}.csv'], \"Posterior e11.5 EedcKO vs WT\")\n",
    "plot_volcano(df_dict[f'DEseq2_CNS_posterior_13_{date}.csv'], \"Posterior e13.5 EedcKO vs WT\")\n",
    "plot_volcano(df_dict[f'DEseq2_CNS_posterior_15_{date}.csv'], \"Posterior e15.5 EedcKO vs WT\")\n",
    "plot_volcano(df_dict[f'DEseq2_CNS_posterior_18_{date}.csv'], \"Posterior e18.5 EedcKO vs WT\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Plot venn diagrams\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "import venn\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "def plot_venn(gene_sets, labels, title, save=True, show=True, colours=None):\n",
    "    output = f'venn_{title.replace(\" \", \"-\")}'\n",
    "    dset = {}\n",
    "    for i, l in enumerate(labels):\n",
    "        dset[l] = gene_sets[i]\n",
    "    if len(gene_sets) > 3:\n",
    "        cmap = ListedColormap(sns.color_palette(colours))\n",
    "        venn.venn(dset, cmap=cmap, fontsize=6, figsize=(2, 2))\n",
    "    else:\n",
    "        venn3(gene_sets, set_labels=labels)\n",
    "    if save:\n",
    "        save_fig(output)\n",
    "    if show:\n",
    "        plt.show()\n",
    "\n",
    "fb_df = df_dict[f'DEseq2_CNS_fb_{date}.csv'][df_dict[f'DEseq2_CNS_fb_{date}.csv']['log2FoldChange'] > 0.5]\n",
    "mb_df = df_dict[f'DEseq2_CNS_mb_{date}.csv'][df_dict[f'DEseq2_CNS_mb_{date}.csv']['log2FoldChange'] > 0.5]\n",
    "hb_df = df_dict[f'DEseq2_CNS_hb_{date}.csv'][df_dict[f'DEseq2_CNS_hb_{date}.csv']['log2FoldChange'] > 0.5]\n",
    "sc_df = df_dict[f'DEseq2_CNS_sc_{date}.csv'][df_dict[f'DEseq2_CNS_sc_{date}.csv']['log2FoldChange'] > 0.5]\n",
    "\n",
    "fb_sig_genes = fb_df[fb_df['padj'] < 0.05][gene_name].values\n",
    "mb_sig_genes = mb_df[mb_df['padj'] < 0.05][gene_name].values\n",
    "hb_sig_genes = hb_df[hb_df['padj'] < 0.05][gene_name].values\n",
    "sc_sig_genes = sc_df[sc_df['padj'] < 0.05][gene_name].values\n",
    "\n",
    "plot_venn([set(fb_sig_genes), \n",
    "           set(mb_sig_genes), set(hb_sig_genes), set(sc_sig_genes)], \n",
    "          ['Forebrain WT vs KO', 'Midbrain WT vs KO', 'Hindbrain WT vs KO', 'Spinalcord WT vs KO'],\n",
    "          'tissue_WTvsKO',  colours=[fb_colour, mb_colour, hb_colour, sc_colour])\n",
    "\n",
    "plot_venn((set(fb_sig_genes), set(mb_sig_genes), set(sc_sig_genes)),\n",
    "          ('Forebrain', 'Midbrain', 'Spinal cord'), 'FbMbSc')\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Plot Anterior Venns\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "e11 = df_dict[f'DEseq2_CNS_anterior_11_{date}.csv'][df_dict[f'DEseq2_CNS_anterior_11_{date}.csv']['log2FoldChange'] > 0.5]\n",
    "e13 = df_dict[f'DEseq2_CNS_anterior_13_{date}.csv'][df_dict[f'DEseq2_CNS_anterior_13_{date}.csv']['log2FoldChange'] > 0.5]\n",
    "e15 = df_dict[f'DEseq2_CNS_anterior_15_{date}.csv'][df_dict[f'DEseq2_CNS_anterior_15_{date}.csv']['log2FoldChange'] > 0.5]\n",
    "e18 = df_dict[f'DEseq2_CNS_anterior_18_{date}.csv'][df_dict[f'DEseq2_CNS_anterior_18_{date}.csv']['log2FoldChange'] > 0.5]\n",
    "\n",
    "e11_sig_genes = e11[e11['padj'] < 0.05][gene_name].values\n",
    "e13_sig_genes = e13[e13['padj'] < 0.05][gene_name].values\n",
    "e15_sig_genes = e15[e15['padj'] < 0.05][gene_name].values\n",
    "e18_sig_genes = e18[e18['padj'] < 0.05][gene_name].values\n",
    "\n",
    "plot_venn([set(e11_sig_genes), \n",
    "           set(e13_sig_genes), set(e15_sig_genes), set(e18_sig_genes)], \n",
    "          ['e11 WT vs KO', 'e13 WT vs KO', 'e15 WT vs KO', 'e18 WT vs KO'],\n",
    "          'posterior_time_WTvsKO', colours=[e11_colour, e13_colour, e15_colour, e18_colour])\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Plot posterior venn diagrams\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "e11 = df_dict[f'DEseq2_CNS_posterior_11_{date}.csv'][df_dict[f'DEseq2_CNS_posterior_11_{date}.csv']['log2FoldChange'] > 0.5]\n",
    "e13 = df_dict[f'DEseq2_CNS_posterior_13_{date}.csv'][df_dict[f'DEseq2_CNS_posterior_13_{date}.csv']['log2FoldChange'] > 0.5]\n",
    "e15 = df_dict[f'DEseq2_CNS_posterior_15_{date}.csv'][df_dict[f'DEseq2_CNS_posterior_15_{date}.csv']['log2FoldChange'] > 0.5]\n",
    "e18 = df_dict[f'DEseq2_CNS_posterior_18_{date}.csv'][df_dict[f'DEseq2_CNS_posterior_18_{date}.csv']['log2FoldChange'] > 0.5]\n",
    "\n",
    "e11_sig_genes = e11[e11['padj'] < 0.05][gene_name].values\n",
    "e13_sig_genes = e13[e13['padj'] < 0.05][gene_name].values\n",
    "e15_sig_genes = e15[e15['padj'] < 0.05][gene_name].values\n",
    "e18_sig_genes = e18[e18['padj'] < 0.05][gene_name].values\n",
    "\n",
    "plot_venn([set(e11_sig_genes), \n",
    "           set(e13_sig_genes), set(e15_sig_genes), set(e18_sig_genes)], \n",
    "          ['e11 WT vs KO', 'e13 WT vs KO', 'e15 WT vs KO', 'e18 WT vs KO'],\n",
    "          'posterior_time_WTvsKO', colours=[e11_colour, e13_colour, e15_colour, e18_colour])\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Print out the number of significant genes\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "fb_sig_genes = fb_df[fb_df['padj'] < 0.05][gene_name].values\n",
    "mb_sig_genes = mb_df[mb_df['padj'] < 0.05][gene_name].values\n",
    "hb_sig_genes = hb_df[hb_df['padj'] < 0.05][gene_name].values\n",
    "sc_sig_genes = sc_df[sc_df['padj'] < 0.05][gene_name].values\n",
    "u.dp([\"P.adj < 0.05 for SC: KO vs WT\", len(sc_sig_genes)])\n",
    "u.dp([\"P.adj < 0.05 for HB: KO vs WT\", len(hb_sig_genes)])\n",
    "u.dp([\"P.adj < 0.05 for MB: KO vs WT\", len(mb_sig_genes)])\n",
    "u.dp([\"P.adj < 0.05 for FB: KO vs WT\", len(fb_sig_genes)])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7) Make bar chart of all comparisons\n",
    "\n",
    "Have a look at the comaprisons and just plot the number of DEGs for a summary figure.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Bar charts for tissue specific DE \n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "dfs = [df_dict[f'DEseq2_CNS_wt_fb-mb_{date}.csv'], df_dict[f'DEseq2_CNS_ko_fb-mb_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_wt_mb-hb_{date}.csv'], df_dict[f'DEseq2_CNS_ko_mb-hb_{date}.csv'],\n",
    "       df_dict[f'DEseq2_CNS_wt_hb-sc_{date}.csv'], df_dict[f'DEseq2_CNS_ko_hb-sc_{date}.csv'],\n",
    "       df_dict[f'DEseq2_CNS_wt_mb-sc_{date}.csv'], df_dict[f'DEseq2_CNS_ko_mb-sc_{date}.csv'],\n",
    "       df_dict[f'DEseq2_CNS_wt_fb-hb_{date}.csv'], df_dict[f'DEseq2_CNS_ko_fb-hb_{date}.csv'],\n",
    "       df_dict[f'DEseq2_CNS_wt_fb-sc_{date}.csv'], df_dict[f'DEseq2_CNS_ko_fb-sc_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_fb_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_mb_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_hb_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_sc_{date}.csv']\n",
    "      ]\n",
    "\n",
    "labels = ['Forebrain vs Midbrain WT',\n",
    "          'Forebrain vs Midbrain KO',\n",
    "          \n",
    "          'Midbrain vs Hindbrain WT',\n",
    "          'Midbrain vs Hindbrain KO',\n",
    "          \n",
    "          'Hindbrain vs Spinalcord WT',\n",
    "          'Hindbrain vs Spinalcord KO',\n",
    "          \n",
    "          'Midbrain vs Spinalcord WT',\n",
    "          'Midbrain vs Spinalcord KO',\n",
    "          \n",
    "          'Forebrain vs Hindbrain WT', \n",
    "          'Forebrain vs Hindbrain KO',\n",
    "          \n",
    "          'Forebrain vs Spinal cord WT',\n",
    "          'Forebrain vs Spinalcord KO',\n",
    "          \n",
    "          'WT vs KO (Forebrain)',\n",
    "          'WT vs KO (Midbrain)',\n",
    "          'WT vs KO (Hindbrain)', \n",
    "          'WT vs KO (Spinal cord)',\n",
    "         ]\n",
    "i = 0\n",
    "values = []\n",
    "logfc_cutoff = 0.5\n",
    "\n",
    "for d in dfs:\n",
    "    sigD = d[d['padj'] < 0.05]\n",
    "    sigD = sigD[abs(sigD['log2FoldChange']) > logfc_cutoff]\n",
    "    num_sig = len(sigD)\n",
    "    values.append(num_sig)\n",
    "    print(f'{num_sig}, {labels[i]}')\n",
    "    i += 1\n",
    "    \n",
    "# Plot a simple bar chart.\n",
    "barchart = Barchart(df, labels, values, order=labels, figsize=(1,1))\n",
    "barchart.palette=sns.color_palette([wt_colour, ko_colour, wt_colour, ko_colour, wt_colour, ko_colour,\n",
    "                                   wt_colour, ko_colour, wt_colour, ko_colour, wt_colour, ko_colour, fb_colour, mb_colour, hb_colour, sc_colour])\n",
    "barchart.figsize=(2,1)\n",
    "barchart.plot()\n",
    "save_fig(f'Barchart_numSigs')\n",
    "u.dp([\"Barchart used p.adj cutoff:\", 0.05, \"\\nLogFC cutoff:\", logfc_cutoff])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Bar charts for time specific DE (anterior)\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "dfs = [df_dict[f'DEseq2_CNS_anterior_wt_11-13_{date}.csv'], df_dict[f'DEseq2_CNS_anterior_ko_11-13_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_anterior_wt_13-15_{date}.csv'], df_dict[f'DEseq2_CNS_anterior_ko_13-15_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_anterior_wt_15-18_{date}.csv'], df_dict[f'DEseq2_CNS_anterior_ko_15-18_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_anterior_wt_11-15_{date}.csv'], df_dict[f'DEseq2_CNS_anterior_ko_11-15_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_anterior_wt_13-18_{date}.csv'], df_dict[f'DEseq2_CNS_anterior_ko_13-18_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_anterior_wt_11-18_{date}.csv'], df_dict[f'DEseq2_CNS_anterior_ko_11-18_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_anterior_11_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_anterior_13_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_anterior_15_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_anterior_18_{date}.csv']\n",
    "      ]\n",
    "\n",
    "labels = ['e11 vs e13 (WT)', 'e11 vs e13 (KO)',\n",
    "          'e13 vs e15 (WT)', 'e13 vs e15 (KO)',\n",
    "          'e15 vs e18 (WT)', 'e15 vs e18 (KO)',\n",
    "          \n",
    "          'e11 vs e15 (WT)', 'e11 vs e15 (KO)',\n",
    "          'e13 vs e18 (WT)', 'e13 vs e18 (KO)',\n",
    "          \n",
    "          'e11 vs e18 (WT)', 'e11 vs e18 (KO)',\n",
    "\n",
    "          'WT vs KO (11)', 'WT vs KO (13)', 'WT vs KO (15)','WT vs KO (18)'\n",
    "         ]\n",
    "i = 0\n",
    "values = []\n",
    "for d in dfs:\n",
    "    sigD = d[d['padj'] < 0.05]\n",
    "    sigD = sigD[abs(sigD['log2FoldChange']) > logfc_cutoff]\n",
    "    num_sig = len(sigD)\n",
    "    values.append(num_sig)\n",
    "    print(f'{num_sig}, {labels[i]}')\n",
    "    i += 1\n",
    "    \n",
    "# Plot a simple bar chart.\n",
    "barchart = Barchart(df, labels, values, title='Temporal genes anterior', order=labels)\n",
    "barchart.palette=sns.color_palette([wt_colour, ko_colour, wt_colour, ko_colour, wt_colour, ko_colour,\n",
    "                                   wt_colour, ko_colour, wt_colour, ko_colour, wt_colour, ko_colour, \n",
    "                                    e11_colour, e13_colour, e15_colour, e18_colour])\n",
    "\n",
    "barchart.plot()\n",
    "save_fig(f'Barchart_anterior_time_numSigs')\n",
    "\n",
    "u.dp([\"Barchart used p.adj cutoff:\", 0.05, \"\\nLogFC cutoff:\", logfc_cutoff])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Bar charts for time specific (anterior & posterior) \n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "dfs = [\n",
    "       df_dict[f'DEseq2_CNS_anterior_11_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_anterior_13_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_anterior_15_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_anterior_18_{date}.csv'],\n",
    "       df_dict[f'DEseq2_CNS_posterior_11_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_posterior_13_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_posterior_15_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_posterior_18_{date}.csv']\n",
    "       \n",
    "      ]\n",
    "\n",
    "labels = [ \n",
    "        'A. WT vs KO (11)', 'A. WT vs KO (13)', 'A. WT vs KO (15)','A. WT vs KO (18)',\n",
    "        'P. WT vs KO (11)', 'P. WT vs KO (13)', 'P. WT vs KO (15)','P. WT vs KO (18)'\n",
    "         ]\n",
    "i = 0\n",
    "values = []\n",
    "logfc_cutoff = 0.5\n",
    "for d in dfs:\n",
    "    sigD = d[d['padj'] < 0.05]\n",
    "    sigD = sigD[abs(sigD['log2FoldChange']) > logfc_cutoff]\n",
    "    num_sig = len(sigD)\n",
    "    values.append(num_sig)\n",
    "    print(f'{num_sig}, {labels[i]}')\n",
    "    i += 1\n",
    "    \n",
    "# Plot a simple bar chart.\n",
    "barchart = Barchart(dfs[0], labels, values, title='Temporal genes anterior', order=labels)\n",
    "barchart.palette=sns.color_palette([e11_colour, e13_colour, e15_colour, e18_colour,\n",
    "                                    e11_colour, e13_colour, e15_colour, e18_colour])\n",
    "barchart.plot()\n",
    "save_fig('Barchart_anterior-posterior_time_numSigs')\n",
    "\n",
    "u.dp([\"Barchart used p.adj cutoff:\", 0.05, \"\\nLogFC cutoff:\", logfc_cutoff])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Bar charts for time specific DE (posterior)\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "dfs = [df_dict[f'DEseq2_CNS_posterior_wt_11-13_{date}.csv'], df_dict[f'DEseq2_CNS_posterior_ko_11-13_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_posterior_wt_13-15_{date}.csv'], df_dict[f'DEseq2_CNS_posterior_ko_13-15_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_posterior_wt_15-18_{date}.csv'], df_dict[f'DEseq2_CNS_posterior_ko_15-18_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_posterior_wt_11-15_{date}.csv'], df_dict[f'DEseq2_CNS_posterior_ko_11-15_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_posterior_wt_13-18_{date}.csv'], df_dict[f'DEseq2_CNS_posterior_ko_13-18_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_posterior_wt_11-18_{date}.csv'], df_dict[f'DEseq2_CNS_posterior_ko_11-18_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_posterior_11_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_posterior_13_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_posterior_15_{date}.csv'], \n",
    "       df_dict[f'DEseq2_CNS_posterior_18_{date}.csv']\n",
    "      ]\n",
    "\n",
    "labels = ['e11 vs e13 (WT)', 'e11 vs e13 (KO)',\n",
    "          'e13 vs e15 (WT)', 'e13 vs e15 (KO)',\n",
    "          'e15 vs e18 (WT)', 'e15 vs e18 (KO)',\n",
    "          \n",
    "          'e11 vs e15 (WT)', 'e11 vs e15 (KO)',\n",
    "          'e13 vs e18 (WT)', 'e13 vs e18 (KO)',\n",
    "          \n",
    "          'e11 vs e18 (WT)', 'e11 vs e18 (KO)',\n",
    "\n",
    "          'WT vs KO (11)', 'WT vs KO (13)', 'WT vs KO (15)','WT vs KO (18)'\n",
    "         ]\n",
    "i = 0\n",
    "values = []\n",
    "logfc_cutoff = 1.0\n",
    "for d in dfs:\n",
    "    sigD = d[d['padj'] < 0.05]\n",
    "    sigD = sigD[abs(sigD['log2FoldChange']) > logfc_cutoff]\n",
    "    num_sig = len(sigD)\n",
    "    values.append(num_sig)\n",
    "    print(f'{num_sig}, {labels[i]}')\n",
    "    i += 1\n",
    "    \n",
    "# Plot a simple bar chart.\n",
    "barchart = Barchart(df, labels, values, title='Temporal genes posterior', order=labels)\n",
    "barchart.palette=sns.color_palette([wt_colour, ko_colour, wt_colour, ko_colour, wt_colour, ko_colour,\n",
    "                                   wt_colour, ko_colour, wt_colour, ko_colour, wt_colour, ko_colour, \n",
    "                                    e11_colour, e13_colour, e15_colour, e18_colour])\n",
    "barchart.plot()\n",
    "save_fig(f'Barchart_posterior_time_numSigs')\n",
    "u.dp([\"Barchart used p.adj cutoff:\", 0.05, \"\\nLogFC cutoff:\", logfc_cutoff])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8) Select log2(TMM + 1) as normalisation\n",
    "\n",
    "We choose the normalisation to be the log2(TMM + 1) as we are keen to look across the genes rather than just run \n",
    "differential expression between the genes. As such, we need to load this add the information from the various \n",
    "differential expression analyses.\n",
    "\n",
    "We also merge our WT vs KO experiments from before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "--------------------------------------------------------\n",
    "Choose normalisation method for data & add annoation\n",
    "--------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "        \n",
    "# Read in normalised RNAseq data\n",
    "tmm = pd.read_csv(f'{r_dir}merged_df_FEATURE_COUNTS_tmm_{date}.csv')\n",
    "\n",
    "tmm.rename(columns={ tmm.columns[0]: gene_id }, inplace = True)\n",
    "\n",
    "for c in tmm:\n",
    "    if c != gene_id:\n",
    "        tmm[c] = np.log2(tmm[c].values + 1)\n",
    "\n",
    "df_all = tmm.copy()\n",
    "df_all[gene_id] = pd.to_numeric(df_all[gene_id])\n",
    "\n",
    "# Read in each our our DE experiments from DEseq2\n",
    "fb_df = df_dict[f'DEseq2_CNS_fb_{date}.csv']\n",
    "mb_df = df_dict[f'DEseq2_CNS_mb_{date}.csv']\n",
    "hb_df = df_dict[f'DEseq2_CNS_hb_{date}.csv']\n",
    "sc_df = df_dict[f'DEseq2_CNS_sc_{date}.csv']\n",
    "\n",
    "a11_df = df_dict[f'DEseq2_CNS_anterior_11_{date}.csv']\n",
    "a13_df = df_dict[f'DEseq2_CNS_anterior_13_{date}.csv']\n",
    "a15_df = df_dict[f'DEseq2_CNS_anterior_15_{date}.csv']\n",
    "a18_df = df_dict[f'DEseq2_CNS_anterior_18_{date}.csv']\n",
    "\n",
    "p11_df = df_dict[f'DEseq2_CNS_posterior_11_{date}.csv']\n",
    "p13_df = df_dict[f'DEseq2_CNS_posterior_13_{date}.csv']\n",
    "p15_df = df_dict[f'DEseq2_CNS_posterior_15_{date}.csv']\n",
    "p18_df = df_dict[f'DEseq2_CNS_posterior_18_{date}.csv']\n",
    "\n",
    "# Let's add all our significant results (and make sure we just indicate where they came from)\n",
    "fb_df[gene_id] = pd.to_numeric(fb_df[gene_id])\n",
    "df_all = df_all.merge(fb_df, how='left', left_on=gene_id, \n",
    "                                       right_on=gene_id, suffixes=('', '_fb'))\n",
    "mb_df[gene_id] = pd.to_numeric(mb_df[gene_id])\n",
    "df_all = df_all.merge(mb_df, how='left', left_on=gene_id, \n",
    "                                       right_on=gene_id, suffixes=('', '_mb'))\n",
    "hb_df[gene_id] = pd.to_numeric(hb_df[gene_id])\n",
    "df_all = df_all.merge(hb_df, how='left', left_on=gene_id, \n",
    "                                       right_on=gene_id, suffixes=('', '_hb'))\n",
    "sc_df[gene_id] = pd.to_numeric(sc_df[gene_id])\n",
    "df_all = df_all.merge(sc_df, how='left', left_on=gene_id, \n",
    "                                       right_on=gene_id, suffixes=('', '_sc'))\n",
    "\n",
    "# Now we also want to add in the results from the anterior and posterior temporal changes\n",
    "a11_df[gene_id] = pd.to_numeric(a11_df[gene_id])\n",
    "df_all = df_all.merge(a11_df, how='left', left_on=gene_id, \n",
    "                                       right_on=gene_id, suffixes=('', '_a11'))\n",
    "a13_df[gene_id] = pd.to_numeric(a13_df[gene_id])\n",
    "df_all = df_all.merge(a13_df, how='left', left_on=gene_id, \n",
    "                                       right_on=gene_id, suffixes=('', '_a13'))\n",
    "a15_df[gene_id] = pd.to_numeric(a15_df[gene_id])\n",
    "df_all = df_all.merge(a15_df, how='left', left_on=gene_id, \n",
    "                                       right_on=gene_id, suffixes=('', '_a15'))\n",
    "a18_df[gene_id] = pd.to_numeric(a18_df[gene_id])\n",
    "df_all = df_all.merge(a18_df, how='left', left_on=gene_id, \n",
    "                                       right_on=gene_id, suffixes=('', '_a18'))\n",
    "\n",
    "p11_df[gene_id] = pd.to_numeric(p11_df[gene_id])\n",
    "df_all = df_all.merge(p11_df, how='left', left_on=gene_id, \n",
    "                                       right_on=gene_id, suffixes=('', '_p11'))\n",
    "p13_df[gene_id] = pd.to_numeric(p13_df[gene_id])\n",
    "df_all = df_all.merge(p13_df, how='left', left_on=gene_id, \n",
    "                                       right_on=gene_id, suffixes=('', '_p13'))\n",
    "p15_df[gene_id] = pd.to_numeric(p15_df[gene_id])\n",
    "df_all = df_all.merge(p15_df, how='left', left_on=gene_id, \n",
    "                                       right_on=gene_id, suffixes=('', '_p15'))\n",
    "p18_df[gene_id] = pd.to_numeric(p18_df[gene_id])\n",
    "df_all = df_all.merge(p18_df, how='left', left_on=gene_id, \n",
    "                                       right_on=gene_id, suffixes=('', '_p18'))\n",
    "\n",
    "# Rename the padj from forebrain and logfc\n",
    "df_all.rename(columns={'padj': 'padj_fb', 'log2FoldChange': 'log2FoldChange_fb'}, inplace=True)\n",
    "\n",
    "# Drop columns that don't have gene names --> leave this out for now.\n",
    "#df_all = df_all.dropna(subset=[gene_name])\n",
    "\n",
    "# This will be on p values that weren't in there \n",
    "p_cols = [c for c in df_all.columns if 'padj' in c]\n",
    "\n",
    "# Make them NS i.e. given them a p value of 1.0\n",
    "df_all[p_cols] = df_all[p_cols].fillna(value=1.0)\n",
    "\n",
    "# Fill everything else with 0's so we don't have NaNs in our counts etc\n",
    "df_all = df_all.fillna(value=0)\n",
    "df_all = df_all.reset_index()  # Ensure the index is reset each time we rejoin everything\n",
    "u.dp(['Length of dataframe', len(df_all)])\n",
    "\n",
    "# We don't want duplicate gene IDs so we drop columns with duplicates in these values\n",
    "no_na_merged = df_all.drop_duplicates()\n",
    "\n",
    "# Collect entrez IDs\n",
    "ens_ids = annot_df['ensembl_gene_id'].values\n",
    "chr_ids = annot_df['chromosome_name'].values\n",
    "gene_id_to_ens = {}\n",
    "gene_id_to_chr = {}\n",
    "for i, g in enumerate(annot_df[gene_id].values):\n",
    "    if not gene_id_to_ens.get(g):\n",
    "        gene_id_to_ens[g] = ens_ids[i]\n",
    "    if not gene_id_to_chr.get(g):\n",
    "        gene_id_to_chr[g] = chr_ids[i]\n",
    "ens_ids = []\n",
    "chrs = []\n",
    "for g in no_na_merged[gene_id].values:\n",
    "    ens_ids.append(gene_id_to_ens.get(g))\n",
    "    chrs.append(gene_id_to_chr.get(g))\n",
    "# Add ensembl gene id \n",
    "no_na_merged['ensembl_gene_id'] = ens_ids\n",
    "no_na_merged['chrs'] = chrs\n",
    "\n",
    "# Organise columns a bit nicer\n",
    "cols = [gene_id, gene_name, 'ensembl_gene_id', 'chrs',\n",
    "        'wt11fb1', 'wt11fb2', 'wt13fb1', 'wt13fb2', 'wt15fb1', 'wt15fb2', 'wt18fb1', 'wt18fb2',\n",
    "        'wt11mb1', 'wt11mb2', 'wt13mb1', 'wt13mb2', 'wt15mb1', 'wt15mb2', 'wt18mb1', 'wt18mb2',\n",
    "        'wt11hb1', 'wt11hb2', 'wt13hb1', 'wt13hb2', 'wt15hb1', 'wt15hb2', 'wt18hb1', 'wt18hb2',\n",
    "        'wt11sc1', 'wt11sc2', 'wt13sc1', 'wt13sc2', 'wt15sc1', 'wt15sc2', 'wt18sc1', 'wt18sc2',\n",
    "        \n",
    "        'ko11fb1', 'ko11fb2', 'ko13fb1', 'ko13fb2', 'ko15fb1', 'ko15fb2', 'ko18fb1', 'ko18fb2',\n",
    "        'ko11mb1', 'ko11mb2', 'ko13mb1', 'ko13mb2', 'ko15mb1', 'ko15mb2', 'ko18mb1', 'ko18mb2',\n",
    "        'ko11hb1', 'ko11hb2', 'ko13hb1', 'ko13hb2', 'ko15hb1', 'ko15hb2', 'ko18hb1', 'ko18hb2',\n",
    "        'ko11sc1', 'ko11sc2', 'ko13sc1', 'ko13sc2', 'ko15sc1', 'ko15sc2', 'ko18sc1', 'ko18sc2',\n",
    "]\n",
    "\n",
    "for c in no_na_merged:\n",
    "    if c not in cols:\n",
    "        cols.append(c)\n",
    "        \n",
    "df_all = no_na_merged[cols]\n",
    "df_all = df_all.reset_index()\n",
    "\n",
    "# Ensure the gene name hasn't been messed up \n",
    "gene_names = []\n",
    "pseudo_id = []\n",
    "\n",
    "for g in df_all[gene_id].values:\n",
    "    gene_n = gene_id_to_name.get(g)\n",
    "    gene_names.append(gene_n)\n",
    "    pseudo_id.append(f'{g}-{gene_n}')\n",
    "df_all[gene_name] = gene_names\n",
    "\n",
    "u.dp(['Length of dataframe (no dups reset index)', len(df_all)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Testing gender of samples\n",
    "\n",
    "Here we just confirm the gender of the samples based on X and Y reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_chr = df_all[df_all['chrs'] == 'Y']\n",
    "x_chr = df_all[df_all['chrs'] == 'X']\n",
    "\n",
    "\n",
    "wt_cols = [c for c in df_all.columns if 'merged' not in c and 'wt' in c]\n",
    "for w in wt_cols:\n",
    "    plt.hist(x_chr[w].values)\n",
    "    plt.title(f'WT X: {w}')\n",
    "    plt.show()\n",
    "    \n",
    "wt_cols = [c for c in df_all.columns if 'merged' not in c and 'wt' in c]\n",
    "for w in wt_cols:\n",
    "    plt.hist(y_chr[w].values)\n",
    "    plt.title(f'WT Y: {w}')\n",
    "    plt.show() \n",
    "\n",
    "    \n",
    "ko_cols = [c for c in df_all.columns if 'merged' not in c and 'ko' in c]\n",
    "for w in ko_cols:\n",
    "    plt.hist(x_chr[w].values)\n",
    "    plt.title(f'KO X: {w}')\n",
    "    plt.show()\n",
    "    \n",
    "for w in ko_cols:\n",
    "    plt.hist(y_chr[w].values)\n",
    "    plt.title(f'KO Y: {w}')\n",
    "    plt.show() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 9) Visualise the merged data top most significant genes\n",
    "\n",
    "Here we visualise the top genes by padj in the forebrain and validate that these align with the other results.\n",
    "\n",
    "We observe the Hox genes as the most diverged.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.colors import ListedColormap\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from sciviso import Scatterplot\n",
    "import umap\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "from sciviso import Vis\n",
    "\n",
    "# Plot PCA of the forebrain genes i.e. so we see how time separates\n",
    "def get_cond_colour(c):\n",
    "    if 'ko' in c:\n",
    "        return ko_colour\n",
    "    elif 'wt' in c:\n",
    "        return wt_colour\n",
    "    return '#FFFFFF'\n",
    "\n",
    "def plt_heatmap(df, idxs, title='', plt_sig=True, mark=None, row_colours=None, cond=''):\n",
    "    # Now let's look at a couple of these, and how similar these are to the other data\n",
    "    heatmap_df = pd.DataFrame()\n",
    "    # Select our genes of interest\n",
    "    if plt_sig:\n",
    "        ps = df['padj'].values[idxs]\n",
    "        p_idxs = np.where(ps < 0.05)\n",
    "        heatmap_df[gene_name] = df[gene_name].values[idxs][p_idxs]\n",
    "    else:\n",
    "        heatmap_df[gene_name] = df[gene_name].values[idxs]\n",
    "\n",
    "    cols = []\n",
    "    col_colours = []\n",
    "    cond_colours = []\n",
    "    tissue_colours = []\n",
    "    time_colours = []\n",
    "    for c in df.columns:\n",
    "        if mark == None:\n",
    "            if (cond in c) and 'merged' in c:\n",
    "                cname = c.split('_')[0]\n",
    "                print(c)\n",
    "                if plt_sig:\n",
    "                    heatmap_df[cname] = df[c].values[idxs][p_idxs]\n",
    "                else:\n",
    "                    heatmap_df[cname] = df[c].values[idxs]\n",
    "                cols.append(cname)\n",
    "                time_colours.append(get_time_colour(c))\n",
    "                tissue_colours.append(get_tissue_colour(c))\n",
    "                cond_colours.append(get_cond_colour(c))\n",
    "        else:\n",
    "            if mark in c and hist_metric in c:\n",
    "                if plt_sig:\n",
    "                    heatmap_df[c] = np.log2(df[c].values[idxs][p_idxs] + 1)\n",
    "                else:\n",
    "                    heatmap_df[c] =  np.log2(df[c].values[idxs] + 1)\n",
    "                    \n",
    "                time_colours.append(get_time_colour(c))\n",
    "                tissue_colours.append(get_tissue_colour(c))\n",
    "                cond_colours.append(get_mark_colour(c))\n",
    "                cols.append(c)\n",
    "    col_colours = [cond_colours, tissue_colours, time_colours]\n",
    "    \n",
    "    heatmap = Heatmap(heatmap_df, cols, gene_name, linewidths=0.0, cmap=\"Purples\", x_tick_labels=0, \n",
    "                      figsize=(2, 1.5), vmin=0, vmax=9,\n",
    "                      title=f'{title}', cluster_cols=False, cluster_rows=True, \n",
    "                     col_colours=col_colours, row_colours=row_colours)\n",
    "    \n",
    "    ax = heatmap.plot()\n",
    "            \n",
    "    ax.ax_heatmap.tick_params(direction='out', length=2, width=0.5)\n",
    "    ax.ax_heatmap.tick_params(labelsize=6)\n",
    "    ax.ax_heatmap.tick_params(axis='x', which='major', pad=2.0)\n",
    "    ax.ax_heatmap.tick_params(axis='y', which='major', pad=2.0)\n",
    "\n",
    "    pplot()\n",
    "    save_fig(title)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot PCA with each of the features "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols, edges = [], []\n",
    "e13, e13_colours = [], []\n",
    "e18, e18_colours = [], []\n",
    "e15, e15_colours = [], []\n",
    "tissue_colours = []\n",
    "time_colours = []\n",
    "cond_colours = []\n",
    "wt_tissue_colours = []\n",
    "wt_time_colours = []\n",
    "wt_cond_colours = []\n",
    "ko_tissue_colours = []\n",
    "ko_time_colours = []\n",
    "ko_cond_colours = []\n",
    "c_i = 0\n",
    "wt_pts = []\n",
    "ko_pts = []\n",
    "line_styles = []\n",
    "for c in df_all.columns:\n",
    "    if ('wt' in c or 'ko' in c) and '11' not in c:\n",
    "        cols.append(c)\n",
    "        if 'wt' in c:\n",
    "            wt_tissue_colours.append(get_tissue_colour(c))\n",
    "            wt_time_colours.append(get_time_colour(c))\n",
    "            wt_pts.append(c_i)\n",
    "        elif 'ko' in c:\n",
    "            ko_tissue_colours.append(get_tissue_colour(c))\n",
    "            ko_time_colours.append(get_time_colour(c))\n",
    "            ko_pts.append(c_i)\n",
    "        if '13' in c:\n",
    "            e13.append(c_i)\n",
    "            e13_colours.append(get_tissue_colour(c))\n",
    "            tissue_colours.append(get_tissue_colour(c))\n",
    "            time_colours.append(get_time_colour(c))\n",
    "            cond_colours.append(get_cond_colour(c))\n",
    "        elif '15' in c:\n",
    "            e15.append(c_i)\n",
    "            e15_colours.append(get_tissue_colour(c))\n",
    "            tissue_colours.append(get_tissue_colour(c))\n",
    "            time_colours.append(get_time_colour(c))\n",
    "            cond_colours.append(get_cond_colour(c))\n",
    "        elif '18' in c:\n",
    "            e18.append(c_i)\n",
    "            e18_colours.append(get_tissue_colour(c))\n",
    "            # Make the edges the colour of our condition\n",
    "            edges.append(get_cond_colour(c))\n",
    "            tissue_colours.append(get_tissue_colour(c))\n",
    "            time_colours.append(get_time_colour(c))\n",
    "            cond_colours.append(get_cond_colour(c))\n",
    "            if 'wt' in c:\n",
    "                line_styles.append('-')\n",
    "            else:\n",
    "                line_styles.append('-')\n",
    "        c_i += 1\n",
    "\n",
    "vals = (df_all[cols].values).T #np.log2(df_all[cols].values + 1).T\n",
    "fb_pca = PCA(n_components=2)\n",
    "fb_pca_values = fb_pca.fit_transform(vals)\n",
    "var_ratio = fb_pca.fit(vals).explained_variance_ratio_\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Plot PCA\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "plt.rcParams['figure.figsize'] = [2, 2]\n",
    "\n",
    "plt.scatter(fb_pca_values[e13,0], fb_pca_values[e13,1], linestyle=line_styles, c=e13_colours, s=100, marker=\">\", edgecolors=edges, linewidths=1.5)\n",
    "plt.scatter(fb_pca_values[e15,0], fb_pca_values[e15,1], linestyle=line_styles, c=e15_colours, s=100, marker=\"o\", edgecolors=edges, linewidths=1.5)\n",
    "plt.scatter(fb_pca_values[e18,0], fb_pca_values[e18,1], linestyle=line_styles, c=e18_colours, s=100, marker=\"X\", edgecolors=edges, linewidths=1.5)\n",
    "\n",
    "plt.title(f'PCA VAR: 0: {var_ratio[0]}, 1: {var_ratio[1]}')\n",
    "save_fig(f'PCA_ne11_fb')\n",
    "# Now we want to fit everything except the gene IDs \n",
    "plt.show()\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            PCA coloured by tissue\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "plt.rcParams['figure.figsize'] = [2, 2]\n",
    "\n",
    "plt.scatter(fb_pca_values[wt_pts,0], fb_pca_values[wt_pts,1], linestyle=line_styles, c=wt_tissue_colours, s=100, marker=\"o\", edgecolors='black', linewidths=0.5)\n",
    "plt.scatter(fb_pca_values[ko_pts,0], fb_pca_values[ko_pts,1], linestyle=line_styles, c=ko_tissue_colours, s=100, marker=\"X\", edgecolors='black', linewidths=0.5)\n",
    "\n",
    "plt.title(f'PCA VAR: 0: {var_ratio[0]}, 1: {var_ratio[1]}')\n",
    "save_fig(f'PCA_ne11_tissue')\n",
    "# Now we want to fit everything except the gene IDs \n",
    "plt.show()\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            PCA coloured by time\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "plt.rcParams['figure.figsize'] = [2, 2]\n",
    "\n",
    "plt.scatter(fb_pca_values[wt_pts,0], fb_pca_values[wt_pts,1], linestyle=line_styles, c=wt_time_colours, s=100, marker=\"o\", edgecolors='black', linewidths=0.5)\n",
    "plt.scatter(fb_pca_values[ko_pts,0], fb_pca_values[ko_pts,1], linestyle=line_styles, c=ko_time_colours, s=100, marker=\"X\", edgecolors='black', linewidths=0.5)\n",
    "\n",
    "plt.title(f'PCA VAR: 0: {var_ratio[0]}, 1: {var_ratio[1]}')\n",
    "save_fig(f'PCA_ne11_time')\n",
    "# Now we want to fit everything except the gene IDs \n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge samples and do sample clustermap\n",
    "\n",
    "Here we just do a sample clustermap "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Sample clustermap for anterior late stage\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "# Smooth out the columns in the data frame i.e. for the clones we only put in the mean of the two replicates\n",
    "log2_df = pd.DataFrame()\n",
    "cols_to_merge = [c for c in df_all.columns if 'wt' in c or 'ko' in c]\n",
    "col_names = []\n",
    "col_values = []\n",
    "i = 0\n",
    "tissue_colours = []\n",
    "cond_colours = []\n",
    "time_colours = []\n",
    "while(i < len(cols_to_merge)):\n",
    "    if ('mb' in cols_to_merge[i] or 'fb' in cols_to_merge[i]) and ('15' in cols_to_merge[i]  or '18' in cols_to_merge[i]):\n",
    "        log2_df[cols_to_merge[i][:-1]] = 0.5 * ((df_all[cols_to_merge[i]].values + 1) +\n",
    "                                                                    df_all[cols_to_merge[i + 1]].values)\n",
    "        tissue_colours.append(get_tissue_colour(cols_to_merge[i]))\n",
    "        time_colours.append(get_time_colour(cols_to_merge[i]))\n",
    "        cond_colours.append(get_cond_colour(cols_to_merge[i]))\n",
    "\n",
    "        print(\"merged\", cols_to_merge[i], cols_to_merge[i + 1])\n",
    "    i += 2\n",
    "    \n",
    "    \n",
    "plt.rcParams['figure.figsize'] = [3, 3]\n",
    "\n",
    "row_colors_t = [cond_colours, tissue_colours, time_colours]\n",
    "corr = log2_df.corr()\n",
    "sns.clustermap(corr, \n",
    "                    xticklabels=corr.columns.values,\n",
    "                    yticklabels=corr.columns.values, cmap='RdBu_r', row_cluster=True, \n",
    "                    col_cluster=True, row_colors=row_colors_t)\n",
    "save_fig(f'Heatmap_sample_cluster')\n",
    "plt.show()\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Sample clustermap for all\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "# Smooth out the columns in the data frame i.e. for the clones we only put in the mean of the two replicates\n",
    "log2_df = pd.DataFrame()\n",
    "cols_to_merge = [c for c in df_all.columns if 'wt' in c or 'ko' in c]\n",
    "col_names = []\n",
    "col_values = []\n",
    "i = 0\n",
    "tissue_colours = []\n",
    "cond_colours = []\n",
    "time_colours = []\n",
    "while(i < len(cols_to_merge)):\n",
    "    log2_df[cols_to_merge[i][:-1]] = 0.5 * ((df_all[cols_to_merge[i]].values + 1) +\n",
    "                                                                df_all[cols_to_merge[i + 1]].values)\n",
    "    tissue_colours.append(get_tissue_colour(cols_to_merge[i]))\n",
    "    time_colours.append(get_time_colour(cols_to_merge[i]))\n",
    "    cond_colours.append(get_cond_colour(cols_to_merge[i]))\n",
    "\n",
    "    print(\"merged\", cols_to_merge[i], cols_to_merge[i + 1])\n",
    "    i += 2\n",
    "    \n",
    "    \n",
    "plt.rcParams['figure.figsize'] = [3, 3]\n",
    "\n",
    "row_colors_t = [cond_colours, tissue_colours, time_colours]\n",
    "corr = log2_df.corr()\n",
    "sns.clustermap(corr, \n",
    "                    xticklabels=corr.columns.values,\n",
    "                    yticklabels=corr.columns.values, cmap='RdBu_r', row_cluster=True, \n",
    "                    col_cluster=True, row_colors=row_colors_t)\n",
    "save_fig(f'Heatmap_sample_cluster_all')\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Sample clustermap for E11\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "# Smooth out the columns in the data frame i.e. for the clones we only put in the mean of the two replicates\n",
    "log2_df = pd.DataFrame()\n",
    "cols_to_merge = [c for c in df_all.columns if 'wt' in c or 'ko' in c]\n",
    "col_names = []\n",
    "col_values = []\n",
    "i = 0\n",
    "tissue_colours = []\n",
    "cond_colours = []\n",
    "time_colours = []\n",
    "while(i < len(cols_to_merge)):\n",
    "    if '11' in cols_to_merge[i]:\n",
    "        log2_df[cols_to_merge[i][:-1]] = 0.5 * ((df_all[cols_to_merge[i]].values + 1) +\n",
    "                                                                    df_all[cols_to_merge[i + 1]].values)\n",
    "        tissue_colours.append(get_tissue_colour(cols_to_merge[i]))\n",
    "        time_colours.append(get_time_colour(cols_to_merge[i]))\n",
    "        cond_colours.append(get_cond_colour(cols_to_merge[i]))\n",
    "\n",
    "        print(\"merged\", cols_to_merge[i], cols_to_merge[i + 1])\n",
    "    i += 2\n",
    "    \n",
    "    \n",
    "plt.rcParams['figure.figsize'] = [3, 3]\n",
    "\n",
    "row_colors_t = [cond_colours, tissue_colours, time_colours]\n",
    "corr = log2_df.corr()\n",
    "sns.clustermap(corr, \n",
    "                    xticklabels=corr.columns.values,\n",
    "                    yticklabels=corr.columns.values, cmap='RdBu_r', row_cluster=True, \n",
    "                    col_cluster=True, row_colors=row_colors_t)\n",
    "save_fig(f'Heatmap_sample_cluster_Ell')\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Sample clustermap for E13\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "# Smooth out the columns in the data frame i.e. for the clones we only put in the mean of the two replicates\n",
    "log2_df = pd.DataFrame()\n",
    "cols_to_merge = [c for c in df_all.columns if 'wt' in c or 'ko' in c]\n",
    "col_names = []\n",
    "col_values = []\n",
    "i = 0\n",
    "tissue_colours = []\n",
    "cond_colours = []\n",
    "time_colours = []\n",
    "while(i < len(cols_to_merge)):\n",
    "    if '13' in cols_to_merge[i]:\n",
    "        log2_df[cols_to_merge[i][:-1]] = 0.5 * ((df_all[cols_to_merge[i]].values + 1) +\n",
    "                                                                    df_all[cols_to_merge[i + 1]].values)\n",
    "        tissue_colours.append(get_tissue_colour(cols_to_merge[i]))\n",
    "        time_colours.append(get_time_colour(cols_to_merge[i]))\n",
    "        cond_colours.append(get_cond_colour(cols_to_merge[i]))\n",
    "\n",
    "        print(\"merged\", cols_to_merge[i], cols_to_merge[i + 1])\n",
    "    i += 2\n",
    "    \n",
    "    \n",
    "plt.rcParams['figure.figsize'] = [3, 3]\n",
    "\n",
    "row_colors_t = [cond_colours, tissue_colours, time_colours]\n",
    "corr = log2_df.corr()\n",
    "sns.clustermap(corr, \n",
    "                    xticklabels=corr.columns.values,\n",
    "                    yticklabels=corr.columns.values, cmap='RdBu_r', row_cluster=True, \n",
    "                    col_cluster=True, row_colors=row_colors_t)\n",
    "save_fig(f'Heatmap_sample_cluster_El3')\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Sample clustermap for E15\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "# Smooth out the columns in the data frame i.e. for the clones we only put in the mean of the two replicates\n",
    "log2_df = pd.DataFrame()\n",
    "cols_to_merge = [c for c in df_all.columns if 'wt' in c or 'ko' in c]\n",
    "col_names = []\n",
    "col_values = []\n",
    "i = 0\n",
    "tissue_colours = []\n",
    "cond_colours = []\n",
    "time_colours = []\n",
    "while(i < len(cols_to_merge)):\n",
    "    if '15' in cols_to_merge[i]:\n",
    "        log2_df[cols_to_merge[i][:-1]] = 0.5 * ((df_all[cols_to_merge[i]].values + 1) +\n",
    "                                                                    df_all[cols_to_merge[i + 1]].values)\n",
    "        tissue_colours.append(get_tissue_colour(cols_to_merge[i]))\n",
    "        time_colours.append(get_time_colour(cols_to_merge[i]))\n",
    "        cond_colours.append(get_cond_colour(cols_to_merge[i]))\n",
    "\n",
    "        print(\"merged\", cols_to_merge[i], cols_to_merge[i + 1])\n",
    "    i += 2\n",
    "    \n",
    "    \n",
    "plt.rcParams['figure.figsize'] = [3, 3]\n",
    "\n",
    "row_colors_t = [cond_colours, tissue_colours, time_colours]\n",
    "corr = log2_df.corr()\n",
    "sns.clustermap(corr, \n",
    "                    xticklabels=corr.columns.values,\n",
    "                    yticklabels=corr.columns.values, cmap='RdBu_r', row_cluster=True, \n",
    "                    col_cluster=True, row_colors=row_colors_t)\n",
    "save_fig(f'Heatmap_sample_cluster_El5')\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Sample clustermap for E18\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "# Smooth out the columns in the data frame i.e. for the clones we only put in the mean of the two replicates\n",
    "log2_df = pd.DataFrame()\n",
    "cols_to_merge = [c for c in df_all.columns if 'wt' in c or 'ko' in c]\n",
    "col_names = []\n",
    "col_values = []\n",
    "i = 0\n",
    "tissue_colours = []\n",
    "cond_colours = []\n",
    "time_colours = []\n",
    "while(i < len(cols_to_merge)):\n",
    "    if '15' in cols_to_merge[i]:\n",
    "        log2_df[cols_to_merge[i][:-1]] = 0.5 * ((df_all[cols_to_merge[i]].values + 1) +\n",
    "                                                                    df_all[cols_to_merge[i + 1]].values)\n",
    "        tissue_colours.append(get_tissue_colour(cols_to_merge[i]))\n",
    "        time_colours.append(get_time_colour(cols_to_merge[i]))\n",
    "        cond_colours.append(get_cond_colour(cols_to_merge[i]))\n",
    "\n",
    "        print(\"merged\", cols_to_merge[i], cols_to_merge[i + 1])\n",
    "    i += 2\n",
    "    \n",
    "plt.rcParams['figure.figsize'] = [3, 3]\n",
    "\n",
    "row_colors_t = [cond_colours, tissue_colours, time_colours]\n",
    "corr = log2_df.corr()\n",
    "sns.clustermap(corr, \n",
    "                    xticklabels=corr.columns.values,\n",
    "                    yticklabels=corr.columns.values, cmap='RdBu_r', row_cluster=True, \n",
    "                    col_cluster=True, row_colors=row_colors_t)\n",
    "save_fig(f'Heatmap_sample_cluster_El8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot heatmaps of samples RNAseq\n",
    "\n",
    "Plot the top and bottom of the heatmaps of each of the DE analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Select top 20 DE genes (by padj in Forebrain)\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "n_genes = 10\n",
    "n_g_neg = -10\n",
    "# Select the locfc that is the forebrain WT vs KO\n",
    "merged_df = df_all.copy()\n",
    "\n",
    "cols_to_merge = [c for c in merged_df.columns if 'wt' in c or 'ko' in c]\n",
    "col_names = []\n",
    "col_values = []\n",
    "values = merged_df[cols_to_merge].values\n",
    "i = 0\n",
    "while(i < len(cols_to_merge)):\n",
    "    merged_df[f'{cols_to_merge[i][:-1]}_merged-rep'] = 0.5 * (merged_df[cols_to_merge[i]].values +\n",
    "                                                                merged_df[cols_to_merge[i + 1]].values)\n",
    "    print(\"merged\", cols_to_merge[i], cols_to_merge[i + 1])\n",
    "    i += 2\n",
    "\n",
    "log2FoldChange = merged_df['stat'].values\n",
    "selected_idxs = list(np.argpartition(log2FoldChange, n_genes)[:n_genes]) # np.argpartition(x, -k)[-k:]\n",
    "\n",
    "col_colours = ['#6666ff'] * n_genes + ['#990033'] * n_genes\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Plot heatmap\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "plt_heatmap(merged_df, selected_idxs, f'Top genes', False, row_colours=None, cond=\"\")\n",
    "plt.show()\n",
    "\n",
    "selected_idxs = list(np.argpartition(log2FoldChange, n_g_neg)[n_g_neg:])\n",
    "plt_heatmap(merged_df, selected_idxs, f'Bottom genes', False, row_colours=None, cond=\"\")\n",
    "plt.show()\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Setup PCA\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "cols, edges = [], []\n",
    "e13, e13_colours = [], []\n",
    "e18, e18_colours = [], []\n",
    "e15, e15_colours = [], []\n",
    "c_i = 0\n",
    "line_styles = []\n",
    "for c in df_all.columns:\n",
    "    if ('wt' in c or 'ko' in c) and '11' not in c:\n",
    "        cols.append(c)\n",
    "        if '13' in c:\n",
    "            e13.append(c_i)\n",
    "            e13_colours.append(get_tissue_colour(c))\n",
    "        elif '15' in c:\n",
    "            e15.append(c_i)\n",
    "            e15_colours.append(get_tissue_colour(c))\n",
    "        elif '18' in c:\n",
    "            e18.append(c_i)\n",
    "            e18_colours.append(get_tissue_colour(c))\n",
    "            # Make the edges the colour of our condition\n",
    "            edges.append(get_cond_colour(c))\n",
    "            if 'wt' in c:\n",
    "                line_styles.append('-')\n",
    "            else:\n",
    "                line_styles.append('-')\n",
    "        c_i += 1\n",
    "\n",
    "vals = np.log2(df_all[cols].values + 1).T\n",
    "fb_pca = PCA(n_components=2)\n",
    "fb_pca_values = fb_pca.fit_transform(vals)\n",
    "var_ratio = fb_pca.fit(vals).explained_variance_ratio_\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Plot PCA\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "plt.rcParams['figure.figsize'] = [2, 2]\n",
    "\n",
    "plt.scatter(fb_pca_values[e13,0], fb_pca_values[e13,1], linestyle=line_styles, c=e13_colours, s=100, marker=\">\", edgecolors=edges, linewidths=1.5)\n",
    "plt.scatter(fb_pca_values[e15,0], fb_pca_values[e15,1], linestyle=line_styles, c=e15_colours, s=100, marker=\"o\", edgecolors=edges, linewidths=1.5)\n",
    "plt.scatter(fb_pca_values[e18,0], fb_pca_values[e18,1], linestyle=line_styles, c=e18_colours, s=100, marker=\"X\", edgecolors=edges, linewidths=1.5)\n",
    "\n",
    "plt.title(f'PCA VAR: 0: {var_ratio[0]}, 1: {var_ratio[1]}')\n",
    "save_fig(f'PCA_ne11_fb')\n",
    "# Now we want to fit everything except the gene IDs \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Select top 20 DE genes (by padj in Forebrain)\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "n_genes = 10\n",
    "n_g_neg = -10\n",
    "val = 'stat'\n",
    "# Select the locfc that is the forebrain WT vs KO\n",
    "for cond in ['a11', 'a13', 'a15', 'a18', 'p11', 'p13', 'p15', 'p18']:\n",
    "    e18_gene_df = merged_df.copy()\n",
    "\n",
    "    log2FoldChange = e18_gene_df[f'{val}_{cond}'].values\n",
    "    selected_idxs = list(np.argpartition(log2FoldChange, n_genes)[:n_genes]) # np.argpartition(x, -k)[-k:]\n",
    "\n",
    "\n",
    "    col_colours = ['#6666ff'] * n_genes + ['#990033'] * n_genes\n",
    "\n",
    "    \"\"\"\n",
    "    ---------------------------------------------------------------\n",
    "                Plot heatmap\n",
    "    ---------------------------------------------------------------\n",
    "    \"\"\"\n",
    "\n",
    "    plt_heatmap(merged_df, selected_idxs, f'Downregulated genes {cond}', False, row_colours=None)\n",
    "    plt.show()\n",
    "\n",
    "    log2FoldChange = e18_gene_df[f'log2FoldChange_{cond}'].values\n",
    "    selected_idxs = list(np.argpartition(log2FoldChange, n_g_neg)[n_g_neg:])\n",
    "    plt_heatmap(merged_df, selected_idxs, f'Upregulated genes {cond}', False, row_colours=None)\n",
    "    plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Select top 20 DE genes (by padj in Forebrain)\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "n_genes = 10\n",
    "n_g_neg = -10\n",
    "# Select the locfc that is the forebrain WT vs KO\n",
    "e18_gene_df = merged_df.copy()\n",
    "\n",
    "log2FoldChange = e18_gene_df['stat_p18'].values\n",
    "selected_idxs = list(np.argpartition(log2FoldChange, n_genes)[:n_genes]) # np.argpartition(x, -k)[-k:]\n",
    "\n",
    "\n",
    "col_colours = ['#6666ff'] * n_genes + ['#990033'] * n_genes\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Plot heatmap\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "plt_heatmap(merged_df, selected_idxs, f'Downregulated genes HB-SC e18', False, row_colours=None)\n",
    "plt.show()\n",
    "\n",
    "log2FoldChange = e18_gene_df['log2FoldChange_p18'].values\n",
    "selected_idxs = list(np.argpartition(log2FoldChange, n_g_neg)[n_g_neg:])\n",
    "plt_heatmap(merged_df, selected_idxs, f'Upregulated genes HB-SC e18', False, row_colours=None)\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Select top 20 DE genes (by padj in Forebrain)\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "n_genes = 10\n",
    "n_g_neg = -10\n",
    "# Select the locfc that is the forebrain WT vs KO\n",
    "e18_gene_df = merged_df.copy()\n",
    "\n",
    "log2FoldChange = e18_gene_df['log2FoldChange_sc'].values\n",
    "selected_idxs = list(np.argpartition(log2FoldChange, n_genes)[:n_genes]) # np.argpartition(x, -k)[-k:]\n",
    "\n",
    "\n",
    "col_colours = ['#6666ff'] * n_genes + ['#990033'] * n_genes\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Plot heatmap\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "plt_heatmap(merged_df, selected_idxs, f'Downregulated genes SC', False, row_colours=None)\n",
    "plt.show()\n",
    "\n",
    "log2FoldChange = e18_gene_df['log2FoldChange_sc'].values\n",
    "selected_idxs = list(np.argpartition(log2FoldChange, n_g_neg)[n_g_neg:])\n",
    "plt_heatmap(merged_df, selected_idxs, f'Upregulaed genes SC', False, row_colours=None)\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Select top 20 DE genes (by padj in Forebrain)\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "n_genes = 10\n",
    "n_g_neg = -10\n",
    "# Select the locfc that is the forebrain WT vs KO\n",
    "e18_gene_df = merged_df.copy()\n",
    "\n",
    "log2FoldChange = e18_gene_df['log2FoldChange_hb'].values\n",
    "selected_idxs = list(np.argpartition(log2FoldChange, n_genes)[:n_genes]) # np.argpartition(x, -k)[-k:]\n",
    "\n",
    "\n",
    "col_colours = ['#6666ff'] * n_genes + ['#990033'] * n_genes\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Plot heatmap\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "plt_heatmap(merged_df, selected_idxs, f'Downregulated genes HB', False, row_colours=None)\n",
    "plt.show()\n",
    "\n",
    "log2FoldChange = e18_gene_df['log2FoldChange_hb'].values\n",
    "selected_idxs = list(np.argpartition(log2FoldChange, n_g_neg)[n_g_neg:])\n",
    "plt_heatmap(merged_df, selected_idxs, f'Upregulaed genes HB', False, row_colours=None)\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Select top 20 DE genes (by padj in Forebrain)\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "n_genes = 10\n",
    "n_g_neg = -10\n",
    "# Select the locfc that is the forebrain WT vs KO\n",
    "e18_gene_df = merged_df.copy()\n",
    "\n",
    "log2FoldChange = e18_gene_df['log2FoldChange_mb'].values\n",
    "selected_idxs = list(np.argpartition(log2FoldChange, n_genes)[:n_genes]) # np.argpartition(x, -k)[-k:]\n",
    "\n",
    "\n",
    "col_colours = ['#6666ff'] * n_genes + ['#990033'] * n_genes\n",
    "\n",
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Plot heatmap\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "plt_heatmap(merged_df, selected_idxs, f'Downregulated genes MB', False, row_colours=None)\n",
    "plt.show()\n",
    "\n",
    "log2FoldChange = e18_gene_df['log2FoldChange_mb'].values\n",
    "selected_idxs = list(np.argpartition(log2FoldChange, n_g_neg)[n_g_neg:])\n",
    "plt_heatmap(merged_df, selected_idxs, f'Upregulaed genes MB', False, row_colours=None)\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Heatmaps of marker genes in WT and KO\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "gene_markers_sep = [['Foxg1', 'Tbr1', 'Emx1', 'Eomes'], \n",
    "                    ['Bhlhe23', 'En2', 'Fev', 'Phox2b'], \n",
    "                    ['Hoxb9', 'Hoxc9', 'Hoxd9', 'Hoxc12'],\n",
    "                    ['Hoxd8', 'Hoxd9', 'Hoxd10', 'Hoxd11', 'Hoxd12', 'Hoxd13', 'Hoxa7', 'Hoxa9', 'Hoxa10', 'Hoxa11', \n",
    "                    'Hoxa13', 'Hoxb9', 'Hoxb13',  'Hoxc8', 'Hoxc9', 'Hoxc10', 'Hoxc11', 'Hoxc12', 'Hoxc13'],\n",
    "                    ['Ccna1', 'Ccna2', 'Ccnd1', 'Ccnd2', 'Ccnd3', 'Ccne1', 'Ccne2', 'Cdc25a', \n",
    "                    'Cdc25b', 'Cdc25c', 'E2f1', 'E2f2', 'E2f3', 'Mcm10', 'Mcm5', 'Mcm3', 'Mcm2', 'Cip2a'],\n",
    "                    ['Cdkn1a', 'Cdkn1b', 'Cdkn1c', 'Cdkn2a', 'Cdkn2b', 'Cdkn2c', 'Cdkn2d'],\n",
    "                    ['Sox1', 'Sox2', 'Sox3'],\n",
    "                    ['Snap25', 'Syt1', 'Slc32a1','Slc17a6', 'Syn1'],\n",
    "                    ['Cspg4', 'Aqp4', 'Slc6a11', 'Olig1', 'Igfbp3']\n",
    "                    ]\n",
    "\n",
    "marker_labels_sep = ['Forebrain', 'Midbrain', 'Hindbrain',  'Spinalcord', \n",
    "                     'Proliferation', 'Neg. reg. Cell Cycle', \n",
    "                    'Progenitors', 'Neurons', 'Glia'\n",
    "                    ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for i, genes in enumerate(gene_markers_sep):\n",
    "    selected_idxs = []\n",
    "    for j, gene in enumerate(df_all[gene_name].values):\n",
    "        if gene in genes:\n",
    "            selected_idxs.append(j)\n",
    "    plt_heatmap(merged_df, selected_idxs, marker_labels_sep[i] + \" genes WT\", False, row_colours=None, cond='wt')\n",
    "    plt_heatmap(merged_df, selected_idxs, marker_labels_sep[i] + \" genes KO\", False, row_colours=None, cond='ko')\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 10) Add Epigenetic data from Encode\n",
    "We downloaded and processed the data according to the notebook DownloadEncode. Now we want to merge the peak files with \n",
    "our RNAseq data. \n",
    "\n",
    "Here we keep the signal with the highest value as multiple may have been added. \n",
    "\n",
    "Keeping in mind, we have potentially the same peak assigned to many genes. This is particularly evident in the H3K27me3 \n",
    "broad peak data.\n",
    "\n",
    "First we download an annotation from \n",
    "We use sciloc2gene v1.0.0 to annotate this information.\n",
    "\n",
    "Note you may need to check the reference used for the peak files (this was not needed since we also used Ensembl to map our gene's and find their location).\n",
    "\n",
    "Need to download the annotation from: https://www.encodeproject.org/references/ENCSR425FOI/\n",
    "ENCFF871VGROpen file informationDownload\tgtf\tgenome reference\tmm10\tM21\n",
    "\n",
    "Convert the GTF to a bed and then sort the bed file (convert using gtf2bed v0.11.0 downloaded frmo https://github.com/fls-bioinformatics-core/GFFUtils/tree/master)\n",
    "gtf2bed ENCFF871VGR.gtf > ENCFF871VGR_mm10-m21_encode.bed\n",
    "\n",
    "Then we convert to a CSV file and sort it by TSS.\n",
    "```\n",
    "def run_parallel(self):\n",
    "    data_dir = 'data/encode/histone_modifications/beds_cns/encode/sorted_bed/'\n",
    "    files = os.listdir(data_dir)\n",
    "    files_to_run = []\n",
    "    for filename in files:\n",
    "        files_to_run.append(filename)\n",
    "    # Run in paralell\n",
    "    pool = ThreadPool(12)\n",
    "    results = pool.map(run_bed, files_to_run)\n",
    "\n",
    "\n",
    "def run_bed(filename):\n",
    "    base_dir = 'data/encode/histone_modifications/beds_cns/encode/'\n",
    "    data_dir = f'{base_dir}sorted_bed/'\n",
    "    output_dir = f'{base_dir}scie2g_28102020/'\n",
    "    overlap_method = 'in_promoter'\n",
    "    if 'H3K9me3' in filename or 'H3K36me3' in filename:\n",
    "        overlap_method = 'overlaps'\n",
    "    #ensembl_gene_id, entrezgene_id, external_gene_name, chromosome_name, start_position, end_position, strand\n",
    "    bed = Bed(f'{data_dir}{filename}', overlap_method=overlap_method,\n",
    "              output_bed_file=f'{output_dir}selected_peaks/{filename}',\n",
    "              buffer_after_tss=1500,\n",
    "              buffer_before_tss=5000,\n",
    "              buffer_gene_overlap=1500,\n",
    "              chr_idx=0, start_idx=1,\n",
    "              end_idx=2, peak_value=6, header_extra='8,9',\n",
    "              header=\"chr,start,end,signal,qvalue,peak\",\n",
    "              gene_start=4, gene_end=5, gene_direction=6, gene_chr=3, gene_name=0\n",
    "              )\n",
    "    # Use your annotation file i.e. for humans\n",
    "    bed.set_annotation_from_file('../supps/mm10Sorted_mmusculus_gene_ensembl-GRCm38.p6.csv')\n",
    "    # Now we can run the assign values\n",
    "    bed.assign_locations_to_genes()\n",
    "    bed.save_loc_to_csv(f'{output_dir}parsed_files/{filename[:-3]}csv', keep_unassigned=True)\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "--------------------------------------------------------\n",
    "Add Epigenetic data to the RNAseq dataframe\n",
    "--------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "def add_peak_file(df, filename):\n",
    "    # Make a dictionary from the key being the gene id to the value (width)\n",
    "    mapping = {}\n",
    "    mapping_signals = {}\n",
    "\n",
    "    file_df = pd.read_csv(f'{chip_dir}{filename}')\n",
    "    \n",
    "    # Keep the max value\n",
    "    signals = file_df['signal'].values\n",
    "    widths = file_df['width'].values\n",
    "    qvalues = file_df['qvalue'].values # q-value at position 9 i.e. 9 - 1 = 8\n",
    "    \n",
    "    # Only keep peaks that were significant i.e. qvalue > -1 * log10(0.05) ~ 1.3\n",
    "    for i, g in enumerate(file_df[gene_id].values):\n",
    "        qvalue = qvalues[i]\n",
    "        if qvalue > 1.3:\n",
    "            if not mapping.get(g):\n",
    "                mapping[g] = widths[i]\n",
    "                mapping_signals[g] = signals[i]\n",
    "            else:\n",
    "                mapping[g] = widths[i] if widths[i] > mapping[g] else mapping[g]\n",
    "                # Keep the signal that corresponds to the widest peak\n",
    "                mapping_signals[g] = signals[i] if widths[i] > mapping[g] else mapping_signals[g]\n",
    "    # Now we want to iterate through the genes in the df and add them\n",
    "    widths_ordered = []\n",
    "    signals_ordered = []\n",
    "    fname = filename[:-4]\n",
    "    for g in df[gene_id].values:\n",
    "        widths_ordered.append(mapping.get(g))\n",
    "        signals_ordered.append(mapping_signals.get(g))\n",
    "        \n",
    "    df[f'{fname}_width'] = widths_ordered\n",
    "    df[f'{fname}_signal'] = signals_ordered\n",
    "\n",
    "    return df\n",
    "\n",
    "# Now we want to add our chipseq data from Encode\n",
    "chip_dir = os.path.join(input_dir, f'scie2g_b2500_g500/parsed_files/')\n",
    "files = os.listdir(chip_dir)\n",
    "\n",
    "chip_files = []\n",
    "mark_locations = []\n",
    "for f in files:\n",
    "    if '.csv' in f and '.swp' not in f and 'embryonic_' in f:\n",
    "        chip_files.append(f)\n",
    "\n",
    "chip_files.sort()\n",
    "\n",
    "# Add them to the dataframe\n",
    "dfs = []\n",
    "for filename in chip_files:\n",
    "    u.dp([\"Adding\", filename])\n",
    "    df_all = add_peak_file(df_all, filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 11) Filter the merged dataframe\n",
    "\n",
    "Since we have merged so many dataframes there are likely to be many missing values.\n",
    "\n",
    "We want to first drop any rows that are duplicates & 2) fill in null values with 0's (otherwise we'll get errors \n",
    "when running analyses).\n",
    "\n",
    "We also want to organise the columns of the dataframe (this makes it easier for charts later on).\n",
    "\n",
    "Lastly, we are only interested in genes that had a significant change between WT and KO (we want to desipher WHY this occured), so we drop any data with a non-sig result.\n",
    "\n",
    "Filtering steps:\n",
    "\n",
    "    1) Keep genes if padj < 0.05 in at least one experiment, THEN\n",
    "    2) Keep genes if genes were expressed in WT or KO tissues with at least an average TMM of 2, THEN\n",
    "    3) Keep genes if genes at least one experiment recorded a log fold change of > 1.0\n",
    "  \n",
    "This dataset will be used for our VAE analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# Keep only the significant data i.e. the data where at least one of the p values was < 0.05\n",
    "genes_to_keep = []\n",
    "ps = ((1.0 * (df_all['padj_sc'].values < 0.05)) + (1.0 * (df_all['padj_mb'].values < 0.05))\n",
    "     +  (1.0 * (df_all['padj_hb'].values < 0.05)) + (1.0 * (df_all['padj_fb'].values < 0.05))\n",
    "     + (1.0 * (df_all['padj_a11'].values < 0.05)) \n",
    "     + (1.0 * (df_all['padj_a13'].values < 0.05)) \n",
    "     + (1.0 * (df_all['padj_a15'].values < 0.05))\n",
    "     + (1.0 * (df_all['padj_a18'].values < 0.05)))\n",
    "\n",
    "\n",
    "df_all['p_mask'] = ps\n",
    "\n",
    "# Now we also want to make sure that in each of these datasets, we have at least 2 TMM average in either WT or KO\n",
    "# We don't want the genes that have super low expression since that is not interesting for our downstream analyses\n",
    "wt_cols = [c for c in df_all.columns if 'wt' in c]\n",
    "ko_cols = [c for c in df_all.columns if 'ko' in c]\n",
    "tmm_cutoff = 0.5\n",
    "\n",
    "expression_mask = ((1.0 * (np.mean(df_all[wt_cols].values, axis=1) >= tmm_cutoff)) + (1.0 * (np.mean(df_all[ko_cols].values, axis=1) >= tmm_cutoff)))\n",
    "df_all['expression_mask'] = expression_mask\n",
    "\n",
    "# Make a DF that only has genes that are significant at the 0.05 level\n",
    "df_sig = df_all[df_all['p_mask'] > 0]\n",
    "\n",
    "# Make a DF that only has genes that are minimally expressed in at least one condition\n",
    "df_sig = df_sig[df_sig['expression_mask'] > 0]\n",
    "\n",
    "u.dp(['Number of significant rows: ', len(df_sig)])\n",
    "\n",
    "# Let's also make sure that we can keep have genes that have at least an absolute logfc of 0.5\n",
    "log_str = 'log2FoldChange'\n",
    "logfc_cols = [c for c in df_sig.columns if log_str in c]\n",
    "mean_logfc = np.mean(abs(df_sig[logfc_cols].values), axis=1)\n",
    "\n",
    "\n",
    "cutoff = 1.0\n",
    "logfc = ((1.0 * (abs(df_sig[f'{log_str}_sc'].values) > cutoff)) \n",
    "         + (1.0 * (abs(df_sig[f'{log_str}_mb'].values) > cutoff)) \n",
    "         + (1.0 * (abs(df_sig[f'{log_str}_hb'].values) > cutoff)) \n",
    "         + (1.0 * (abs(df_sig[f'{log_str}_fb'].values) > cutoff)) \n",
    "         + (1.0 * (abs(df_sig[f'{log_str}_a11'].values) > cutoff))\n",
    "         + (1.0 * (abs(df_sig[f'{log_str}_a13'].values) > cutoff)) \n",
    "         + (1.0 * (abs(df_sig[f'{log_str}_a15'].values) > cutoff)) \n",
    "         + (1.0 * (abs(df_sig[f'{log_str}_a18'].values) > cutoff))\n",
    "        )\n",
    "\n",
    "df_sig['logfc_mask'] = logfc\n",
    "df_sig_2 = df_sig[df_sig['logfc_mask'] > 2]  # Require two of the conditions it to be met\n",
    "df_sig_1 = df_sig[df_sig['logfc_mask'] > 0]  # Require two of the conditions it to be met\n",
    "\n",
    "u.dp(['Number of significant rows with an absolute FC > 1.0: ', len(df_sig_1)])\n",
    "u.dp(['Number of significant rows with 2 values absolute FC > 2.0: ', len(df_sig_2)])\n",
    "\n",
    "# Save all to dataframes\n",
    "df_sig_1.to_csv(f'{output_dir}df-sig_1_epi-2500_{date}.csv', index=False)\n",
    "df_sig_2.to_csv(f'{output_dir}df-consistent_epi-2500_{date}.csv', index=False)\n",
    "df_sig.to_csv(f'{output_dir}df-significant_epi-2500_{date}.csv', index=False)\n",
    "df_all.to_csv(f'{output_dir}df-all_epi-2500_{date}.csv', index=False)\n",
    "\n",
    "# --------------------------------------------------------------------------------\n",
    "#                       Number of significant rows: \t12532\t                       \n",
    "# --------------------------------------------------------------------------------\n",
    "# --------------------------------------------------------------------------------\n",
    "#           Number of significant rows with an absolute FC > 1.0: \t2944\t          \n",
    "# --------------------------------------------------------------------------------\n",
    "# --------------------------------------------------------------------------------\n",
    "#        Number of significant rows with 2 values absolute FC > 2.0: \t1928\t       \n",
    "# --------------------------------------------------------------------------------\n",
    "\n",
    "gene_markers_sep = [['Emx1', 'Eomes', 'Tbr1', 'Foxg1', 'Lhx6'], \n",
    "                    ['En1', 'En2', 'Lmx1a', 'Bhlhe23', 'Sall4'], \n",
    "                    ['Hoxb1', 'Krox20', 'Fev', 'Hoxd3', 'Phox2b'],\n",
    "                    ['Hoxd8', 'Hoxd9', 'Hoxd10', 'Hoxd11', 'Hoxd12', 'Hoxd13', 'Hoxa7', 'Hoxa9', 'Hoxa10', 'Hoxa11', \n",
    "                    'Hoxa13', 'Hoxb9', 'Hoxb13',  'Hoxc8', 'Hoxc9', 'Hoxc10', 'Hoxc11', 'Hoxc12', 'Hoxc13'],\n",
    "                    ['Ccna1', 'Ccna2', 'Ccnd1', 'Ccnd2', 'Ccnd3', 'Ccne1', 'Ccne2', 'Cdc25a', \n",
    "                    'Cdc25b', 'Cdc25c', 'E2f1', 'E2f2', 'E2f3', 'Mcm10', 'Mcm5', 'Mcm3', 'Mcm2', 'Cip2a'],\n",
    "                    ['Cdkn1a', 'Cdkn1b', 'Cdkn1c', 'Cdkn2a', 'Cdkn2b', 'Cdkn2c', 'Cdkn2d'],\n",
    "                    ['Sox2', 'Sox1', 'Sox3', 'Hes1', 'Hes5'],\n",
    "                    ['Snap25', 'Syt1', 'Slc32a1','Slc17a6', 'Syn1'],\n",
    "                    ['Cspg4', 'Aqp4', 'Slc6a11', 'Olig1', 'Igfbp3'],\n",
    "                    ['Foxg1'], \n",
    "                    ['En2'], \n",
    "                    ['Phox2b'],\n",
    "                    ['Hoxc9'],\n",
    "                    ['Sox3']\n",
    "                    ]\n",
    "for ml in gene_markers_sep:\n",
    "    print(\"\\n\")\n",
    "    for m in ml:\n",
    "        for g in df_sig_1[gene_name].values:\n",
    "            try:\n",
    "                if g == m:\n",
    "                    print(g)\n",
    "            except:\n",
    "                x = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 12) Setup functions for simple analyses\n",
    "\n",
    "\n",
    "We want to test whether some simple things are enriched or not. i.e. do we see different annotations if we look at the genes that were 1) significant a) with H3K27me3, b) without H3K27me3, 2) non-significant a) with H3K27me3, b) without H3K27me3.\n",
    "\n",
    "For each of these groups we want to:\n",
    "\n",
    "    1) test for annotation enrichment\n",
    "    2) plot the overal HM profile \n",
    "    3) plot the overal gene expression profile."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import venn\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "def plot_venn(gene_sets, labels, title, save=True, show=True, colours=None):\n",
    "    output = f'venn_{title.replace(\" \", \"-\")}'\n",
    "    dset = {}\n",
    "    for i, l in enumerate(labels):\n",
    "        dset[l] = gene_sets[i]\n",
    "    if len(gene_sets) > 3:\n",
    "        cmap = ListedColormap(sns.color_palette(colours))\n",
    "        venn.venn(dset, cmap=cmap, fontsize=6, figsize=(2, 2))\n",
    "    else:\n",
    "        venn3(gene_sets, set_labels=labels)\n",
    "    if save:\n",
    "        save_fig(output)\n",
    "    if show:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats.multitest import multipletests\n",
    "import string\n",
    "from scipy import stats\n",
    "from sciviso import Heatmap\n",
    "import matplotlib \n",
    "def run_annot_plot(df_bg, df_fg, title=\"\"):\n",
    "    changes = []\n",
    "    order = ['Pr-A', 'Pr-W', 'Pr-B', 'Pr-F', 'En-Sd', 'En-Sp', \n",
    "         'En-W', 'En-Pd', 'En-Pp', 'Tr-S', 'Tr-P', 'Tr-I', \n",
    "          'Hc-P', 'Hc-H', 'NS']\n",
    "    total = len(df_bg) * 1.0\n",
    "    total_sig = len(df_fg) * 1.0\n",
    "    perc_all = []\n",
    "    perc_sig = []\n",
    "    \n",
    "    for o in order:\n",
    "        o_all = len(df_bg[df_bg['peak_value'] == o])\n",
    "        o_sig = len(df_fg[df_fg['peak_value'] == o])\n",
    "        perc_all.append(o_all)\n",
    "        perc_sig.append(o_sig)\n",
    "        if o_all == 0 and o_sig == 0:\n",
    "            changes.append(0)\n",
    "        else:\n",
    "            changes.append(((o_sig) - (o_all)))\n",
    "\n",
    "    odds_ratios = []\n",
    "    pvalues = []\n",
    "    for i, o in enumerate(order):\n",
    "        # Do a FET on each one\n",
    "        oddsratio, pvalue = stats.fisher_exact([[perc_sig[i], perc_all[i]], \n",
    "                                                [total_sig - perc_sig[i],\n",
    "                                                 total - perc_all[i]]])\n",
    "\n",
    "        print(o, oddsratio, pvalue,[[perc_sig[i], perc_all[i]], \n",
    "                                                [total_sig - perc_sig[i], \n",
    "                                                 total - perc_all[i]]])\n",
    "        odds_ratios.append(oddsratio)\n",
    "        pvalues.append(pvalue)\n",
    "\n",
    "    reg, padj, a, b = multipletests(pvalues, alpha=0.1, \n",
    "                                    method='fdr_bh', returnsorted=False)\n",
    "\n",
    "    p_sigs = []\n",
    "    for p in padj: \n",
    "        if p > 0.05:\n",
    "            p_sigs.append('')\n",
    "        elif p <= 0.05 and p > 0.01:\n",
    "            p_sigs.append('*')\n",
    "        elif p <= 0.01 and p > 0.001:\n",
    "            p_sigs.append('**')\n",
    "        elif p <= 0.001 and p > 0.0001:\n",
    "            p_sigs.append('***')\n",
    "        elif p <= 0.0001:\n",
    "            p_sigs.append('****')\n",
    "        else:\n",
    "            print(p)\n",
    "            p_sigs.append('')\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(1.5,1.0))\n",
    "    c = [grey] * 14\n",
    "    c[2] = \"green\"\n",
    "    c[12] = \"green\"\n",
    "    plt.bar(order, odds_ratios, color=c, linewidth=0.5, edgecolor='black')\n",
    "    rects = ax.patches\n",
    "    # Make some labels.\n",
    "    labels = [f'{p_sigs[i]}' for i in range(0, len(rects))]\n",
    "\n",
    "    for rect, label in zip(rects, labels):\n",
    "        height = rect.get_height()\n",
    "        ax.text(rect.get_x() + rect.get_width() / 2, height, label,\n",
    "                ha='center', va='bottom')\n",
    "    plt.ylim(0, 7)\n",
    "    plt.yticks(np.arange(0, 7, 2.0))\n",
    "    ax.set_title(f'Odds ratio FET for significant dataset {title} compared to all genes ChromHMM annotations')\n",
    "    ax.set_xlabel('', fontsize=6)\n",
    "    ax.set_ylabel('Odds ratio', fontsize=6)\n",
    "    ax.set_xticklabels(order, rotation=90, ha=\"center\")\n",
    "\n",
    "        \n",
    "    ax.tick_params(direction='out', length=2, width=0.5)\n",
    "    ax.spines['bottom'].set_linewidth(0.5)\n",
    "    ax.spines['top'].set_linewidth(0)\n",
    "    ax.spines['left'].set_linewidth(0.5)\n",
    "    ax.spines['right'].set_linewidth(0)\n",
    "    ax.tick_params(labelsize=6)\n",
    "    ax.tick_params(axis='x', which='major', pad=0)\n",
    "    ax.tick_params(axis='y', which='major', pad=0)\n",
    "\n",
    "    save_fig(f'OddsRatio-ChromHMM-{title}')\n",
    "    \n",
    "    plt.show()\n",
    "\n",
    "tissues = ['forebrain', 'midbrain', 'hindbrain', 'neural-tube', \n",
    "           'embryonic-facial-prominence', 'limb',\n",
    "           'heart', 'liver']\n",
    "marks = ['H3K9me3', 'H3K36me3', 'H3K27me3', 'H3K27ac', 'H3K4me1', 'H3K4me2', 'H3K4me3','H3K9ac']\n",
    "\n",
    "def get_median_naninc(df, mark, hist_metric=\"\", time_1=\"\", time_2=\"\", time_3=\"\", tissue=\"brain\"):\n",
    "    cols = []\n",
    "    for c in df.columns:\n",
    "        if tissue in c and mark in c and hist_metric in c and (time_1 in c or time_2 in c or time_3 in c):\n",
    "            cols.append(c)\n",
    "    # get nan median\n",
    "    vals = np.nanmean(df[cols].values, axis=1)\n",
    "    return vals\n",
    "\n",
    "    \n",
    "def plot_mark_heatmap(df, idxs, title, mark_cutoff=3.0):\n",
    "    mark_df = pd.DataFrame()\n",
    "    mark_values = []\n",
    "    mean_col = []\n",
    "    num_genes = 0\n",
    "    if idxs is None:\n",
    "        num_genes = len(df)\n",
    "    else:\n",
    "        num_genes = len(idxs) * 1.0\n",
    "    for t in [['10.5', '11.5', '12.5'], ['15.5', '15.5', '16.5']]:\n",
    "        tissue_titles = []\n",
    "        marks_title = []\n",
    "        for m in marks:\n",
    "            mark_col = []\n",
    "            for tissue in tissues:\n",
    "                median_all_data = get_median_naninc(df, m, \"signal\", t[0], t[1], t[2], tissue)\n",
    "                if idxs is None:\n",
    "                    median_genes = median_all_data\n",
    "                else:\n",
    "                    median_genes = median_all_data[idxs]     \n",
    "                has_mark = 1.0 * len(np.where(median_genes > mark_cutoff)[0])\n",
    "                if has_mark == 0:\n",
    "                    mark_col.append(0)\n",
    "                else:\n",
    "                    mark_col.append(has_mark/num_genes) #np.nan_to_num(np.nanmean(median_genes)))\n",
    "                if string.capwords(tissue) not in tissue_titles:\n",
    "                    if tissue == 'embryonic-facial-prominence':\n",
    "                        if 'E.F.P' not in tissue_titles:\n",
    "                            tissue_titles.append('E.F.P')\n",
    "                    else:\n",
    "                        tissue_titles.append(string.capwords(tissue))\n",
    "            marks_title.append(string.capwords(m))\n",
    "            mark_df[string.capwords(m)] = mark_col\n",
    "\n",
    "        mark_df['Tissue'] = tissue_titles\n",
    "\n",
    "        heatmap = Heatmap(mark_df, marks_title, 'Tissue', vmin=0, vmax=1, cmap='Greens', figsize=(2, 2),\n",
    "                          title=f'{title} {\"-\".join(t)}', cluster_rows=False, cluster_cols=False)\n",
    "        heatmap.plot()\n",
    "        pplot()\n",
    "        print(mark_df.head())\n",
    "        save_fig(f'mark_all_tissues_signal-{title}_{\"-\".join(t)}')\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 13) Setup the groups\n",
    "\n",
    "    1) Significnat (p < 0.05 in at least one experiment (anterior time or spatial wt/ko))\n",
    "        1.a) has H3K27me3 mark (signal at e16.5 > 3.0)\n",
    "            1.a.i) Expressed (mean (log2(TMM + 1) > 2 in WT OR KO) \n",
    "            1.a.j) Not expressed\n",
    "        1.b) doesn't have H3K27me3 (signal at e16.5 < 3.0)\n",
    "            1.b.i) Expressed\n",
    "            1.b.j) Not expressed\n",
    "    2) Not significant \n",
    "        2.a) Has (as above)\n",
    "            2.a.i) Exprerssed\n",
    "            2.a.j) Not expressed\n",
    "        2.b) Has not (as above)\n",
    "            2.b.i) Expressed\n",
    "            2.b.j) Not expressed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmm_cutoff = 0.5\n",
    "\n",
    "df_bg = df_all.copy()  # Make a copy so we don't override the other one\n",
    "\n",
    "# Also have a look at how the marks plot over genes that are not significant\n",
    "chrom_dir = os.path.join(input_dir, \"supps\", \"annot\")\n",
    "chromhmm_annot = pd.read_csv(f'{chrom_dir}/e16.5_forebrain_15_segments.csv')\n",
    "print(len(chromhmm_annot), len(df_bg))\n",
    "# Make sure we only keep 1 annotation per gene\n",
    "chromhmm_annot = chromhmm_annot.groupby('external_gene_name').first()\n",
    "# Merge this with our dataframe\n",
    "df_bg = df_bg.merge(chromhmm_annot, on='external_gene_name', how='left', suffixes=('', '_chmm'))\n",
    "# Ensure the new length is the same as the old length\n",
    "ps = ((1.0 * (df_all['padj_sc'].values < 0.05)) + (1.0 * (df_all['padj_mb'].values < 0.05))\n",
    "     +  (1.0 * (df_all['padj_hb'].values < 0.05)) + (1.0 * (df_all['padj_fb'].values < 0.05))\n",
    "     + (1.0 * (df_all['padj_a11'].values < 0.05)) \n",
    "     + (1.0 * (df_all['padj_a13'].values < 0.05)) \n",
    "     + (1.0 * (df_all['padj_a15'].values < 0.05))\n",
    "     + (1.0 * (df_all['padj_a18'].values < 0.05)))\n",
    "\n",
    "\n",
    "df_bg['p_mask'] = ps\n",
    "df_bg = df_bg.fillna(0)\n",
    "ko_cols = [c for c in df_bg if 'ko' in c]\n",
    "wt_cols = [c for c in df_bg if 'wt' in c]\n",
    "\n",
    "expression_mask = ((1.0 * (np.mean(df_all[wt_cols].values, axis=1) >= tmm_cutoff)) + (1.0 * (np.mean(df_all[ko_cols].values, axis=1) >= tmm_cutoff)))\n",
    "df_bg['expression_mask'] = expression_mask\n",
    "df_bg = df_bg[df_bg['expression_mask'] > tmm_cutoff]\n",
    "\n",
    "cutoff = 1.0\n",
    "log_str = 'log2FoldChange'\n",
    "logfc = ((1.0 * (abs(df_bg[f'{log_str}_sc'].values) > cutoff)) \n",
    "         + (1.0 * (abs(df_bg[f'{log_str}_mb'].values) > cutoff)) \n",
    "         + (1.0 * (abs(df_bg[f'{log_str}_hb'].values) > cutoff)) \n",
    "         + (1.0 * (abs(df_bg[f'{log_str}_fb'].values) > cutoff)) \n",
    "         + (1.0 * (abs(df_bg[f'{log_str}_a11'].values) > cutoff))\n",
    "         + (1.0 * (abs(df_bg[f'{log_str}_a13'].values) > cutoff)) \n",
    "         + (1.0 * (abs(df_bg[f'{log_str}_a15'].values) > cutoff)) \n",
    "         + (1.0 * (abs(df_bg[f'{log_str}_a18'].values) > cutoff))\n",
    "        )\n",
    "\n",
    "df_bg['logfc_mask'] = logfc\n",
    "\n",
    "# Add in the H3K mark --> we'll just look in the brain regions at e16.5\n",
    "h3k_cols = [c for c in df_bg.columns if 'brain' in c and '16.5' in c and 'signal' in c and 'H3K27me3' in c]\n",
    "df_bg['hk3_mean'] = np.nanmean(df_bg[h3k_cols].values, axis=1)\n",
    "\n",
    "ectopic = ((1.0 * (np.mean(df_bg[wt_cols].values, axis=1) <= tmm_cutoff)) + (1.0 * (np.mean(df_bg[ko_cols].values, axis=1) >= tmm_cutoff)))\n",
    "df_bg['ectopic'] = ectopic\n",
    "\n",
    "df_sig = df_bg[df_bg['p_mask'] > 0]\n",
    "df_ns = df_bg[df_bg['p_mask'] == 0]\n",
    "h3k_cutoff = 0.0\n",
    "\n",
    "df_sig_hasH3 = df_sig[df_sig['hk3_mean'] > h3k_cutoff]\n",
    "df_hasH3_exp = df_bg[df_bg['hk3_mean'] > h3k_cutoff]\n",
    "\n",
    "df_noH3_exp = df_bg[df_bg['hk3_mean'] == 0]\n",
    "df_noH3_exp = df_noH3_exp[df_noH3_exp['expression_mask'] > tmm_cutoff]\n",
    "\n",
    "pert_cutoff = 2\n",
    "\n",
    "df_sig_hasH3_pert = df_sig_hasH3[df_sig_hasH3['logfc_mask'] > pert_cutoff]  # Consistently perturbed\n",
    "df_sig_hasH3_unpert = df_sig_hasH3[df_sig_hasH3['logfc_mask'] <= pert_cutoff]  # Inconsistently perturbed\n",
    "\n",
    "df_sig_hasH3_ectopic = df_sig_hasH3[df_sig_hasH3['ectopic'] == pert_cutoff]\n",
    "\n",
    "df_sig_hasH3_exp = df_sig_hasH3[df_sig_hasH3['expression_mask'] > tmm_cutoff]\n",
    "df_sig_hasH3_exp_pert = df_sig_hasH3_exp[df_sig_hasH3_exp['logfc_mask'] > pert_cutoff]  # Consistently perturbed\n",
    "df_sig_hasH3_exp_unpert = df_sig_hasH3_exp[df_sig_hasH3_exp['logfc_mask'] <= pert_cutoff]  # Inonsistently perturbed\n",
    "\n",
    "df_sig_hasH3_not_exp = df_sig_hasH3[df_sig_hasH3['expression_mask'] <= tmm_cutoff]\n",
    "\n",
    "df_sig_noH3 = df_sig[df_sig['hk3_mean'] <= h3k_cutoff]\n",
    "df_sig_noH3_pert = df_sig_noH3[df_sig_noH3['logfc_mask'] > pert_cutoff]  # Consistently perturbed\n",
    "df_sig_noH3_unpert = df_sig_noH3[df_sig_noH3['logfc_mask'] <= pert_cutoff]  # Inonsistently perturbed\n",
    "\n",
    "df_sig_noH3_exp = df_sig_noH3[df_sig_noH3['expression_mask'] > tmm_cutoff]\n",
    "# #df_bg = df_bg[df_bg['logfc_mask'] > 2] \n",
    "df_sig_noH3_exp_pert = df_sig_noH3_exp[df_sig_noH3_exp['logfc_mask'] > pert_cutoff]\n",
    "df_sig_noH3_exp_unpert = df_sig_noH3_exp[df_sig_noH3_exp['logfc_mask'] <= pert_cutoff]\n",
    "\n",
    "\n",
    "df_sig_noH3_ectopic = df_sig_noH3[df_sig_noH3['ectopic'] == 2]\n",
    "\n",
    "df_sig_noH3_notexp = df_sig_noH3[df_sig_noH3['expression_mask'] <= tmm_cutoff]\n",
    "\n",
    "\n",
    "df_ns_hasH3 = df_ns[df_ns['hk3_mean'] > h3k_cutoff]\n",
    "\n",
    "df_ns_hasH3_exp = df_ns_hasH3[df_ns_hasH3['expression_mask'] > tmm_cutoff]\n",
    "\n",
    "df_ns_hasH3_not_exp = df_ns_hasH3[df_ns_hasH3['expression_mask'] <= tmm_cutoff]\n",
    "df_ns_noH3 = df_ns[df_ns['hk3_mean'] <= h3k_cutoff]\n",
    "df_ns_noH3_exp = df_ns_noH3[df_ns_noH3['expression_mask'] > tmm_cutoff]\n",
    "df_ns_noH3_notexp = df_ns_noH3[df_ns_noH3['expression_mask'] <= tmm_cutoff]\n",
    "\n",
    "df_ns_noH3_ectopic = df_ns_noH3[df_ns_noH3['ectopic'] == 2]\n",
    "df_ns_H3_ectopic = df_ns_hasH3[df_ns_hasH3['ectopic'] == 2]\n",
    "\n",
    "dfs = [df_bg, df_sig, df_ns, df_sig_hasH3, df_ns_hasH3, df_sig_noH3, df_ns_noH3,\n",
    "       df_sig_hasH3_pert, df_sig_hasH3_unpert, df_sig_noH3_pert, df_sig_noH3_unpert,\n",
    "       df_sig_hasH3_exp_pert, df_sig_hasH3_exp_unpert, df_sig_noH3_exp_pert, df_sig_noH3_exp_unpert,\n",
    "       df_sig_hasH3_exp, df_sig_hasH3_not_exp, df_sig_noH3_exp, df_sig_noH3_notexp, \n",
    "      df_ns_hasH3_exp, df_ns_hasH3_not_exp, df_ns_noH3_exp, df_ns_noH3_notexp,\n",
    "      df_sig_hasH3_ectopic, df_sig_noH3_ectopic]\n",
    "df_dict = {'All genes': df_bg, \n",
    "           'Genes with H3K27me3': df_hasH3_exp,\n",
    "           'Genes without H3K27me3': df_noH3_exp,\n",
    "             'Genes with at least 1 sig': df_sig, \n",
    "             'Non-sig genes': df_ns, \n",
    "             'Sig with H3K27me3': df_sig_hasH3, \n",
    "             'NS with H3K27me3': df_ns_hasH3,\n",
    "             'Sig unmarked': df_sig_noH3, \n",
    "             'NS unmarked': df_ns_noH3,\n",
    "             'Sig Marked Perturbed': df_sig_hasH3_pert,  \n",
    "             'Sig Marked un-Perturbed': df_sig_hasH3_unpert, \n",
    "             'Sig Unmarked Perturbed': df_sig_noH3_pert,  \n",
    "             'Sig Unmarked un-Perturbed': df_sig_noH3_unpert,\n",
    "             'Exp. Sig Marked Perturbed': df_sig_hasH3_exp_pert, \n",
    "             'Exp. Sig Marked un-Perturbed': df_sig_hasH3_exp_unpert, \n",
    "             'Exp. Sig Unmarked Perturbed': df_sig_noH3_exp_pert,  \n",
    "             'Exp. Sig Unmarked un-Perturbed': df_sig_noH3_exp_unpert,\n",
    "            'Sig genes with H3K27me3 and expression': df_sig_hasH3_exp, \n",
    "             'Sig genes with H3K27me3 NO expression': df_sig_hasH3_not_exp, \n",
    "            'Sig genes unmarked and expression': df_sig_noH3_exp, \n",
    "             'Sig genes unmarked NO expression': df_sig_noH3_notexp,\n",
    "            'NS genes with H3K27me3 and expression': df_ns_hasH3_exp,\n",
    "             'NS genes with H3K27me3 NO expression': df_ns_hasH3_not_exp, \n",
    "            'NS genes unmarked and expression': df_ns_noH3_exp,\n",
    "             'NS genes unmarked NO expression': df_ns_noH3_notexp, \n",
    "            'Sig ectopic marked': df_sig_hasH3_ectopic, \n",
    "             'Sig ectopic unmarked': df_sig_noH3_ectopic, \n",
    "            'NS ectopic marked': df_ns_H3_ectopic,\n",
    "            'NS ectopic unmarked': df_ns_noH3_ectopic,\n",
    "          }\n",
    "\n",
    "\n",
    "emg = [g for g in df_dict['Sig ectopic marked'][gene_name].values]\n",
    "print(\", \".join(emg))\n",
    "emg = [g for g in df_dict['Sig ectopic unmarked'][gene_name].values]\n",
    "print(\", \".join(emg))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot correlation between FB logFC and median H3K27me3 mark"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "from scipy.stats import spearmanr\n",
    "from sciviso import Histogram, Scatterplot\n",
    "from numpy.polynomial.polynomial import polyfit\n",
    "import statsmodels.api as sm\n",
    "\n",
    "# Do this on the consistently affected dataset\n",
    "df_const = pd.read_csv(f'{output_dir}df-consistent_epi-2500_{date}.csv')\n",
    "\n",
    "\n",
    "#calculate Spearman Rank correlation and corre\n",
    "# Also do the main markers\n",
    "fb_genes = ['Emx1', 'Eomes', 'Tbr1', 'Foxg1', 'Lhx6']\n",
    "mb_genes = ['En1', 'En2', 'Lmx1a', 'Bhlhe23', 'Sall4']\n",
    "hb_genes = ['Phox2b', 'Krox20', 'Fev', 'Hoxb1',  'Hoxd3']\n",
    "sc_genes = ['Hoxd8', 'Hoxd9', 'Hoxd10', 'Hoxd11','Hoxa7', 'Hoxa9', 'Hoxa10',\n",
    "              'Hoxb9', 'Hoxb13',  'Hoxc8', 'Hoxc9', 'Hoxc10', 'Hoxc11', 'Hoxc12', 'Hoxc13']\n",
    "\n",
    "# Fill NAs in genes that didn't have the H3K27me3 mark\n",
    "df_const_na = df_const.fillna(0)\n",
    "h3k_cols = [c for c in df_const_na.columns if 'brain' in c and 'H3K27me3' in c and 'signal' in c]\n",
    "Y = np.nan_to_num(np.nanmedian(df_const_na[h3k_cols].values, axis=1))\n",
    "df_const_na['Median_H3K'] = Y\n",
    "X = df_const_na['log2FoldChange_fb'].values\n",
    "results = sm.OLS(Y,sm.add_constant(X)).fit()\n",
    "print(results.summary())\n",
    "corr, _ = spearmanr(X, Y)\n",
    "u.dp([\"H3K27me3 median signal in brain regions vs FB logFC:\\n\", 'Spearmans correlation: %.3f' % corr])\n",
    "\n",
    "# now run the scatter\n",
    "s = Scatterplot(df_const_na,'Median_H3K', 'log2FoldChange_fb', title='H3K27me3 vs FB logFC', colour=\"green\", \n",
    "                points_to_annotate=['Foxg1', 'Otx2', 'Pax2', 'Hoxd9'], \n",
    "                annotation_label=gene_name, add_legend=False)\n",
    "s.opacity=0.1\n",
    "ax = s.plot()\n",
    "ax.set_xlim(-3, 20)\n",
    "ax.set_ylim(-5, 12)\n",
    "\n",
    "save_fig(f'Scatter_H3K27me3vsLogFC')\n",
    "plt.show()\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "--------------------------------------------------------\n",
    "Do the same with H3K27ac as a control\n",
    "--------------------------------------------------------\n",
    "\"\"\"\n",
    "h3k_cols = [c for c in df_const_na.columns if 'brain' in c and 'H3K27ac' in c and 'signal' in c]\n",
    "Y = np.nan_to_num(np.nanmedian(df_const_na[h3k_cols].values, axis=1))\n",
    "df_const_na['Median_H3K'] = Y\n",
    "X = df_const_na['log2FoldChange_fb'].values\n",
    "results = sm.OLS(Y,sm.add_constant(X)).fit()\n",
    "print(results.summary())\n",
    "corr, _ = spearmanr(X, Y)\n",
    "u.dp([\"H3K27ac median signal in brain regions vs FB logFC:\\n\", 'Spearmans correlation: %.3f' % corr])\n",
    "\n",
    "# now run the scatter\n",
    "s = Scatterplot(df_const_na,'Median_H3K', 'log2FoldChange_fb', title='H3K27ac vs FB logFC', colour=\"grey\", \n",
    "                points_to_annotate=['Foxg1', 'Otx2', 'Pax2', 'Hoxd9'], \n",
    "                annotation_label=gene_name, add_legend=False)\n",
    "s.opacity=0.1\n",
    "ax = s.plot()\n",
    "ax.set_xlim(-3, 20)\n",
    "ax.set_ylim(-5, 12)\n",
    "\n",
    "save_fig(f'Scatter_H3K27acvsLogFC')\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "h3k_cols = [c for c in df_bg.columns if 'brain' in c and '10.5' in c and 'signal' in c and 'H3K27me3' in c]\n",
    "df_bg['hk3_mean'] = np.nanmean(df_bg[h3k_cols].values, axis=1)\n",
    "e10 = df_bg[df_bg['hk3_mean'] > h3k_cutoff][gene_name].values\n",
    "\n",
    "h3k_cols = [c for c in df_bg.columns if 'brain' in c and '11.5' in c and 'signal' in c and 'H3K27me3' in c]\n",
    "df_bg['hk3_mean'] = np.nanmean(df_bg[h3k_cols].values, axis=1)\n",
    "e11 = df_bg[df_bg['hk3_mean'] > h3k_cutoff][gene_name].values\n",
    "\n",
    "h3k_cols = [c for c in df_bg.columns if 'brain' in c and '12.5' in c and 'signal' in c and 'H3K27me3' in c]\n",
    "df_bg['hk3_mean'] = np.nanmean(df_bg[h3k_cols].values, axis=1)\n",
    "e12 = df_bg[df_bg['hk3_mean'] > h3k_cutoff][gene_name].values\n",
    "\n",
    "h3k_cols = [c for c in df_bg.columns if 'brain' in c and '13.5' in c and 'signal' in c and 'H3K27me3' in c]\n",
    "df_bg['hk3_mean'] = np.nanmean(df_bg[h3k_cols].values, axis=1)\n",
    "e13 = df_bg[df_bg['hk3_mean'] > h3k_cutoff][gene_name].values\n",
    "\n",
    "h3k_cols = [c for c in df_bg.columns if 'brain' in c and '14.5' in c and 'signal' in c and 'H3K27me3' in c]\n",
    "df_bg['hk3_mean'] = np.nanmean(df_bg[h3k_cols].values, axis=1)\n",
    "e14 = df_bg[df_bg['hk3_mean'] > h3k_cutoff][gene_name].values\n",
    "\n",
    "h3k_cols = [c for c in df_bg.columns if 'brain' in c and '15.5' in c and 'signal' in c and 'H3K27me3' in c]\n",
    "df_bg['hk3_mean'] = np.nanmean(df_bg[h3k_cols].values, axis=1)\n",
    "e15 = df_bg[df_bg['hk3_mean'] > h3k_cutoff][gene_name].values\n",
    "\n",
    "h3k_cols = [c for c in df_bg.columns if 'brain' in c and '16.5' in c and 'signal' in c and 'H3K27me3' in c]\n",
    "df_bg['hk3_mean'] = np.nanmean(df_bg[h3k_cols].values, axis=1)\n",
    "e16 = df_bg[df_bg['hk3_mean'] > h3k_cutoff][gene_name].values\n",
    "\n",
    "### ----------- \n",
    "plt.plot([10.5, 11.5, 12.5, 13.5, 14.5, 15.5, 16.5], \n",
    "         [len(e10), len(e11), len(e12), len(e13), len(e14), len(e15), len(e16)]\n",
    "        )\n",
    "afct = set(df_sig[gene_name].values)\n",
    "cafct = set(df_sig[df_sig['logfc_mask'] > pert_cutoff][gene_name].values)\n",
    "\n",
    "plt.plot([10.5, 11.5, 12.5, 13.5, 14.5, 15.5, 16.5], \n",
    "         [len(set(afct) & set(e10)), len(set(afct) & set(e11)), \n",
    "          len(set(afct) & set(e12)), len(set(afct) & set(e13)), len(set(afct) & set(e14)), \n",
    "          len(set(afct) & set(e15)), len(set(afct) & set(e16))]\n",
    "        )\n",
    "\n",
    "plt.plot([10.5, 11.5, 12.5, 13.5, 14.5, 15.5, 16.5], \n",
    "         [len(set(cafct) & set(e10)), len(set(cafct) & set(e11)), \n",
    "          len(set(cafct) & set(e12)), len(set(cafct) & set(e13)), len(set(cafct) & set(e14)), \n",
    "          len(set(cafct) & set(e15)), len(set(cafct) & set(e16))]\n",
    "        )\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# ==============================\n",
    "plt.plot([10.5,  12.5, 13.5, 14.5, 15.5, 16.5], \n",
    "         [len(e10), len(e12), len(e13), len(e14), len(e15), len(e16)], label=\"marked\"\n",
    "        )\n",
    "plt.plot([10.5,  12.5, 13.5, 14.5, 15.5, 16.5], \n",
    "         [len(df_bg) - len(e10), len(df_bg) - len(e12), len(df_bg) - len(e13), len(df_bg) - len(e14), len(df_bg) - len(e15), len(df_bg) - len(e16)]\n",
    "        , label=\"bg - marked\")\n",
    "afct = set(df_sig[gene_name].values)\n",
    "cafct = set(df_sig[df_sig['logfc_mask'] > pert_cutoff][gene_name].values)\n",
    "afct = set(df_sig[gene_name].values)\n",
    "\n",
    "plt.plot([10.5, 12.5, 13.5, 14.5, 15.5, 16.5], \n",
    "         [len(set(afct) & set(e10)), \n",
    "          len(set(afct) & set(e12)), len(set(afct) & set(e13)), len(set(afct) & set(e14)), \n",
    "          len(set(afct) & set(e15)), len(set(afct) & set(e16))], label=\"Intersect, affect and marked\"\n",
    "        )\n",
    "plt.plot([10.5, 12.5, 13.5, 14.5, 15.5, 16.5], \n",
    "         [len(set(afct)) - len(set(afct) & set(e10)), \n",
    "          len(set(afct)) - len(set(afct) & set(e12)), len(set(afct)) - len(set(afct) & set(e13)),\n",
    "          len(set(afct)) - len(set(afct) & set(e14)), \n",
    "          len(set(afct)) - len(set(afct) & set(e15)), len(set(afct)) - len(set(afct) & set(e16))],\n",
    "         label=\"unmarked affected\"\n",
    "        )\n",
    "plt.plot([10.5, 12.5, 13.5, 14.5, 15.5, 16.5], \n",
    "         [len(set(cafct) & set(e10)), \n",
    "          len(set(cafct) & set(e12)), len(set(cafct) & set(e13)), len(set(cafct) & set(e14)), \n",
    "          len(set(cafct) & set(e15)), len(set(cafct) & set(e16))], \n",
    "         label=\"consist. affected, marked\"\n",
    "        )\n",
    "plt.legend(loc=\"upper left\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "#------------ \n",
    "h3k_cols = [c for c in df_bg.columns if 'brain' in c and 'signal' in c and 'H3K27me3' in c and '16.5' in c]\n",
    "df_bg['hk3_mean'] = np.nanmean(df_bg[h3k_cols].values, axis=1)\n",
    "plot_venn((set(df_bg[df_bg['hk3_mean'] == 0][gene_name].values), \n",
    "           set(df_sig[gene_name].values), set(df_sig[df_sig['logfc_mask'] > pert_cutoff][gene_name].values)),\n",
    "          ('Un-Marked genes', 'Affected Genes', 'Const. Affected Genes'), 'venn_unmaked_marked_const_aff')\n",
    "\n",
    "#------------ \n",
    "h3k_cols = [c for c in df_bg.columns if 'brain' in c and 'signal' in c and 'H3K27me3' in c and '10.5' in c]\n",
    "df_bg['hk3_mean'] = np.nanmean(df_bg[h3k_cols].values, axis=1)\n",
    "plot_venn((set(df_bg[df_bg['hk3_mean'] > h3k_cutoff][gene_name].values), \n",
    "           set(df_sig[gene_name].values), set(df_sig[df_sig['logfc_mask'] > pert_cutoff][gene_name].values)),\n",
    "          ('Marked genes', 'Affected Genes', 'Const. Affected Genes'), 'venn_marked_const_aff')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print out the size of each DF\n",
    "for label, d in df_dict.items():\n",
    "    u.dp([label, len(d)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 13) Test for enrichment of specific marks using the Gorkin et al annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save each DF so we can test each of them for ORA (just save the gene name to avoid dups.)\n",
    "for label, d in df_dict.items():\n",
    "    print(label)\n",
    "    dn = d[gene_id]\n",
    "    dn.to_csv(os.path.join(output_dir, f'{label.replace(\" \", \"_\")}_gene-name_{date}.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "--------------------------------------------------------\n",
    "Test for enriched gorkin et al states\n",
    "--------------------------------------------------------\n",
    "\"\"\"\n",
    "for label, d in df_dict.items():\n",
    "    run_annot_plot(df_bg, d, title=f'Annot {label}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for label, d in df_dict.items():\n",
    "    try:\n",
    "        logFC_up = d[d['log2FoldChange_fb'] > 0.5]\n",
    "        logFC_down = d[d['log2FoldChange_fb'] < -0.5]\n",
    "        u.dp([label, len(d), 100 * (len(d)/len(df_all))])\n",
    "        print('LogFC up: ', len(logFC_up), 100 * (len(logFC_up)/len(d)))\n",
    "        print('LogFC down: ', len(logFC_down), 100 * (len(logFC_down)/len(d)))\n",
    "    except:\n",
    "        print(\"unable to run:\", label)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 15) Plot the general trend of the RNA seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fb_genes = ['Emx1', 'Eomes', 'Tbr1', 'Foxg1', 'Lhx6']\n",
    "mb_genes = ['En1', 'En2', 'Lmx1a', 'Bhlhe23', 'Sall4']\n",
    "hb_genes = ['Phox2b', 'Krox20', 'Fev', 'Hoxb1',  'Hoxd3']\n",
    "sc_genes = ['Hoxd8', 'Hoxd9', 'Hoxd10', 'Hoxd11','Hoxa7', 'Hoxa9', 'Hoxa10',\n",
    "            'Hoxb9', 'Hoxb13',  'Hoxc8', 'Hoxc9', 'Hoxc10', 'Hoxc11', 'Hoxc12', 'Hoxc13']\n",
    "progenitors = ['Sox2', 'Sox1', 'Sox3', 'Hes1', 'Hes5']\n",
    "neurons = ['Tubb3', 'Snap25', 'Syt1', 'Slc32a1','Slc17a6']\n",
    "glia = ['Pdgfra', 'Cspg4', 'Aqp4', 'Egfr', 'Slc6a11']\n",
    "\n",
    "from collections import defaultdict\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from statannot import add_stat_annotation\n",
    "\n",
    "hox_genes = sc_genes\n",
    "forebrain_genes = fb_genes\n",
    "h3cols = []\n",
    "fb_rna_cols = []\n",
    "sc_rna_cols = []\n",
    "\n",
    "# Do it in this way so they are correctly ordered\n",
    "for c in df_bg.columns:\n",
    "    if 'H3K27me' in c and 'signal' in c:\n",
    "        h3cols.append(c)\n",
    "    elif 'wt' in c:\n",
    "        if 'fb' in c:\n",
    "            fb_rna_cols.append(c)\n",
    "        if 'sc' in c:\n",
    "            sc_rna_cols.append(c)\n",
    "for c in df_bg.columns:\n",
    "    if 'ko' in c:\n",
    "        if 'fb' in c:\n",
    "            fb_rna_cols.append(c)\n",
    "        if 'sc' in c:\n",
    "            sc_rna_cols.append(c)\n",
    "\n",
    "wt_colour = \"#AADFF1\"   \n",
    "    \n",
    "def plot_gene_boxplot(df, title, cluster_id, cols):\n",
    "    boxplot = Boxplot(df, gene_name, cols[0])\n",
    "    boxplot.label_font_size = 6\n",
    "    boxplot.title_font_size = 8\n",
    "    sns.set(rc={'figure.figsize': (1.5, 1.5), 'font.family': 'sans-serif',\n",
    "                    'font.sans-serif': 'Arial', 'font.size': 8.0}, style='ticks')\n",
    "    boxplot.palette = sci_colour\n",
    "    box_df = boxplot.format_data_for_boxplot(df, cols, gene_name, df[gene_name].values)\n",
    "    is_hox = []\n",
    "    sns.set_style(\"ticks\")\n",
    "    for c in box_df['Samples'].values:\n",
    "        if 'Hox' in c:\n",
    "            is_hox.append('Hox')\n",
    "        else:\n",
    "            is_hox.append('Hox like')\n",
    "    box_df['is_hox'] = is_hox\n",
    "\n",
    "    boxplot = Boxplot(box_df, \"Conditions\", \"Values\", add_stats=False, add_dots=False, \n",
    "                       order=cols)\n",
    "    boxplot.label_font_size = 6\n",
    "    boxplot.title_font_size = 8\n",
    "    plt.rcParams['figure.figsize'] = (1.5,1.5)\n",
    "    sns.set(rc={'figure.figsize': (1.5, 1.5), 'font.family': 'sans-serif',\n",
    "                    'font.sans-serif': 'Arial', 'font.size': 8.0}, style='ticks')\n",
    "    boxplot.palette = sci_colour\n",
    "    ax = boxplot.plot()\n",
    "    ax.tick_params(labelsize=6)\n",
    "    \n",
    "    plt.title(title, {'fontsize': 8, 'fontweight': 700})\n",
    "    ax.set_ylim(0, 2.5)\n",
    "    c = 0\n",
    "    for b in ax.artists:\n",
    "        if c < len(cols)/2:\n",
    "            b.set_facecolor(wt_colour)\n",
    "        else:\n",
    "            b.set_facecolor(ko_colour)\n",
    "        c += 1\n",
    "    save_fig(f'{title}')\n",
    "    \n",
    "    plt.show()\n",
    "\n",
    "avgs_fb = [['wt11fb1',\n",
    "         'wt11fb2'],\n",
    "         ['wt13fb1',\n",
    "         'wt13fb2'],\n",
    "         ['wt15fb1',\n",
    "         'wt15fb2'],\n",
    "         ['wt18fb1',\n",
    "         'wt18fb2'],\n",
    "         ['ko11fb1',\n",
    "         'ko11fb2'],\n",
    "         ['ko13fb1',\n",
    "         'ko13fb2'],\n",
    "         ['ko15fb1',\n",
    "         'ko15fb2'],\n",
    "         ['ko18fb1',\n",
    "         'ko18fb2']]\n",
    "\n",
    "avgs_sc = [['wt11sc1',\n",
    "         'wt11sc2'],\n",
    "         ['wt13sc1',\n",
    "         'wt13sc2'],\n",
    "         ['wt15sc1',\n",
    "         'wt15sc2'],\n",
    "         ['wt18sc1',\n",
    "         'wt18sc2'],\n",
    "         ['ko11sc1',\n",
    "         'ko11sc2'],\n",
    "         ['ko13sc1',\n",
    "         'ko13sc2'],\n",
    "         ['ko15sc1',\n",
    "         'ko15sc2'],\n",
    "         ['ko18sc1',\n",
    "         'ko18sc2']]\n",
    "avgs_hb = [['wt11hb1',\n",
    "         'wt11hb2'],\n",
    "         ['wt13hb1',\n",
    "         'wt13hb2'],\n",
    "         ['wt15hb1',\n",
    "         'wt15hb2'],\n",
    "         ['wt18hb1',\n",
    "         'wt18hb2'],\n",
    "         ['ko11hb1',\n",
    "         'ko11hb2'],\n",
    "         ['ko13hb1',\n",
    "         'ko13hb2'],\n",
    "         ['ko15hb1',\n",
    "         'ko15hb2'],\n",
    "         ['ko18hb1',\n",
    "         'ko18hb2']]\n",
    "\n",
    "avgs_mb = [['wt11mb1',\n",
    "         'wt11mb2'],\n",
    "         ['wt13mb1',\n",
    "         'wt13mb2'],\n",
    "         ['wt15mb1',\n",
    "         'wt15mb2'],\n",
    "         ['wt18mb1',\n",
    "         'wt18mb2'],\n",
    "         ['ko11mb1',\n",
    "         'ko11mb2'],\n",
    "         ['ko13mb1',\n",
    "         'ko13mb2'],\n",
    "         ['ko15mb1',\n",
    "         'ko15mb2'],\n",
    "         ['ko18mb1',\n",
    "         'ko18mb2']]\n",
    "for label, df in df_dict.items():\n",
    "    if 'ectopic' in label:\n",
    "\n",
    "        avgs_df = pd.DataFrame()\n",
    "        avgs_df[gene_name] = df[gene_name].values\n",
    "        fb_cols = []\n",
    "        for f in avgs_fb:\n",
    "            new_col = f'{f[0][:2]} {f[0][2:4]} FB'\n",
    "            avgs_df[new_col] = (df[f[0]].values + df[f[1]].values) / 2.0\n",
    "            fb_cols.append(new_col)\n",
    "\n",
    "        sc_cols = []\n",
    "        for f in avgs_sc:\n",
    "            new_col = f'{f[0][:2]} {f[0][2:4]} SC'\n",
    "            avgs_df[new_col] = (df[f[0]].values + df[f[1]].values) / 2.0\n",
    "            sc_cols.append(new_col)\n",
    "\n",
    "        mb_cols = []\n",
    "        for f in avgs_mb:\n",
    "            new_col = f'{f[0][:2]} {f[0][2:4]} MB'\n",
    "            avgs_df[new_col] = (df[f[0]].values + df[f[1]].values) / 2.0\n",
    "            mb_cols.append(new_col)\n",
    "\n",
    "        hb_cols = []\n",
    "        for f in avgs_hb:\n",
    "            new_col = f'{f[0][:2]} {f[0][2:4]} HB'\n",
    "            avgs_df[new_col] = (df[f[0]].values + df[f[1]].values) / 2.0\n",
    "            hb_cols.append(new_col)\n",
    "        plot_gene_boxplot(avgs_df, f'{label} FB', '', fb_cols)\n",
    "        plot_gene_boxplot(avgs_df, f'{label} MB', '', mb_cols)\n",
    "        plot_gene_boxplot(avgs_df, f'{label} HB', '', hb_cols)\n",
    "\n",
    "        plot_gene_boxplot(avgs_df, f'{label} SC', '', sc_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "---------------------------------------------------------------\n",
    "            Read in results from DEseq2 and format DFs nicer\n",
    "---------------------------------------------------------------\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "# Want to consider genes that are DE in FB & MB vs SC and HB and then look at those \n",
    "# which are \"consistently affected\"\n",
    "# i.e. for each gene make a \"code of how it was affected\"\n",
    "df_all = pd.read_csv(f'{output_dir}df-all_epi-2500_{date}.csv')\n",
    "\n",
    "df_dict = {}\n",
    "deseq2_files = os.listdir(r_dir)\n",
    "for f in deseq2_files:\n",
    "    if 'DEseq2' in f:\n",
    "        try:\n",
    "            de_df = pd.read_csv(os.path.join(r_dir, f))\n",
    "            de_df = de_df.rename(columns={de_df.columns[0]: 'u_id'})\n",
    "            gene_names = [s.split('-')[1] for s in list(de_df['u_id'].values)]\n",
    "            gene_ids = [s.split('-')[0] for s in list(de_df['u_id'].values)]\n",
    "            de_df['padj'] = de_df['padj'].fillna(1) # Replace Nan p values with 1.0s\n",
    "            # Replace Nans with 0's for other values\n",
    "            de_df = de_df.replace(np.nan, 0)\n",
    "            de_df[gene_id] = gene_ids\n",
    "            de_df[gene_name] = gene_names\n",
    "            df_dict[f] = de_df\n",
    "        except:\n",
    "            print(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot the venn diagrams of overlapping DE genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dict.keys()\n",
    "\n",
    "fb_unique_genes = []\n",
    "mb_unique_genes = []\n",
    "hb_unique_genes = []\n",
    "sc_unique_genes = []\n",
    "anterior_genes = []\n",
    "posterior_genes = []\n",
    "\n",
    "# DEseq2_CNS_wt_fb-hb_20210124.csv\n",
    "fb_genes_all = []\n",
    "mb_genes_all = []\n",
    "hb_genes_all = []\n",
    "sc_genes_all = []\n",
    "cutoff = 2.0\n",
    "neg_cutoff = -2.0\n",
    "for k, d in df_dict.items():\n",
    "    if 'wt_fb-' in k:\n",
    "        gs = d[d['padj'] < 0.05]\n",
    "        gs = gs[gs['log2FoldChange'] > cutoff] # Only want those that are \"selectively on \" in FB\n",
    "        fb_genes_all += list(gs[gene_name].values)\n",
    "    if 'wt_' in k and 'mb' in k:\n",
    "        gs = d[d['padj'] < 0.05]\n",
    "        if k == 'DEseq2_CNS_wt_fb-mb_20210124.csv':\n",
    "            gs = gs[gs['log2FoldChange'] < neg_cutoff]\n",
    "        else:\n",
    "            gs = gs[gs['log2FoldChange'] > cutoff]\n",
    "        mb_genes_all += list(gs[gene_name].values)\n",
    "    if 'wt_' in k and 'hb' in k:\n",
    "        gs = d[d['padj'] < 0.05]\n",
    "        if k == 'DEseq2_CNS_wt_hb-sc_20210124.csv':\n",
    "            gs = gs[gs['log2FoldChange'] > cutoff]\n",
    "        else:\n",
    "            gs = gs[gs['log2FoldChange'] < neg_cutoff]\n",
    "        hb_genes_all += list(gs[gene_name].values)\n",
    "    if 'wt_' in k and 'sc' in k:\n",
    "        gs = d[d['padj'] < 0.05]\n",
    "        gs = gs[gs['log2FoldChange'] < neg_cutoff] # Only want those that are \"selectively on \" in FB\n",
    "        sc_genes_all += list(gs[gene_name].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now see which are also in the consistently affected genes\n",
    "const_aff = pd.read_csv(f'{output_dir}df-consistent_epi-2500_{date}.csv')\n",
    "const_aff_genes = const_aff[gene_name]\n",
    "\n",
    "plot_venn([set(fb_genes_all), \n",
    "           set(mb_genes_all), set(hb_genes_all), set(sc_genes_all)],\n",
    "          ['FB', 'MB', 'HB', 'SC'], 'SpecificGenes_loc', colours=[fb_colour, mb_colour, hb_colour, sc_colour])\n",
    "\n",
    "plot_venn([set(fb_genes_all) & set(const_aff_genes), \n",
    "           set(mb_genes_all) & set(const_aff_genes), \n",
    "           set(hb_genes_all) & set(const_aff_genes), \n",
    "           set(sc_genes_all) & set(const_aff_genes)],\n",
    "          ['FB', 'MB', 'HB', 'SC'], 'SpecificGenes_CAFF_loc', colours=[fb_colour, mb_colour, hb_colour, sc_colour])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Print out the genes for the table\n",
    "\n",
    "This is just for supp. info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mb_caff = set(mb_genes_all) & set(const_aff_genes)\n",
    "fb_caff = set(fb_genes_all) & set(const_aff_genes)\n",
    "hb_caff = set(hb_genes_all) & set(const_aff_genes)\n",
    "sc_caff = set(sc_genes_all) & set(const_aff_genes)\n",
    "\n",
    "fb_spec = []\n",
    "anterior_spec = []\n",
    "for g in fb_caff:\n",
    "    if g not in mb_caff and g not in hb_caff and g not in sc_caff:\n",
    "        fb_spec.append(g)\n",
    "    if g in mb_caff and g not in hb_caff and g not in sc_caff:\n",
    "        anterior_spec.append(g)\n",
    "        \n",
    "mb_spec = []\n",
    "for g in mb_caff:\n",
    "    if g not in fb_caff and g not in hb_caff and g not in sc_caff:\n",
    "        mb_spec.append(g)\n",
    "        \n",
    "posterior_spec = []\n",
    "hb_spec = []\n",
    "for g in hb_caff:\n",
    "    if g not in mb_caff and g not in fb_caff and g not in sc_caff:\n",
    "        hb_spec.append(g)\n",
    "    if g in sc_caff and g not in mb_caff and g not in fb_caff:\n",
    "        posterior_spec.append(g)\n",
    "        \n",
    "sc_spec = []\n",
    "for g in sc_caff:\n",
    "    if g not in mb_caff and g not in hb_caff and g not in fb_caff:\n",
    "        sc_spec.append(g)\n",
    "       \n",
    "\n",
    "\n",
    "print(', '.join(fb_spec), '\\n')\n",
    "print(', '.join(mb_spec), '\\n')\n",
    "print(', '.join(hb_spec), '\\n')\n",
    "print(', '.join(sc_spec), '\\n')\n",
    "print(', '.join(anterior_spec), '\\n')\n",
    "print(', '.join(posterior_spec), '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run AP_axis functional\n",
    "\n",
    "Run the functional enrichment and then re-read in the values from the GSEA and use this to plot the\n",
    "pathway terms.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from adjustText import adjust_text\n",
    "\n",
    "from sciviso import Vis\n",
    "\n",
    "\n",
    "class Volcanoplot(Vis):\n",
    "\n",
    "    def __init__(self, df: pd.DataFrame, log_fc: str, p_val: str, label_column: str, title='',\n",
    "                 xlabel='', ylabel='', invert=False, p_val_cutoff=0.05,\n",
    "                 log_fc_cuttoff=2, label_big_sig=False, colours=None, offset=None,\n",
    "                 text_colours=None, values_to_label=None, max_labels=10, values_colours=None,\n",
    "                 figsize=(1.5, 1.5), title_font_size=8, label_font_size=6, title_font_weight=700):\n",
    "        super().__init__(df, figsize=figsize, title_font_size=title_font_size, label_font_size=label_font_size,\n",
    "                         title_font_weight=title_font_weight)\n",
    "        super().__init__(df)\n",
    "        self.log_fc = log_fc\n",
    "        self.p_val = p_val\n",
    "        self.p_val_cutoff = p_val_cutoff\n",
    "        self.log_fc_cuttoff = log_fc_cuttoff\n",
    "        self.values_to_label = values_to_label\n",
    "        self.label_big_sig = label_big_sig\n",
    "        self.invert = invert\n",
    "        self.label_column = label_column\n",
    "        self.offset = offset\n",
    "        self.label = 'volcanoplot'\n",
    "        self.colours = {'ns_small-neg-logFC': 'lightgrey',\n",
    "                        'ns_small-pos-logFC': 'lightgrey',\n",
    "                        'ns_big-neg-logFC': 'grey',\n",
    "                        'ns_big-pos-logFC': 'grey',\n",
    "                        'sig_small-neg-logFC': 'lightgrey',\n",
    "                        'sig_small-pos-logFC': 'lightgrey',\n",
    "                        'sig_big-neg-logFC': '#df80ff',\n",
    "                        'sig_big-pos-logFC': '#ffa366'} if colours is None else colours\n",
    "        self.xlabel = xlabel\n",
    "        self.ylabel = ylabel\n",
    "        self.title = title\n",
    "        self.figsize = figsize\n",
    "        self.max_labels = max_labels\n",
    "        self.values_colours = values_colours or {}\n",
    "        self.text_colours = text_colours or {}\n",
    "\n",
    "    def add_scatter_and_annotate(self, fig: plt, x_all: np.array, y_all: np.array,\n",
    "                                 colour: str, idxs: np.array, annotate=False, s=20):\n",
    "        x = x_all[idxs]\n",
    "        y = y_all[idxs]\n",
    "        ax = fig.scatter(x, y, c=colour, alpha=self.opacity, s=s, vmin=-10.0, vmax=10.0)\n",
    "\n",
    "        # Check if we want to annotate any of these with their gene IDs\n",
    "\n",
    "        if self.values_to_label is not None:\n",
    "            texts = []\n",
    "            labels = self.df[self.label_column].values[idxs]\n",
    "            for i, name in enumerate(labels):\n",
    "                if name in self.values_to_label:\n",
    "                    lbl_bg = self.values_colours.get(name)\n",
    "                    color = self.text_colours.get(name)\n",
    "                    texts.append(fig.text(x[i], y[i], name, color=color, fontsize=6,\n",
    "                                          bbox=dict(fc=\"white\", alpha=0.5, boxstyle='round,pad=0.1', lw=0)))\n",
    "            adjust_text(texts, force_text=2.0)\n",
    "        # Check if the user wants these labeled\n",
    "        if self.label_big_sig and annotate:\n",
    "            # If they do have a limit on the number of ones we show (i.e. we don't want 10000 gene names...)\n",
    "            max_values = -1 * self.max_labels\n",
    "            if len(y) < self.max_labels:\n",
    "                max_values = -1 * (len(y) - 1)\n",
    "            most_sig_idxs = np.argpartition(y, max_values)[max_values:]\n",
    "            labels = self.df[self.label_column].values[idxs][most_sig_idxs]\n",
    "            x = x[most_sig_idxs]\n",
    "            y = y[most_sig_idxs]\n",
    "            # We only label the ones with the max log fc\n",
    "            for i, name in enumerate(labels):\n",
    "                name = (' ').join(name.split('_')[1:]) #Format nicer\n",
    "                fig.annotate(name, (x[i], y[i]),\n",
    "                             xytext=(0, 10),\n",
    "                             textcoords='offset points', ha='center', va='bottom',\n",
    "                             bbox=dict(boxstyle='round,pad=0.25', \n",
    "                                       fc=colour, alpha=0.2)\n",
    "                             )\n",
    "        return ax\n",
    "\n",
    "    def plot(self):\n",
    "        \"\"\"\n",
    "        For annotation styling see: https://matplotlib.org/api/pyplot_api.html#matplotlib.pyplot.annotate\n",
    "        Returns\n",
    "        -------\n",
    "\n",
    "        \"\"\"\n",
    "        # if offset is not given, make the offset the smallest value in the dataset\n",
    "        if not self.offset:\n",
    "            vals = self.df[self.p_val].values\n",
    "            self.offset = np.min(vals[np.nonzero(vals)])\n",
    "            self.u.warn_p(['No offset was provided, setting offset to be smallest value recorded in dataset: ',\n",
    "                           self.offset])\n",
    "\n",
    "        # x axis has log_fc, first only plot the values < cutoff\n",
    "        x = self.df[self.log_fc].values\n",
    "        y = -1 * np.log10(self.df[self.p_val].values + self.offset)\n",
    "\n",
    "        log_fc_np = self.df[self.log_fc].values\n",
    "        p_val_np = self.df[self.p_val].values\n",
    "\n",
    "        if self.invert:\n",
    "            x = -1 * np.log10(self.df[self.p_val].values + self.offset)\n",
    "            y = self.df[self.log_fc].values\n",
    "        sig_small_pos_logfc = np.where((p_val_np <= self.p_val_cutoff) & (np.abs(log_fc_np) < self.log_fc_cuttoff)\n",
    "                                       & (log_fc_np > 0))\n",
    "        sig_big_pos_logfc = np.where((p_val_np <= self.p_val_cutoff) & (np.abs(log_fc_np) >= self.log_fc_cuttoff)\n",
    "                                     & (log_fc_np > 0))\n",
    "\n",
    "        sig_small_neg_logfc = np.where((p_val_np <= self.p_val_cutoff) & (np.abs(log_fc_np) < self.log_fc_cuttoff)\n",
    "                                       & (log_fc_np <= 0))\n",
    "        sig_big_neg_logfc = np.where((p_val_np <= self.p_val_cutoff) & (np.abs(log_fc_np) >= self.log_fc_cuttoff)\n",
    "                                     & (log_fc_np <= 0))\n",
    "\n",
    "        # Plot the points\n",
    "        fig, ax = plt.subplots(figsize=self.figsize)\n",
    "        self.add_scatter_and_annotate(ax, x, y, self.colours['sig_small-pos-logFC'], sig_small_pos_logfc)\n",
    "        self.add_scatter_and_annotate(ax, x, y, self.colours['sig_big-pos-logFC'], \n",
    "                                      sig_big_pos_logfc, annotate=True, s=40)\n",
    "\n",
    "        # Negative\n",
    "        self.add_scatter_and_annotate(ax, x, y, self.colours['sig_small-neg-logFC'], sig_small_neg_logfc)\n",
    "        self.add_scatter_and_annotate(ax, x, y, self.colours['sig_big-neg-logFC'], sig_big_neg_logfc, annotate=True, s=40)\n",
    "        self.add_labels()\n",
    "        ax.tick_params(labelsize=6)\n",
    "        ax.tick_params(direction='out', length=2, width=0.5)\n",
    "        ax.spines['bottom'].set_linewidth(0.5)\n",
    "        ax.spines['top'].set_linewidth(0)\n",
    "        ax.spines['left'].set_linewidth(0.5)\n",
    "        ax.spines['right'].set_linewidth(0)\n",
    "        ax.tick_params(labelsize=6)\n",
    "        ax.tick_params(axis='x', which='major', pad=0)\n",
    "        ax.tick_params(axis='y', which='major', pad=0)\n",
    "        return ax\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for f in os.listdir(fig_dir):\n",
    "    if 'GSEA' in f and 'KEGG' in f and 'padj' in f:\n",
    "        \n",
    "        gsea_out = pd.read_csv(fig_dir + f)\n",
    "        gsea_out = gsea_out.fillna(1.0)\n",
    "        if len(gsea_out) > 1:\n",
    "            volcanoplot = Volcanoplot(gsea_out, 'NES', 'padj', 'pathway', \n",
    "                                          f.split(\"_\")[2], 'NES', '-log10(p adj)', \n",
    "                                          p_val_cutoff=1.0, max_labels=5,\n",
    "                                          label_big_sig=True, log_fc_cuttoff=1.5, figsize=(1.5,1.5))\n",
    "            sns.set_style(\"ticks\")\n",
    "            volcanoplot.plot()\n",
    "            save_fig(f'Volcano_gsea_{f.split(\"_\")[2]}')\n",
    "            plt.show()\n",
    "        else:\n",
    "            print(\"Nothing significant for: \", f)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for f in os.listdir(fig_dir):\n",
    "    if 'GSEA' in f and 'Go' in f and 'svg' not in f:\n",
    "        gsea_out = pd.read_csv(fig_dir + f)\n",
    "        gsea_out = gsea_out.fillna(1.0)\n",
    "        volcanoplot = Volcanoplot(gsea_out, 'NES', 'padj', 'pathway', \n",
    "                                      f.split(\"_\")[2], 'NES', '-log10(p adj)', \n",
    "                                      p_val_cutoff=1.0, max_labels=5,\n",
    "                                      label_big_sig=True, log_fc_cuttoff=1.5, figsize=(2,2))\n",
    "        sns.set_style(\"ticks\")\n",
    "        volcanoplot.plot()\n",
    "        save_fig(f'Volcano_gsea_GO_{f.split(\"_\")[1]}_{f.split(\"_\")[2]}')\n",
    "        plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "metadata": {
     "collapsed": false
    },
    "source": []
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
