{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1a582222",
   "metadata": {},
   "source": [
    "## Here we look at what the effect is of integrating all the marks\n",
    "\n",
    "We want to see if we can pick out specific genes based on the latent space"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91ec1a4d",
   "metadata": {},
   "source": [
    "## Read in from here\n",
    "\n",
    "\n",
    "We want to get the weightings for different classes from this point onwards!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e6f229a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_all = pd.read_csv('../data/df_all.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ea759c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>VAE0</th>\n",
       "      <th>VAE1</th>\n",
       "      <th>VAE2</th>\n",
       "      <th>Log2FC FB</th>\n",
       "      <th>Log2FC MB</th>\n",
       "      <th>Log2FC HB</th>\n",
       "      <th>Log2FC SC</th>\n",
       "      <th>GroupLabel</th>\n",
       "      <th>H3K27me3</th>\n",
       "      <th>...</th>\n",
       "      <th>stomach_16.5-days_embryonic_H3K4me1_ChIP-seq_ENCFF814BNR_width</th>\n",
       "      <th>stomach_16.5-days_embryonic_H3K4me1_ChIP-seq_ENCFF814BNR_signal</th>\n",
       "      <th>stomach_16.5-days_embryonic_H3K4me2_ChIP-seq_ENCFF501CJA_width</th>\n",
       "      <th>stomach_16.5-days_embryonic_H3K4me2_ChIP-seq_ENCFF501CJA_signal</th>\n",
       "      <th>stomach_16.5-days_embryonic_H3K4me3_ChIP-seq_ENCFF569KWB_width</th>\n",
       "      <th>stomach_16.5-days_embryonic_H3K4me3_ChIP-seq_ENCFF569KWB_signal</th>\n",
       "      <th>stomach_16.5-days_embryonic_H3K9ac_ChIP-seq_ENCFF068FWP_width</th>\n",
       "      <th>stomach_16.5-days_embryonic_H3K9ac_ChIP-seq_ENCFF068FWP_signal</th>\n",
       "      <th>stomach_16.5-days_embryonic_H3K9me3_ChIP-seq_ENCFF544RGQ_width</th>\n",
       "      <th>stomach_16.5-days_embryonic_H3K9me3_ChIP-seq_ENCFF544RGQ_signal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>0.260962</td>\n",
       "      <td>0.369031</td>\n",
       "      <td>-1.299443</td>\n",
       "      <td>0.436612</td>\n",
       "      <td>-0.141153</td>\n",
       "      <td>-0.246419</td>\n",
       "      <td>-0.095730</td>\n",
       "      <td>Partly affected unmarked</td>\n",
       "      <td>2.82776</td>\n",
       "      <td>...</td>\n",
       "      <td>459.0</td>\n",
       "      <td>4.17547</td>\n",
       "      <td>2522.0</td>\n",
       "      <td>32.56543</td>\n",
       "      <td>2456.0</td>\n",
       "      <td>37.44113</td>\n",
       "      <td>1852.0</td>\n",
       "      <td>6.81303</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sox17</td>\n",
       "      <td>0.407478</td>\n",
       "      <td>-0.646836</td>\n",
       "      <td>0.337369</td>\n",
       "      <td>0.092069</td>\n",
       "      <td>0.278734</td>\n",
       "      <td>0.488633</td>\n",
       "      <td>0.088057</td>\n",
       "      <td>Partly affected marked</td>\n",
       "      <td>5.30733</td>\n",
       "      <td>...</td>\n",
       "      <td>702.0</td>\n",
       "      <td>4.40404</td>\n",
       "      <td>1327.0</td>\n",
       "      <td>17.05396</td>\n",
       "      <td>1257.0</td>\n",
       "      <td>14.96443</td>\n",
       "      <td>507.0</td>\n",
       "      <td>4.75484</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>1.325021</td>\n",
       "      <td>1.099265</td>\n",
       "      <td>-0.278983</td>\n",
       "      <td>-0.551568</td>\n",
       "      <td>-0.132808</td>\n",
       "      <td>-0.126111</td>\n",
       "      <td>0.041484</td>\n",
       "      <td>Partly affected unmarked</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>479.0</td>\n",
       "      <td>2.85033</td>\n",
       "      <td>1334.0</td>\n",
       "      <td>32.48818</td>\n",
       "      <td>2048.0</td>\n",
       "      <td>41.08784</td>\n",
       "      <td>1396.0</td>\n",
       "      <td>18.10043</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tcea1</td>\n",
       "      <td>1.577630</td>\n",
       "      <td>1.028430</td>\n",
       "      <td>-0.621913</td>\n",
       "      <td>-0.289489</td>\n",
       "      <td>-0.151392</td>\n",
       "      <td>-0.142735</td>\n",
       "      <td>0.015326</td>\n",
       "      <td>Partly affected unmarked</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1381.0</td>\n",
       "      <td>12.89607</td>\n",
       "      <td>2526.0</td>\n",
       "      <td>55.09949</td>\n",
       "      <td>1176.0</td>\n",
       "      <td>12.09817</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rgs20</td>\n",
       "      <td>0.316957</td>\n",
       "      <td>1.250120</td>\n",
       "      <td>-0.743550</td>\n",
       "      <td>-0.290937</td>\n",
       "      <td>-0.094171</td>\n",
       "      <td>-0.466210</td>\n",
       "      <td>0.249359</td>\n",
       "      <td>Partly affected unmarked</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12792</th>\n",
       "      <td>Frmpd4</td>\n",
       "      <td>-0.385455</td>\n",
       "      <td>0.081109</td>\n",
       "      <td>-1.109779</td>\n",
       "      <td>0.392005</td>\n",
       "      <td>-0.463031</td>\n",
       "      <td>-0.259205</td>\n",
       "      <td>-0.120672</td>\n",
       "      <td>Partly affected unmarked</td>\n",
       "      <td>3.20311</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12793</th>\n",
       "      <td>Msl3</td>\n",
       "      <td>1.032435</td>\n",
       "      <td>0.199148</td>\n",
       "      <td>-0.023354</td>\n",
       "      <td>-0.406618</td>\n",
       "      <td>-0.192482</td>\n",
       "      <td>-0.072598</td>\n",
       "      <td>-0.112576</td>\n",
       "      <td>Partly affected unmarked</td>\n",
       "      <td>2.93122</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2402.0</td>\n",
       "      <td>21.82412</td>\n",
       "      <td>3127.0</td>\n",
       "      <td>33.87626</td>\n",
       "      <td>2257.0</td>\n",
       "      <td>13.40337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12794</th>\n",
       "      <td>Arhgap6</td>\n",
       "      <td>-0.129757</td>\n",
       "      <td>-0.457979</td>\n",
       "      <td>0.349736</td>\n",
       "      <td>0.978575</td>\n",
       "      <td>0.381812</td>\n",
       "      <td>0.111678</td>\n",
       "      <td>-0.803318</td>\n",
       "      <td>Partly affected marked</td>\n",
       "      <td>3.56165</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2878.0</td>\n",
       "      <td>23.78164</td>\n",
       "      <td>2420.0</td>\n",
       "      <td>18.13809</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>5.96359</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12795</th>\n",
       "      <td>Hccs</td>\n",
       "      <td>1.083603</td>\n",
       "      <td>0.476223</td>\n",
       "      <td>-0.351159</td>\n",
       "      <td>-0.236171</td>\n",
       "      <td>-0.180150</td>\n",
       "      <td>-0.213555</td>\n",
       "      <td>-0.096767</td>\n",
       "      <td>Partly affected unmarked</td>\n",
       "      <td>3.30158</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1698.0</td>\n",
       "      <td>17.11209</td>\n",
       "      <td>2270.0</td>\n",
       "      <td>30.84611</td>\n",
       "      <td>1158.0</td>\n",
       "      <td>12.58762</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12796</th>\n",
       "      <td>Mid1</td>\n",
       "      <td>0.877479</td>\n",
       "      <td>0.196111</td>\n",
       "      <td>0.664959</td>\n",
       "      <td>-1.016361</td>\n",
       "      <td>-0.954067</td>\n",
       "      <td>-0.388715</td>\n",
       "      <td>-0.516899</td>\n",
       "      <td>Partly affected unmarked</td>\n",
       "      <td>3.60893</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1298.0</td>\n",
       "      <td>7.41834</td>\n",
       "      <td>1159.0</td>\n",
       "      <td>18.90378</td>\n",
       "      <td>853.0</td>\n",
       "      <td>6.55106</td>\n",
       "      <td>1094.0</td>\n",
       "      <td>2.23479</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12797 rows Ã— 1204 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      external_gene_name      VAE0      VAE1      VAE2  Log2FC FB  Log2FC MB  \\\n",
       "0                   Xkr4  0.260962  0.369031 -1.299443   0.436612  -0.141153   \n",
       "1                  Sox17  0.407478 -0.646836  0.337369   0.092069   0.278734   \n",
       "2                 Mrpl15  1.325021  1.099265 -0.278983  -0.551568  -0.132808   \n",
       "3                  Tcea1  1.577630  1.028430 -0.621913  -0.289489  -0.151392   \n",
       "4                  Rgs20  0.316957  1.250120 -0.743550  -0.290937  -0.094171   \n",
       "...                  ...       ...       ...       ...        ...        ...   \n",
       "12792             Frmpd4 -0.385455  0.081109 -1.109779   0.392005  -0.463031   \n",
       "12793               Msl3  1.032435  0.199148 -0.023354  -0.406618  -0.192482   \n",
       "12794            Arhgap6 -0.129757 -0.457979  0.349736   0.978575   0.381812   \n",
       "12795               Hccs  1.083603  0.476223 -0.351159  -0.236171  -0.180150   \n",
       "12796               Mid1  0.877479  0.196111  0.664959  -1.016361  -0.954067   \n",
       "\n",
       "       Log2FC HB  Log2FC SC                GroupLabel  H3K27me3  ...  \\\n",
       "0      -0.246419  -0.095730  Partly affected unmarked   2.82776  ...   \n",
       "1       0.488633   0.088057    Partly affected marked   5.30733  ...   \n",
       "2      -0.126111   0.041484  Partly affected unmarked   0.00000  ...   \n",
       "3      -0.142735   0.015326  Partly affected unmarked   0.00000  ...   \n",
       "4      -0.466210   0.249359  Partly affected unmarked   0.00000  ...   \n",
       "...          ...        ...                       ...       ...  ...   \n",
       "12792  -0.259205  -0.120672  Partly affected unmarked   3.20311  ...   \n",
       "12793  -0.072598  -0.112576  Partly affected unmarked   2.93122  ...   \n",
       "12794   0.111678  -0.803318    Partly affected marked   3.56165  ...   \n",
       "12795  -0.213555  -0.096767  Partly affected unmarked   3.30158  ...   \n",
       "12796  -0.388715  -0.516899  Partly affected unmarked   3.60893  ...   \n",
       "\n",
       "       stomach_16.5-days_embryonic_H3K4me1_ChIP-seq_ENCFF814BNR_width  \\\n",
       "0                                                  459.0                \n",
       "1                                                  702.0                \n",
       "2                                                  479.0                \n",
       "3                                                    NaN                \n",
       "4                                                    NaN                \n",
       "...                                                  ...                \n",
       "12792                                                NaN                \n",
       "12793                                                NaN                \n",
       "12794                                                NaN                \n",
       "12795                                                NaN                \n",
       "12796                                                NaN                \n",
       "\n",
       "       stomach_16.5-days_embryonic_H3K4me1_ChIP-seq_ENCFF814BNR_signal  \\\n",
       "0                                                4.17547                 \n",
       "1                                                4.40404                 \n",
       "2                                                2.85033                 \n",
       "3                                                    NaN                 \n",
       "4                                                    NaN                 \n",
       "...                                                  ...                 \n",
       "12792                                                NaN                 \n",
       "12793                                                NaN                 \n",
       "12794                                                NaN                 \n",
       "12795                                                NaN                 \n",
       "12796                                                NaN                 \n",
       "\n",
       "       stomach_16.5-days_embryonic_H3K4me2_ChIP-seq_ENCFF501CJA_width  \\\n",
       "0                                                 2522.0                \n",
       "1                                                 1327.0                \n",
       "2                                                 1334.0                \n",
       "3                                                 1381.0                \n",
       "4                                                    NaN                \n",
       "...                                                  ...                \n",
       "12792                                                NaN                \n",
       "12793                                             2402.0                \n",
       "12794                                             2878.0                \n",
       "12795                                             1698.0                \n",
       "12796                                             1298.0                \n",
       "\n",
       "       stomach_16.5-days_embryonic_H3K4me2_ChIP-seq_ENCFF501CJA_signal  \\\n",
       "0                                               32.56543                 \n",
       "1                                               17.05396                 \n",
       "2                                               32.48818                 \n",
       "3                                               12.89607                 \n",
       "4                                                    NaN                 \n",
       "...                                                  ...                 \n",
       "12792                                                NaN                 \n",
       "12793                                           21.82412                 \n",
       "12794                                           23.78164                 \n",
       "12795                                           17.11209                 \n",
       "12796                                            7.41834                 \n",
       "\n",
       "       stomach_16.5-days_embryonic_H3K4me3_ChIP-seq_ENCFF569KWB_width  \\\n",
       "0                                                 2456.0                \n",
       "1                                                 1257.0                \n",
       "2                                                 2048.0                \n",
       "3                                                 2526.0                \n",
       "4                                                    NaN                \n",
       "...                                                  ...                \n",
       "12792                                                NaN                \n",
       "12793                                             3127.0                \n",
       "12794                                             2420.0                \n",
       "12795                                             2270.0                \n",
       "12796                                             1159.0                \n",
       "\n",
       "       stomach_16.5-days_embryonic_H3K4me3_ChIP-seq_ENCFF569KWB_signal  \\\n",
       "0                                               37.44113                 \n",
       "1                                               14.96443                 \n",
       "2                                               41.08784                 \n",
       "3                                               55.09949                 \n",
       "4                                                    NaN                 \n",
       "...                                                  ...                 \n",
       "12792                                                NaN                 \n",
       "12793                                           33.87626                 \n",
       "12794                                           18.13809                 \n",
       "12795                                           30.84611                 \n",
       "12796                                           18.90378                 \n",
       "\n",
       "       stomach_16.5-days_embryonic_H3K9ac_ChIP-seq_ENCFF068FWP_width  \\\n",
       "0                                                 1852.0               \n",
       "1                                                  507.0               \n",
       "2                                                 1396.0               \n",
       "3                                                 1176.0               \n",
       "4                                                    NaN               \n",
       "...                                                  ...               \n",
       "12792                                                NaN               \n",
       "12793                                             2257.0               \n",
       "12794                                             2004.0               \n",
       "12795                                             1158.0               \n",
       "12796                                              853.0               \n",
       "\n",
       "       stomach_16.5-days_embryonic_H3K9ac_ChIP-seq_ENCFF068FWP_signal  \\\n",
       "0                                                6.81303                \n",
       "1                                                4.75484                \n",
       "2                                               18.10043                \n",
       "3                                               12.09817                \n",
       "4                                                    NaN                \n",
       "...                                                  ...                \n",
       "12792                                                NaN                \n",
       "12793                                           13.40337                \n",
       "12794                                            5.96359                \n",
       "12795                                           12.58762                \n",
       "12796                                            6.55106                \n",
       "\n",
       "       stomach_16.5-days_embryonic_H3K9me3_ChIP-seq_ENCFF544RGQ_width  \\\n",
       "0                                                    NaN                \n",
       "1                                                    NaN                \n",
       "2                                                    NaN                \n",
       "3                                                    NaN                \n",
       "4                                                    NaN                \n",
       "...                                                  ...                \n",
       "12792                                                NaN                \n",
       "12793                                                NaN                \n",
       "12794                                                NaN                \n",
       "12795                                                NaN                \n",
       "12796                                             1094.0                \n",
       "\n",
       "       stomach_16.5-days_embryonic_H3K9me3_ChIP-seq_ENCFF544RGQ_signal  \n",
       "0                                                    NaN                \n",
       "1                                                    NaN                \n",
       "2                                                    NaN                \n",
       "3                                                    NaN                \n",
       "4                                                    NaN                \n",
       "...                                                  ...                \n",
       "12792                                                NaN                \n",
       "12793                                                NaN                \n",
       "12794                                                NaN                \n",
       "12795                                                NaN                \n",
       "12796                                            2.23479                \n",
       "\n",
       "[12797 rows x 1204 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "49c1ec62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Unmarked proliferation                                             111\n",
       "Marked Anterior                                                     94\n",
       "Development                                                         93\n",
       "Marked Posterior                                                    44\n",
       "Unmarked proliferation Unmarked immune response                     34\n",
       "Marked Posterior Development                                        30\n",
       "Unmarked proliferation Development                                  21\n",
       "Unmarked immune response Development                                12\n",
       "Unmarked proliferation Marked Anterior                              12\n",
       "Unmarked immune response Marked Anterior                             7\n",
       "Marked Posterior Marked Anterior                                     6\n",
       "Unmarked proliferation Unmarked immune response Marked Anterior      2\n",
       "Name: GroupLabel, dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pos_subset = df_all[~df_all['GroupLabel'].isin(['Unmarked immune response', 'Partly affected unmarked', 'Partly affected marked', 'Consistently affected marked', 'Consistently affected unmarked'])]\n",
    "df_pos_subset['GroupLabel'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "ad137af7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Consistently affected marked      437\n",
      "Consistently affected unmarked    301\n",
      "Name: GroupLabel, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "df_neg_subset = df_all[df_all['GroupLabel'].isin(['Consistently affected marked', 'Consistently affected unmarked'])]\n",
    "print(df_neg_subset['GroupLabel'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "c96bc141",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_tfs = pd.read_csv(f'../data/doro_mouse.csv')\n",
    "mouse_tfs = mouse_tfs[mouse_tfs['confidence'].isin(['A'])]\n",
    "tfs = set(mouse_tfs['tf'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "a516131e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_training = df_all[['external_gene_name'] + [c for c in df_all.columns if 'H3' in c and '_' in c and 'signal' in c]]\n",
    "df_training = df_training.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "f8bea14c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pos = df_training[df_training['external_gene_name'].isin(df_pos_subset['external_gene_name'].values)]\n",
    "df_pos.set_index('external_gene_name', inplace=True)\n",
    "df_pos = df_pos.T.sample(frac=1) # Randomly shuffle the data\n",
    "\n",
    "df_neg = df_training[df_training['external_gene_name'].isin(df_neg_subset['external_gene_name'].values)]\n",
    "df_neg.set_index('external_gene_name', inplace=True)\n",
    "df_neg = df_neg.T.sample(frac=1)\n",
    "\n",
    "df_tfs = df_training[df_training['external_gene_name'].isin(tfs)]\n",
    "df_tfs.set_index('external_gene_name', inplace=True)\n",
    "df_tfs = df_tfs.T.sample(frac=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "079d92ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(588, 466)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pos.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "48f717c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'8-weeks', '11.5-days', '16.5-days', '15.5-days', '0-days', '10.5-days', '12.5-days', '13.5-days', '14.5-days'}\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                   VAE identified genes on time:\t(588, 466)\t                    \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.4191086691086692\n",
      "0.1009920634920635\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                    TFs identified genes on time:\t(588, 84)\t                    \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.2709435626102292\n",
      "0.13086419753086417\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                Consistently affected genes on time:\t(588, 738)\t                \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.29360269360269364\n",
      "0.12222222222222223\n"
     ]
    }
   ],
   "source": [
    "from scivae import Validate\n",
    "from sciutil import SciUtil\n",
    "\n",
    "label = 'time'\n",
    "label_i = 1\n",
    "df_pos_labels = [i.split('_')[label_i] for i in df_pos.index]\n",
    "df_neg_labels = [i.split('_')[label_i] for i in df_neg.index]\n",
    "df_tfs_labels = [i.split('_')[label_i] for i in df_tfs.index]\n",
    "print(set(df_tfs_labels))\n",
    "\n",
    "\n",
    "u = SciUtil()\n",
    "u.dp([f'VAE identified genes on {label}:', df_pos.shape])\n",
    "vd = Validate(df_pos.values, df_pos_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'TFs identified genes on {label}:', df_tfs.shape])\n",
    "vd = Validate(df_tfs.values, df_tfs_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'Consistently affected genes on {label}:', df_neg.shape])\n",
    "vd = Validate(df_neg.values, df_neg_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "c797d6c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'embryonic', 'adult', 'postnatal'}\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                   VAE identified genes on time:\t(588, 466)\t                    \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.32882882882882886\n",
      "0.3333333333333333\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                    TFs identified genes on time:\t(588, 84)\t                    \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.5\n",
      "0.3333333333333333\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                Consistently affected genes on time:\t(588, 738)\t                \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.35555555555555557\n",
      "0.3333333333333333\n"
     ]
    }
   ],
   "source": [
    "from scivae import Validate\n",
    "from sciutil import SciUtil\n",
    "\n",
    "label = 'time'\n",
    "label_i = 2\n",
    "df_pos_labels = [i.split('_')[label_i] for i in df_pos.index]\n",
    "df_neg_labels = [i.split('_')[label_i] for i in df_neg.index]\n",
    "df_tfs_labels = [i.split('_')[label_i] for i in df_tfs.index]\n",
    "print(set(df_tfs_labels))\n",
    "\n",
    "\n",
    "u = SciUtil()\n",
    "u.dp([f'VAE identified genes on {label}:', df_pos.shape])\n",
    "vd = Validate(df_pos.values, df_pos_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'TFs identified genes on {label}:', df_tfs.shape])\n",
    "vd = Validate(df_tfs.values, df_tfs_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'Consistently affected genes on {label}:', df_neg.shape])\n",
    "vd = Validate(df_neg.values, df_neg_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "958f8fd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'embryonic-facial-prominence', 'neural-tube', 'hindbrain', 'intestine', 'lung', 'kidney', 'forebrain', 'liver', 'midbrain', 'limb', 'stomach', 'heart'}\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                        VAE identified genes on tissue:\t                        \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.6137784601019894\n",
      "0.4600184945773181\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                        TFs identified genes on tissue:\t                        \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.24431216931216929\n",
      "0.1597222222222222\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                    Consistently affected genes on tissue:\t                     \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.31620370370370365\n",
      "0.1529040404040404\n"
     ]
    }
   ],
   "source": [
    "\n",
    "label = 'tissue'\n",
    "label_i = 0\n",
    "df_pos_labels = [i.split('_')[label_i] for i in df_pos.index]\n",
    "df_neg_labels = [i.split('_')[label_i] for i in df_neg.index]\n",
    "df_tfs_labels = [i.split('_')[label_i] for i in df_tfs.index]\n",
    "print(set(df_tfs_labels))\n",
    "\n",
    "\n",
    "u = SciUtil()\n",
    "u.dp([f'VAE identified genes on {label}:'])\n",
    "vd = Validate(df_pos.values, df_pos_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'TFs identified genes on {label}:'])\n",
    "vd = Validate(df_tfs.values, df_tfs_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'Consistently affected genes on {label}:'])\n",
    "vd = Validate(df_neg.values, df_neg_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "70ceaf8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'H3K9ac', 'H3K27me3', 'H3K4me1', 'H3K4me2', 'H3K9me3', 'H3K36me3', 'H3K27ac', 'H3K4me3'}\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                         VAE identified genes on mark:\t                         \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.971875\n",
      "0.9782986111111112\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                         TFs identified genes on mark:\t                         \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.979020979020979\n",
      "0.9903846153846154\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                     Consistently affected genes on mark:\t                      \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.9696691176470589\n",
      "0.9527007918552036\n"
     ]
    }
   ],
   "source": [
    "\n",
    "label = 'mark'\n",
    "label_i = 3\n",
    "df_pos_labels = [i.split('_')[label_i] for i in df_pos.index]\n",
    "df_neg_labels = [i.split('_')[label_i] for i in df_neg.index]\n",
    "df_tfs_labels = [i.split('_')[label_i] for i in df_tfs.index]\n",
    "print(set(df_tfs_labels))\n",
    "\n",
    "\n",
    "u = SciUtil()\n",
    "u.dp([f'VAE identified genes on {label}:'])\n",
    "vd = Validate(df_pos.values, df_pos_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'TFs identified genes on {label}:'])\n",
    "vd = Validate(df_tfs.values, df_tfs_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'Consistently affected genes on {label}:'])\n",
    "vd = Validate(df_neg.values, df_neg_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6147882",
   "metadata": {},
   "source": [
    "## Look at human brain expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b86f7bb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a map from mouse to human\n",
    "map_df = pd.read_csv('../data/mm10_orthologs_hsapiens_gene_ensembl-GRCh38.p13.csv')\n",
    "map_df = map_df.dropna()\n",
    "plt.hist(map_df['mmusculus_homolog_perc_id_r1'].values)\n",
    "plt.show()\n",
    "\n",
    "# Keep ones with over 80%\n",
    "mapping = map_df[map_df['mmusculus_homolog_perc_id_r1'] > 80]\n",
    "mapping = dict(zip(mapping.mmusculus_homolog_ensembl_gene, mapping.ensembl_gene_id))\n",
    "\n",
    "# Now map each of the sets to the human version\n",
    "gene_name_to_ensembl = pd.read_csv('../data/mmusculus_gene_ensembl-GRCm39.csv')\n",
    "gene_name_to_ensembl = dict(zip(gene_name_to_ensembl['external_gene_name'].values, gene_name_to_ensembl['ensembl_gene_id'].values))\n",
    "\n",
    "# For each of them make the map\n",
    "df_neg_human = [gene_name_to_ensembl.get(g) for g in df_neg_subset.external_gene_name.values]\n",
    "df_neg_human = [mapping.get(g) for g in df_neg_human if g is not None]\n",
    "df_neg_human = [g for g in df_neg_human if g is not None]\n",
    "\n",
    "df_pos_human = [gene_name_to_ensembl.get(g) for g in df_pos_subset.external_gene_name.values]\n",
    "df_pos_human = [mapping.get(g) for g in df_pos_human if g is not None]\n",
    "df_pos_human = [g for g in df_pos_human if g is not None]\n",
    "\n",
    "print(len(df_neg_human), len(df_pos_human))\n",
    "\n",
    "# For TFs we just use human TFs from dorothea\n",
    "human_tfs = pd.read_csv('../data/dorothea_hs_ABCD.csv')\n",
    "human_tfs = human_tfs[human_tfs['confidence'].isin(['A'])]\n",
    "human_tfs = set(human_tfs['tf'].values)\n",
    "print(len(human_tfs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "f34f998b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/gq/6ljhmvm1713fykdjqbl188pm0000gn/T/ipykernel_13767/1843951305.py:35: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  human_emb['ensembl_gene_id'] = human_rm['ensembl_gene_id'].values\n",
      "/var/folders/gq/6ljhmvm1713fykdjqbl188pm0000gn/T/ipykernel_13767/1843951305.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  human_emb['gene_symbol'] = human_rm['gene_symbol'].values\n",
      "/var/folders/gq/6ljhmvm1713fykdjqbl188pm0000gn/T/ipykernel_13767/1843951305.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  human_emb['entrez_id'] = human_rm['entrez_id'].values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ocx_8-pcw_W1_M_13058\n",
      "M1C-S1C_8-pcw_W1_M_13058\n",
      "AMY_8-pcw_W1_M_13058\n",
      "MGE_8-pcw_W1_M_13058\n",
      "STC_8-pcw_W1_M_13058\n",
      "URL_8-pcw_W1_M_13058\n",
      "CGE_8-pcw_W1_M_13058\n",
      "DTH_8-pcw_W1_M_13058\n",
      "MFC_8-pcw_W1_M_13058\n",
      "DFC_8-pcw_W1_M_13058\n",
      "OFC_8-pcw_W1_M_13058\n",
      "LGE_8-pcw_W1_M_13058\n",
      "ITC_8-pcw_W1_M_13058\n",
      "HIP_8-pcw_W1_M_13058\n",
      "VFC_8-pcw_W1_M_13058\n",
      "PCx_8-pcw_W1_M_13058\n",
      "DFC_9-pcw_W1_M_12833\n",
      "MFC_9-pcw_W1_M_12833\n",
      "AMY_9-pcw_W1_M_12833\n",
      "DTH_9-pcw_W1_M_12833\n",
      "URL_9-pcw_W1_M_12833\n",
      "LGE_9-pcw_W1_M_12833\n",
      "M1C-S1C_9-pcw_W1_M_12833\n",
      "MGE_9-pcw_W1_M_12833\n",
      "TCx_9-pcw_W1_M_12833\n",
      "HIP_9-pcw_W1_M_12833\n",
      "Ocx_9-pcw_W1_M_12833\n",
      "CGE_9-pcw_W1_M_12833\n",
      "OFC_9-pcw_W1_M_12833\n",
      "PCx_9-pcw_W1_M_12833\n",
      "HIP_12-pcw_W2_F_12835\n",
      "DTH_12-pcw_W2_F_12835\n",
      "AMY_12-pcw_W2_F_12835\n",
      "A1C_12-pcw_W2_F_12835\n",
      "V1C_12-pcw_W2_F_12835\n",
      "MFC_12-pcw_W2_F_12835\n",
      "STR_12-pcw_W2_F_12835\n",
      "M1C_12-pcw_W2_F_12835\n",
      "DFC_12-pcw_W2_F_12835\n",
      "ITC_12-pcw_W2_F_12835\n",
      "VFC_12-pcw_W2_F_12835\n",
      "STC_12-pcw_W2_F_12835\n",
      "IPC_12-pcw_W2_F_12835\n",
      "OFC_12-pcw_W2_F_12835\n",
      "S1C_12-pcw_W2_F_12835\n",
      "STC_12-pcw_W2_F_12960\n",
      "V1C_12-pcw_W2_F_12960\n",
      "A1C_12-pcw_W2_F_12960\n",
      "CB_12-pcw_W2_F_12960\n",
      "ITC_12-pcw_W2_F_12960\n",
      "IPC_12-pcw_W2_F_12960\n",
      "S1C_12-pcw_W2_F_12960\n",
      "DTH_12-pcw_W2_F_12960\n",
      "HIP_12-pcw_W2_F_12960\n",
      "DFC_12-pcw_W2_F_12960\n",
      "OFC_12-pcw_W2_F_12960\n",
      "M1C_12-pcw_W2_F_12960\n",
      "VFC_12-pcw_W2_F_12960\n",
      "AMY_12-pcw_W2_F_12960\n",
      "STR_12-pcw_W2_F_12960\n",
      "AMY_12-pcw_W2_F_13060\n",
      "M1C_12-pcw_W2_F_13060\n",
      "ITC_12-pcw_W2_F_13060\n",
      "VFC_12-pcw_W2_F_13060\n",
      "STR_12-pcw_W2_F_13060\n",
      "CBC_12-pcw_W2_F_13060\n",
      "DTH_12-pcw_W2_F_13060\n",
      "DFC_12-pcw_W2_F_13060\n",
      "OFC_12-pcw_W2_F_13060\n",
      "S1C_12-pcw_W2_F_13060\n",
      "V1C_12-pcw_W2_F_13060\n",
      "MFC_12-pcw_W2_F_13060\n",
      "HIP_12-pcw_W2_F_13060\n",
      "IPC_12-pcw_W2_F_13060\n",
      "A1C_12-pcw_W2_F_13060\n",
      "STR_13-pcw_W2_M_12820\n",
      "M1C_13-pcw_W2_M_12820\n",
      "AMY_13-pcw_W2_M_12820\n",
      "S1C_13-pcw_W2_M_12820\n",
      "V1C_13-pcw_W2_M_12820\n",
      "A1C_13-pcw_W2_M_12820\n",
      "HIP_13-pcw_W2_M_12820\n",
      "VFC_13-pcw_W2_M_12820\n",
      "MFC_13-pcw_W2_M_12820\n",
      "STC_13-pcw_W2_M_12820\n",
      "OFC_13-pcw_W2_M_12820\n",
      "IPC_13-pcw_W2_M_12820\n",
      "DFC_13-pcw_W2_M_12820\n",
      "ITC_13-pcw_W2_M_12820\n",
      "STR_13-pcw_W2_F_12834\n",
      "S1C_13-pcw_W2_F_12834\n",
      "V1C_13-pcw_W2_F_12834\n",
      "AMY_13-pcw_W2_F_12834\n",
      "IPC_13-pcw_W2_F_12834\n",
      "CB_13-pcw_W2_F_12834\n",
      "DFC_13-pcw_W2_F_12834\n",
      "OFC_13-pcw_W2_F_12834\n",
      "STC_13-pcw_W2_F_12834\n",
      "A1C_13-pcw_W2_F_12834\n",
      "HIP_13-pcw_W2_F_12834\n",
      "M1C_13-pcw_W2_F_12834\n",
      "ITC_13-pcw_W2_F_12834\n",
      "MD_13-pcw_W2_F_12834\n",
      "MFC_13-pcw_W2_F_12834\n",
      "VFC_13-pcw_W2_F_12834\n",
      "S1C_13-pcw_W2_M_12888\n",
      "DFC_13-pcw_W2_M_12888\n",
      "CB_13-pcw_W2_M_12888\n",
      "AMY_13-pcw_W2_M_12888\n",
      "ITC_13-pcw_W2_M_12888\n",
      "HIP_13-pcw_W2_M_12888\n",
      "IPC_13-pcw_W2_M_12888\n",
      "A1C_13-pcw_W2_M_12888\n",
      "VFC_13-pcw_W2_M_12888\n",
      "STR_13-pcw_W2_M_12888\n",
      "OFC_13-pcw_W2_M_12888\n",
      "MFC_13-pcw_W2_M_12888\n",
      "V1C_13-pcw_W2_M_12888\n",
      "M1C_13-pcw_W2_M_12888\n",
      "VFC_16-pcw_W3_M_12287\n",
      "OFC_16-pcw_W3_M_12287\n",
      "STC_16-pcw_W3_M_12287\n",
      "DFC_16-pcw_W3_M_12287\n",
      "IPC_16-pcw_W3_M_12287\n",
      "A1C_16-pcw_W3_M_12287\n",
      "MFC_16-pcw_W3_M_12287\n",
      "STR_16-pcw_W3_M_12287\n",
      "MD_16-pcw_W3_M_12287\n",
      "V1C_16-pcw_W3_M_12287\n",
      "VFC_16-pcw_W3_M_12837\n",
      "STC_16-pcw_W3_M_12837\n",
      "IPC_16-pcw_W3_M_12837\n",
      "A1C_16-pcw_W3_M_12837\n",
      "MD_16-pcw_W3_M_12837\n",
      "AMY_16-pcw_W3_M_12837\n",
      "V1C_16-pcw_W3_M_12837\n",
      "ITC_16-pcw_W3_M_12837\n",
      "CBC_16-pcw_W3_M_12837\n",
      "HIP_16-pcw_W3_M_12837\n",
      "STR_16-pcw_W3_M_12837\n",
      "OFC_16-pcw_W3_M_12837\n",
      "DFC_16-pcw_W3_M_12837\n",
      "M1C_16-pcw_W3_M_12837\n",
      "MFC_16-pcw_W3_M_12837\n",
      "S1C_16-pcw_W3_M_12837\n",
      "HIP_16-pcw_W3_M_12879\n",
      "ITC_16-pcw_W3_M_12879\n",
      "DFC_16-pcw_W3_M_12879\n",
      "IPC_16-pcw_W3_M_12879\n",
      "STC_16-pcw_W3_M_12879\n",
      "V1C_16-pcw_W3_M_12879\n",
      "MD_16-pcw_W3_M_12879\n",
      "STR_16-pcw_W3_M_12879\n",
      "M1C-S1C_16-pcw_W3_M_12879\n",
      "A1C_16-pcw_W3_M_12879\n",
      "VFC_16-pcw_W3_M_12879\n",
      "AMY_16-pcw_W3_M_12879\n",
      "MFC_16-pcw_W3_M_12879\n",
      "STR_17-pcw_W3_F_12880\n",
      "STC_17-pcw_W3_F_12880\n",
      "MD_17-pcw_W3_F_12880\n",
      "A1C_17-pcw_W3_F_12880\n",
      "HIP_17-pcw_W3_F_12880\n",
      "DFC_17-pcw_W3_F_12880\n",
      "VFC_17-pcw_W3_F_12880\n",
      "IPC_17-pcw_W3_F_12880\n",
      "M1C-S1C_17-pcw_W3_F_12880\n",
      "V1C_17-pcw_W3_F_12880\n",
      "CBC_17-pcw_W3_F_12880\n",
      "OFC_17-pcw_W3_F_12880\n",
      "AMY_17-pcw_W3_F_12880\n",
      "MFC_17-pcw_W3_F_12880\n",
      "STC_19-pcw_W4_F_12885\n",
      "IPC_19-pcw_W4_F_12885\n",
      "M1C-S1C_19-pcw_W4_F_12885\n",
      "MFC_19-pcw_W4_F_12885\n",
      "V1C_19-pcw_W4_F_12885\n",
      "HIP_19-pcw_W4_F_12885\n",
      "A1C_19-pcw_W4_F_12885\n",
      "MD_19-pcw_W4_F_12885\n",
      "VFC_19-pcw_W4_F_12885\n",
      "STR_19-pcw_W4_F_12885\n",
      "DFC_19-pcw_W4_F_12885\n",
      "CBC_21-pcw_W4_F_12365\n",
      "ITC_21-pcw_W4_F_12365\n",
      "MFC_21-pcw_W4_M_12886\n",
      "M1C_21-pcw_W4_M_12886\n",
      "STR_21-pcw_W4_M_12886\n",
      "HIP_21-pcw_W4_M_12886\n",
      "STC_21-pcw_W4_M_12886\n",
      "ITC_21-pcw_W4_M_12886\n",
      "AMY_21-pcw_W4_M_12886\n",
      "VFC_21-pcw_W4_M_12886\n",
      "OFC_21-pcw_W4_M_12886\n",
      "V1C_21-pcw_W4_M_12886\n",
      "S1C_21-pcw_W4_M_12886\n",
      "CBC_21-pcw_W4_M_12886\n",
      "IPC_21-pcw_W4_M_12886\n",
      "DFC_21-pcw_W4_M_12886\n",
      "HIP_24-pcw_W4_M_12288\n",
      "AMY_24-pcw_W4_M_12288\n",
      "DFC_24-pcw_W4_M_12288\n",
      "S1C_24-pcw_W4_M_12288\n",
      "MD_24-pcw_W4_M_12288\n",
      "STR_24-pcw_W4_M_12288\n",
      "A1C_24-pcw_W4_M_12288\n",
      "IPC_24-pcw_W4_M_12288\n",
      "OFC_24-pcw_W4_M_12288\n",
      "STC_24-pcw_W4_M_12288\n",
      "ITC_24-pcw_W4_M_12288\n",
      "VFC_24-pcw_W4_M_12288\n",
      "M1C_24-pcw_W4_M_12288\n",
      "MFC_24-pcw_W4_M_12288\n",
      "CBC_24-pcw_W4_M_12288\n",
      "V1C_24-pcw_W4_M_12288\n",
      "A1C_25-pcw_W4_F_12948\n",
      "STC_26-pcw_W4_F_12949\n",
      "V1C_26-pcw_W4_F_12949\n",
      "DFC_26-pcw_W4_F_12949\n",
      "CBC_35-pcw_W4_F_12295\n",
      "VFC_35-pcw_W4_F_12295\n",
      "A1C_37-pcw_W4_M_263195015\n",
      "OFC_37-pcw_W4_M_263195015\n",
      "AMY_37-pcw_W4_M_263195015\n",
      "V1C_37-pcw_W4_M_263195015\n",
      "MFC_37-pcw_W4_M_263195015\n",
      "MD_37-pcw_W4_M_263195015\n",
      "CBC_37-pcw_W4_M_263195015\n",
      "STR_37-pcw_W4_M_263195015\n",
      "IPC_37-pcw_W4_M_263195015\n",
      "ITC_37-pcw_W4_M_263195015\n",
      "S1C_37-pcw_W4_M_263195015\n",
      "DFC_37-pcw_W4_M_263195015\n",
      "STC_37-pcw_W4_M_263195015\n",
      "M1C_37-pcw_W4_M_263195015\n",
      "VFC_37-pcw_W4_M_263195015\n",
      "HIP_37-pcw_W4_M_263195015\n",
      "ensembl_gene_id\n",
      "gene_symbol\n"
     ]
    }
   ],
   "source": [
    "from statannot import add_stat_annotation\n",
    "data_dir = '../data/'\n",
    "r_dir = f'{data_dir}results/deseq2/'\n",
    "supp_dir = f'{data_dir}input/supps/'\n",
    "\n",
    "human_b = pd.read_csv(supp_dir + 'genes_matrix_csv/expression_matrix.csv', header=None)\n",
    "human_rm = pd.read_csv(supp_dir + 'genes_matrix_csv/rows_metadata.csv')\n",
    "human_cm = pd.read_csv(supp_dir + 'genes_matrix_csv/columns_metadata.csv')\n",
    "\n",
    "ages = human_cm['age'].values\n",
    "gender = human_cm['gender'].values\n",
    "regions = human_cm['structure_acronym'].values\n",
    "col_names = ['row-id']\n",
    "donor_id = human_cm['donor_id'].values\n",
    "\n",
    "# Rename columns to early, mid, w1 = e8, 9, w2 = 12, 13, w3 = 16, 17, w4 = > 17\n",
    "for i, r in enumerate(regions):\n",
    "    a = int(ages[i].split(' ')[0])\n",
    "    stage = 0\n",
    "    if a > 17:\n",
    "        stage = 'W4'\n",
    "    elif a > 13:\n",
    "        stage = 'W3'\n",
    "    elif a > 9:\n",
    "        stage = 'W2'\n",
    "    else:\n",
    "        stage = 'W1'\n",
    "    \n",
    "    cn = regions[i] + '_' + ages[i].replace(' ', '-') + '_' + stage + '_' + gender[i]  + '_' + str(donor_id[i])\n",
    "    col_names.append(cn)\n",
    "\n",
    "human_b.columns = col_names\n",
    "human_emb = human_b[[c for c in human_b.columns if 'pcw' in c]]# and ('OFC' in c or 'MD' in c or 'CBC' in c)]]\n",
    "\n",
    "human_emb['ensembl_gene_id'] = human_rm['ensembl_gene_id'].values\n",
    "human_emb['gene_symbol'] = human_rm['gene_symbol'].values\n",
    "human_emb['entrez_id'] = human_rm['entrez_id'].values\n",
    "\n",
    "human_emb_a = pd.DataFrame()\n",
    "human_emb_num = pd.DataFrame()\n",
    "human_emb_a['ensembl_gene_id'] = human_rm['ensembl_gene_id'].values\n",
    "human_emb = human_emb.fillna(0)\n",
    "for c in human_emb.columns:\n",
    "    if 'entrez_id' not in c:\n",
    "        try:\n",
    "            human_emb_a[c] = np.log2(human_emb[c].values + 1)\n",
    "            human_emb_num[c] = np.log2(human_emb[c].values + 1)\n",
    "        except:\n",
    "            print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "60d7813a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pos_h = human_emb[human_emb['ensembl_gene_id'].isin(df_pos_human)]\n",
    "df_pos_h.set_index('gene_symbol', inplace=True)\n",
    "df_pos_h = df_pos_h[[c for c in df_pos_h.columns if 'entrez' not in c and 'ensembl' not in c]]\n",
    "df_pos_h = df_pos_h.T.sample(frac=1) # Randomly shuffle the data\n",
    "\n",
    "df_neg_h = human_emb[human_emb['ensembl_gene_id'].isin(df_neg_human)]\n",
    "df_neg_h.set_index('gene_symbol', inplace=True)\n",
    "df_neg_h = df_neg_h[[c for c in df_neg_h.columns if 'entrez' not in c and 'ensembl' not in c]]\n",
    "\n",
    "df_neg_h = df_neg_h.T.sample(frac=1)\n",
    "\n",
    "df_tfs_h = human_emb[human_emb['gene_symbol'].isin(human_tfs)]\n",
    "df_tfs_h.set_index('gene_symbol', inplace=True)\n",
    "df_tfs_h = df_tfs_h[[c for c in df_tfs_h.columns if 'entrez' not in c and 'ensembl' not in c]]\n",
    "\n",
    "df_tfs_h = df_tfs_h.T.sample(frac=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "5c39b8ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'AMY', 'M1C', 'DTH', 'HIP', 'URL', 'S1C', 'Ocx', 'CB', 'DFC', 'M1C-S1C', 'VFC', 'STR', 'IPC', 'CBC', 'ITC', 'TCx', 'V1C', 'LGE', 'CGE', 'PCx', 'MFC', 'A1C', 'OFC', 'MGE', 'STC', 'MD'}\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                  VAE identified genes on tissue:\t(588, 466)\t                   \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.4\n",
      "0.1\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                   TFs identified genes on tissue:\t(588, 84)\t                   \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.1398148148148148\n",
      "0.12037037037037036\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m               Consistently affected genes on tissue:\t(588, 738)\t               \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.2490196078431372\n",
      "0.17647058823529413\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ariane/opt/miniconda3/envs/scivae_10/lib/python3.10/site-packages/sklearn/metrics/_classification.py:2006: UserWarning: y_pred contains classes not in y_true\n",
      "  warnings.warn(\"y_pred contains classes not in y_true\")\n",
      "/Users/ariane/opt/miniconda3/envs/scivae_10/lib/python3.10/site-packages/sklearn/metrics/_classification.py:2006: UserWarning: y_pred contains classes not in y_true\n",
      "  warnings.warn(\"y_pred contains classes not in y_true\")\n",
      "/Users/ariane/opt/miniconda3/envs/scivae_10/lib/python3.10/site-packages/sklearn/metrics/_classification.py:2006: UserWarning: y_pred contains classes not in y_true\n",
      "  warnings.warn(\"y_pred contains classes not in y_true\")\n"
     ]
    }
   ],
   "source": [
    "from scivae import Validate\n",
    "from sciutil import SciUtil\n",
    "\n",
    "label = 'tissue'\n",
    "label_i = 0\n",
    "df_pos_labels = [i.split('_')[label_i] for i in df_pos_h.index]\n",
    "df_neg_labels = [i.split('_')[label_i] for i in df_neg_h.index]\n",
    "df_tfs_labels = [i.split('_')[label_i] for i in df_tfs_h.index]\n",
    "print(set(df_tfs_labels))\n",
    "\n",
    "\n",
    "u = SciUtil()\n",
    "u.dp([f'VAE identified genes on {label}:', df_pos.shape])\n",
    "vd = Validate(df_pos_h.values, df_pos_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'TFs identified genes on {label}:', df_tfs.shape])\n",
    "vd = Validate(df_tfs_h.values, df_tfs_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'Consistently affected genes on {label}:', df_neg.shape])\n",
    "vd = Validate(df_neg_h.values, df_neg_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "7be8160d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'12-pcw', '24-pcw', '25-pcw', '26-pcw', '16-pcw', '19-pcw', '21-pcw', '8-pcw', '17-pcw', '9-pcw', '35-pcw', '13-pcw', '37-pcw'}\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                    VAE identified genes on age:\t(588, 466)\t                    \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.8001443001443\n",
      "0.22857142857142856\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                    TFs identified genes on age:\t(588, 84)\t                     \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.59004329004329\n",
      "0.29696969696969694\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                Consistently affected genes on age:\t(588, 738)\t                 \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.6833333333333333\n",
      "0.41666666666666663\n"
     ]
    }
   ],
   "source": [
    "from scivae import Validate\n",
    "from sciutil import SciUtil\n",
    "\n",
    "label = 'age'\n",
    "label_i = 1\n",
    "df_pos_labels = [i.split('_')[label_i] for i in df_pos_h.index]\n",
    "df_neg_labels = [i.split('_')[label_i] for i in df_neg_h.index]\n",
    "df_tfs_labels = [i.split('_')[label_i] for i in df_tfs_h.index]\n",
    "print(set(df_tfs_labels))\n",
    "\n",
    "\n",
    "u = SciUtil()\n",
    "u.dp([f'VAE identified genes on {label}:', df_pos.shape])\n",
    "vd = Validate(df_pos_h.values, df_pos_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'TFs identified genes on {label}:', df_tfs.shape])\n",
    "vd = Validate(df_tfs_h.values, df_tfs_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'Consistently affected genes on {label}:', df_neg.shape])\n",
    "vd = Validate(df_neg_h.values, df_neg_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "501b95ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'W4', 'W1', 'W2', 'W3'}\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                    VAE identified genes on age:\t(588, 466)\t                    \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.791517857142857\n",
      "0.5563988095238095\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                    TFs identified genes on age:\t(588, 84)\t                     \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.7840909090909091\n",
      "0.4409090909090909\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                Consistently affected genes on age:\t(588, 738)\t                 \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.94375\n",
      "0.81875\n"
     ]
    }
   ],
   "source": [
    "from scivae import Validate\n",
    "from sciutil import SciUtil\n",
    "\n",
    "label = 'age'\n",
    "label_i = 2\n",
    "df_pos_labels = [i.split('_')[label_i] for i in df_pos_h.index]\n",
    "df_neg_labels = [i.split('_')[label_i] for i in df_neg_h.index]\n",
    "df_tfs_labels = [i.split('_')[label_i] for i in df_tfs_h.index]\n",
    "print(set(df_tfs_labels))\n",
    "\n",
    "\n",
    "u = SciUtil()\n",
    "u.dp([f'VAE identified genes on {label}:', df_pos.shape])\n",
    "vd = Validate(df_pos_h.values, df_pos_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'TFs identified genes on {label}:', df_tfs.shape])\n",
    "vd = Validate(df_tfs_h.values, df_tfs_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'Consistently affected genes on {label}:', df_neg.shape])\n",
    "vd = Validate(df_neg_h.values, df_neg_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "31980092",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'M', 'F'}\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                    VAE identified genes on sex:\t(588, 466)\t                    \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.825\n",
      "0.5\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                    TFs identified genes on sex:\t(588, 84)\t                     \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.8307692307692307\n",
      "0.676923076923077\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m                Consistently affected genes on sex:\t(588, 738)\t                 \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "0.8701298701298701\n",
      "0.8928571428571428\n"
     ]
    }
   ],
   "source": [
    "from scivae import Validate\n",
    "from sciutil import SciUtil\n",
    "\n",
    "label = 'sex'\n",
    "label_i = 3\n",
    "df_pos_labels = [i.split('_')[label_i] for i in df_pos_h.index]\n",
    "df_neg_labels = [i.split('_')[label_i] for i in df_neg_h.index]\n",
    "df_tfs_labels = [i.split('_')[label_i] for i in df_tfs_h.index]\n",
    "print(set(df_tfs_labels))\n",
    "\n",
    "\n",
    "u = SciUtil()\n",
    "u.dp([f'VAE identified genes on {label}:', df_pos.shape])\n",
    "vd = Validate(df_pos_h.values, df_pos_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'TFs identified genes on {label}:', df_tfs.shape])\n",
    "vd = Validate(df_tfs_h.values, df_tfs_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n",
    "\n",
    "u.dp([f'Consistently affected genes on {label}:', df_neg.shape])\n",
    "vd = Validate(df_neg_h.values, df_neg_labels)\n",
    "print(vd.predict('rf', 'balanced_accuracy'))\n",
    "print(vd.predict('svm', 'balanced_accuracy'))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
